{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c921c4-4457-4409-9ced-9309fbfe322f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#05 统计bayes的每个样本reads数，写成A256G的形式，过滤，只过滤测序深度,不过滤频率。\n",
    "#统计每个样本的突变数量\n",
    "#dp>100并且在至少2个样本中出现过的才算。\n",
    "#变异类型是snp或者del或ins，这里也筛选了。\n",
    "#最终输出csv文件。\n",
    "\n",
    "import csv\n",
    "import os\n",
    "import vcf\n",
    "wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "for i in wenjianjia_list:\n",
    "    dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{i}'\n",
    "    if not os.path.exists(dirs):\n",
    "        os.makedirs(dirs)\n",
    "\n",
    "for p in os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair'):\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair/{p}'\n",
    "    \n",
    "    bayes_filenames=os.listdir(prj_path)\n",
    "    type_list= []\n",
    "    dict_mut_num = {}#在几个样本中出现过。形成列表，在列表中的才算真正的突变。\n",
    "    #找出突变在几个样本中出现\n",
    "    for i in bayes_filenames:\n",
    "        if i.endswith('vcf'):\n",
    "            i_1 = i.split('.')[0]\n",
    "            reads_pos_dict = {}\n",
    "            path1 = prj_path+'/'+i\n",
    "            vcf_reader = vcf.Reader(filename=path1)\n",
    "            for record in vcf_reader:\n",
    "    #             print(record)\n",
    "    #             print(record.INFO)\n",
    "\n",
    "                alt_count = record.INFO['AO']\n",
    "                ref_count = record.INFO['RO']\n",
    "                ref = str(record).split(',')[2].split('=')[1]\n",
    "                type0 = record.INFO['TYPE']\n",
    "                position = str(record).split(',')[1].split('=')[1]\n",
    "                dp = record.INFO['DP']\n",
    "                if type0 not in type_list:\n",
    "                    type_list.append(type0)\n",
    "                for i in range(0,len(record.INFO['TYPE'])):\n",
    "                    alt1 = str(record).split('=')[4].replace(')','').replace('[','').replace(']','').replace(' ','')\n",
    "                    alt2 = alt1.split(\",\")#转成列表,碱基\n",
    "                    type2 = record.INFO['TYPE'][i]\n",
    "                    info = f'{ref}_{position}_{alt2[i]}'\n",
    "                    alt_count1 = alt_count[i]\n",
    "                    af = record.INFO['AO'][i]/(record.INFO['AO'][i] + record.INFO['RO'])\n",
    "    #                 print(af)\n",
    "                    if dp > 100:\n",
    "    #                     reads_pos_dict[info] = {'alt':alt_count1,'ref':ref_count,'type':type2,'allele_frequency':af}\n",
    "                        if info not in dict_mut_num:\n",
    "                            dict_mut_num[info] = 1\n",
    "                        else:\n",
    "                            dict_mut_num[info] += 1\n",
    "    #         print(reads_pos_dict)\n",
    "\n",
    "\n",
    "\n",
    "    filtered_list = []# 测序深度>100且出现在至少2个样本中的突变。\n",
    "    for k,v in dict_mut_num.items():\n",
    "        if v > 1:\n",
    "            filtered_list.append(k)\n",
    "    # print(filtered_list)\n",
    "\n",
    "\n",
    "    #根据filtered_list筛选突变，输出csv。\n",
    "    for i in bayes_filenames:\n",
    "        if i.endswith('vcf'):\n",
    "            i_1 = i.split('.')[0]\n",
    "            reads_pos_dict = {}\n",
    "            path1 = prj_path +'/'+i\n",
    "            vcf_reader = vcf.Reader(filename=path1)\n",
    "            for record in vcf_reader:\n",
    "    #             print(record)\n",
    "    #             print(record.INFO)\n",
    "\n",
    "                alt_count = record.INFO['AO']\n",
    "                ref_count = record.INFO['RO']\n",
    "                ref = str(record).split(',')[2].split('=')[1]\n",
    "                type0 = record.INFO['TYPE']\n",
    "                position = str(record).split(',')[1].split('=')[1]\n",
    "                dp = record.INFO['DP']\n",
    "                if type0 not in type_list:\n",
    "                    type_list.append(type0)\n",
    "                for i in range(0,len(record.INFO['TYPE'])):\n",
    "                    alt1 = str(record).split('=')[4].replace(')','').replace('[','').replace(']','').replace(' ','')\n",
    "                    alt2 = alt1.split(\",\")#转成列表,碱基\n",
    "                    type2 = record.INFO['TYPE'][i]\n",
    "                    info = f'{ref}_{position}_{alt2[i]}'\n",
    "                    alt_count1 = alt_count[i]\n",
    "                    af = record.INFO['AO'][i]/(record.INFO['AO'][i] + record.INFO['RO'])    \n",
    "                    if dp > 100:#即使有列表了也还是需要测序深度>100的这个条件，不然有可能是出现在样本多但是测序深度不行的。\n",
    "                        if info in filtered_list:\n",
    "                            if type2 == 'snp' or type2 == 'del' or type2 == 'ins':#筛选变异类型\n",
    "                                reads_pos_dict[info] = {'alt':alt_count1,'ref':ref_count,'type':type2,'allele_frequency':af,'sample_exist':dict_mut_num[info]}\n",
    "    #         print(reads_pos_dict)\n",
    "\n",
    "            with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{p}/{i_1}.csv','w') as f:#表格可以加个频率\n",
    "                writer = csv.writer(f)\n",
    "                writer.writerow([\"mutation\",\"alt_reads\",\"ref_reads\",\"type\",\"app\",\"allele_frequency\",\"sample_exist'\"])\n",
    "                for k,v in reads_pos_dict.items():\n",
    "                    writer.writerow([k,v['alt'],v['ref'],v['type'],'freebayes',v['allele_frequency'],v['sample_exist']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc88bc62-de72-4e38-8e3d-29acee0b6d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install pyvcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e32282e-e97a-46a1-8379-cb4848602bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "#04统计突变类型，3%-80%和80%以上的比例，先不管allele frequency，只看次数\n",
    "#二代测序中3%-80%的突变，对频率有错误影响，对蛋白分布没有影响。可以加一列属于哪个样本的。加上了，次数是准的了。\n",
    "\n",
    "#80%以下：\n",
    "import os\n",
    "import csv\n",
    "\n",
    "# wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "# for i in wenjianjia_list:\n",
    "#     dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/{i}'\n",
    "#     if not os.path.exists(dirs):\n",
    "#         os.makedirs(dirs)\n",
    "\n",
    "# wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "# for i in wenjianjia_list:\n",
    "#     dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/07mutation_proportion/{i}'\n",
    "#     if not os.path.exists(dirs):\n",
    "#         os.makedirs(dirs)\n",
    "\n",
    "for p in os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample'):\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{p}'\n",
    "\n",
    "    dict_003_08_next_generation = {}\n",
    "    for i in os.listdir(prj_path):\n",
    "        if i.endswith('.csv'):\n",
    "            csv_path = f'{prj_path}/{i}'\n",
    "            with open(csv_path) as csvfile:\n",
    "                csv_reader = csv.reader(csvfile)  \n",
    "                for row in csv_reader:\n",
    "                    if row[0] != 'mutation':\n",
    "                        first = row[0].split('_')[0]\n",
    "    #                     last = row[0].split('_')[2]\n",
    "    #                     position = row[0].split('_')[1]\n",
    "                        frequnecy = float(row[5])\n",
    "                        mutation = row[0]\n",
    "                        type1 = row[3]\n",
    "                        if 0.03 <= frequnecy <= 0.8:\n",
    "    #                         if len(first) == 1:#找的snv很奇怪，明明是单个突变非要写一堆，所以就不用type筛选了。\n",
    "    #                             dict_003_08_next_generation[mutation] = frequnecy\n",
    "                            if mutation not in dict_003_08_next_generation.keys():\n",
    "                                dict_003_08_next_generation[mutation] = {}\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "                            else:#一定要加，不然每次都是赋值一个空字典就没办法新加进来样本。\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "    # print(dict_003_08_next_generation)\n",
    "    # print(p)\n",
    "# ----------------------------------------------------------------------------------------------------------    \n",
    "# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!统计所有的突变及其数量，频率。 后面看在哪个突变上要用。             \n",
    "    with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06isnv_mutation_{p}.csv','w') as f:\n",
    "        writer = csv.writer(f)\n",
    "        writer.writerow(['isnv','sample','frequency','type'])\n",
    "        for k,v in dict_003_08_next_generation.items():\n",
    "            for s,info in v.items():\n",
    "                frequency = info[0]\n",
    "                type3 = info[1]\n",
    "                writer.writerow([k,s.split('.')[0],frequency,type3])\n",
    "            \n",
    "#-----------------------------------------------------------------------------------------------------------\n",
    "    \n",
    "#统计数量\n",
    "    dict_num_mutation = {'A_G':0,'A_T':0,'A_C':0,'G_A':0,'G_C':0,'G_T':0,'T_C':0,'T_A':0,'T_G':0,'C_A':0,'C_T':0,'C_G':0,'del':0,'ins':0}\n",
    "    for k,v in dict_003_08_next_generation.items():\n",
    "        ref = k.split('_')[0]\n",
    "        alt = k.split('_')[2]\n",
    "        for s in v.values():\n",
    "            type2 = s[1]\n",
    "    #         print(type2)\n",
    "        if type2 =='snp':\n",
    "            name = f'{ref}_{alt}'\n",
    "            add_count = len(v)\n",
    "        #     print(name)\n",
    "            for m,c in dict_num_mutation.items():\n",
    "                if name == m:\n",
    "                    dict_num_mutation[m] += add_count  #!!!赋值不可以只写c，一定要写成键的值。\n",
    "        if type2 =='del':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['del'] += add_count\n",
    "        if type2 =='ins':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['ins'] += add_count\n",
    "\n",
    "\n",
    "\n",
    "    print(dict_num_mutation)\n",
    "    count_all = 0\n",
    "    for k,v in dict_num_mutation.items():\n",
    "        count_all += v\n",
    "    print(count_all)\n",
    "    \n",
    "    dict_proportion =  dict.fromkeys(dict_num_mutation.keys())\n",
    "    for k in dict_proportion.keys():\n",
    "        dict_proportion[k] = {}\n",
    "        dict_proportion[k]['count'] = dict_num_mutation[k]\n",
    "        if count_all!= 0:\n",
    "            dict_proportion[k]['proportion'] = dict_proportion[k]['count']/count_all\n",
    "        else:\n",
    "            dict_proportion[k]['proportion'] = '/'\n",
    "    print(dict_proportion)  \n",
    "    with open(f\"/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/07mutation_proportion/07isnv_mutation_type_{p}.csv\",\"w\") as csvfile: \n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow(['mutation_type','count','proportion','group'])\n",
    "        for k,v in dict_proportion.items():\n",
    "            writer.writerow([k,v['count'],v['proportion'],'iSNV'])\n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "# #-----------------------------------------------------------------------\n",
    "# #80%上：\n",
    "\n",
    "    dict_003_08_next_generation = {}\n",
    "    for i in os.listdir(prj_path):\n",
    "        if i.endswith('.csv'):\n",
    "            csv_path = f'{prj_path}/{i}'\n",
    "            with open(csv_path) as csvfile:\n",
    "                csv_reader = csv.reader(csvfile)  \n",
    "                for row in csv_reader:\n",
    "                    if row[0] != 'mutation':\n",
    "                        first = row[0].split('_')[0]\n",
    "    #                     last = row[0].split('_')[2]\n",
    "    #                     position = row[0].split('_')[1]\n",
    "                        frequnecy = float(row[5])\n",
    "                        mutation = row[0]\n",
    "                        type1 = row[3]\n",
    "                        if  frequnecy > 0.8:\n",
    "    #                         if len(first) == 1:#找的snv很奇怪，明明是单个突变非要写一堆，所以就不用type筛选了。\n",
    "    #                             dict_003_08_next_generation[mutation] = frequnecy\n",
    "                            if mutation not in dict_003_08_next_generation.keys():\n",
    "                                dict_003_08_next_generation[mutation] = {}\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "                            else:#一定要加，不然每次都是赋值一个空字典就没办法新加进来样本。\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "\n",
    "    #统计数量\n",
    "    dict_num_mutation = {'A_G':0,'A_T':0,'A_C':0,'G_A':0,'G_C':0,'G_T':0,'T_C':0,'T_A':0,'T_G':0,'C_A':0,'C_T':0,'C_G':0,'del':0,'ins':0}\n",
    "    for k,v in dict_003_08_next_generation.items():\n",
    "        ref = k.split('_')[0]\n",
    "        alt = k.split('_')[2]\n",
    "        for s in v.values():\n",
    "            type2 = s[1]\n",
    "    #         print(type2)\n",
    "        if type2 =='snp':\n",
    "            name = f'{ref}_{alt}'\n",
    "            add_count = len(v)\n",
    "        #     print(name)\n",
    "            for m,c in dict_num_mutation.items():\n",
    "                if name == m:\n",
    "                    dict_num_mutation[m] += add_count  #!!!赋值不可以只写c，一定要写成键的值。\n",
    "        if type2 =='del':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['del'] += add_count\n",
    "        if type2 =='ins':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['ins'] += add_count\n",
    "\n",
    "\n",
    "\n",
    "    # print(dict_num_mutation)\n",
    "\n",
    "    #算比例的分母\n",
    "    count_all = 0\n",
    "    for k,v in dict_num_mutation.items():\n",
    "        count_all += v\n",
    "    # print(count_all)\n",
    "    dict_proportion =  dict.fromkeys(dict_num_mutation.keys())\n",
    "    for k in dict_proportion.keys():\n",
    "        dict_proportion[k] = {}\n",
    "        dict_proportion[k]['count'] = dict_num_mutation[k]\n",
    "        if count_all!= 0:\n",
    "            dict_proportion[k]['proportion'] = dict_proportion[k]['count']/count_all\n",
    "        else:\n",
    "            dict_proportion[k]['proportion'] = '/'\n",
    "    print(dict_proportion)  \n",
    "    with open(f\"/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/07mutation_proportion/07snp_mutation_type_{p}.csv\",\"w\") as csvfile: \n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow(['mutation_type','count','proportion','group'])\n",
    "        for k,v in dict_proportion.items():\n",
    "            writer.writerow([k,v['count'],v['proportion'],'SNP'])\n",
    "\n",
    "    #!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!统计所有的突变及其数量，频率。 后面看在哪个突变上要用。             \n",
    "    with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06snp_mutation_{p}.csv','w') as f:\n",
    "        writer = csv.writer(f)\n",
    "        writer.writerow(['isnv','sample','frequency','type'])\n",
    "        for k,v in dict_003_08_next_generation.items():\n",
    "            for s,info in v.items():\n",
    "                frequency = info[0]\n",
    "                type3 = info[1]\n",
    "                writer.writerow([k,s.split('.')[0],frequency,type3])\n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2eb0b72-f9f2-4a4a-9c2c-fa798d036898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 05 翻译蛋白质，要加3-80%的。isnv的\n",
    "#分批次\n",
    "#非同义氨基酸突变的具体情况和不同蛋白非同义突变的数量。\n",
    "#插入加了。\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "for i in wenjianjia_list:\n",
    "    dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{i}'\n",
    "    if not os.path.exists(dirs):\n",
    "        os.makedirs(dirs)\n",
    "dict_pep_cds = {'99_518':'NS1_1C','628_1002':\"NS2_1B\",'1140_2315':\"N\",'2348_3073':\"P\",'3233_4003':\"M\",'4274_4468':\"SH\",'4659_5555':\"G\",'5632_7356':\"F\",'7576_8160':\"M2\",'8468_14965':\"L\"}\n",
    "\n",
    "csv_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist')\n",
    "for c in csv_list:\n",
    "    if c.endswith('csv'):\n",
    "        c1 = c.split('_')[2].split('.')[0]\n",
    "        csv_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06isnv_mutation_{c1}.csv'\n",
    "        with open(csv_path) as csvfile:\n",
    "            dict_info = {}\n",
    "            info_list = []\n",
    "            csv_reader = csv.reader(csvfile)  \n",
    "            for row in csv_reader:\n",
    "                if row[0] != 'isnv':\n",
    "                    mut = row[0]\n",
    "                    if mut not in info_list:\n",
    "                        info_list.append(mut)\n",
    "        dict_info = dict.fromkeys(info_list)\n",
    "\n",
    "\n",
    "        for i in dict_info.keys():#嵌套字典可以这样赋值：1、dict_info = dict.fromkeys(info_list)2、每个键都赋值成空字典，再正常加就好了\n",
    "            dict_info[i] = {}\n",
    "\n",
    "\n",
    "    #----------------\n",
    "    #新加的，统计每个突变的样本数量。\n",
    "        df = pd.read_csv(csv_path)\n",
    "        df_counts = df.isnv.value_counts()\n",
    "        #print(df_counts)\n",
    "        outputpath=f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/mut_sample_number/{c1}_mut_sample_number.csv'\n",
    "        df_counts.to_csv(outputpath,sep=',',index=True,header=False)\n",
    "\n",
    "        dict_mut_num = {}\n",
    "        \n",
    "        mut_number_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/mut_sample_number/{c1}_mut_sample_number.csv'\n",
    "        with open(mut_number_path) as csvfile:\n",
    "            csv_reader = csv.reader(csvfile) \n",
    "            for row in csv_reader:\n",
    "                isnv = row[0]\n",
    "                num = row[1]\n",
    "                dict_mut_num[isnv] = num\n",
    "        # print(dict_mut_num)\n",
    "\n",
    "\n",
    "    #--------------\n",
    "        with open(csv_path) as csvfile:\n",
    "            csv_reader = csv.reader(csvfile)  \n",
    "            for row in csv_reader:\n",
    "                if row[0] != 'isnv':\n",
    "                    mut = row[0]\n",
    "                    pos = int(row[0].split('_')[1])\n",
    "                    fre = float(row[2])\n",
    "                    type4 = row[3]\n",
    "                    first = row[0].split('_')[0]\n",
    "                    last = row[0].split('_')[2]\n",
    "                    #分一下位置，del和ins的有可能结束位置和开始位置不在同一段蛋白里面。不太可能\n",
    "                    if type4 == 'del':#按照突变前的长度算\n",
    "                        pos_end = pos + len(first) -1\n",
    "                    if type4 == 'ins':#按照突变后的长度算\n",
    "                        pos_end = pos + len(last) -1          \n",
    "                    for k,v in dict_pep_cds.items():\n",
    "                        start = int(k.split('_')[0])\n",
    "                        end = int(k.split('_')[1])\n",
    "                        if type4 == 'snp':\n",
    "                            if start <= pos <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "                                dict_info[mut]['product'] = v\n",
    "                                dict_info[mut]['type'] = type4\n",
    "                                dict_info[mut]['start'] = start\n",
    "                                dict_info[mut]['end'] = end\n",
    "                                dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "        #                         dict_info[mut]['frequency'] = fre 频率先不考虑了，有好多样本的\n",
    "                        if type4 == 'del' or type4 == 'ins':\n",
    "                            if start <= pos <= end and start <= pos_end <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "                                dict_info[mut]['product'] = v\n",
    "                                dict_info[mut]['type'] = type4\n",
    "                                dict_info[mut]['start'] = start\n",
    "                                dict_info[mut]['end'] = end\n",
    "        #                         dict_info[mut]['frequency'] = fre\n",
    "                                dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # print(dict_info)\n",
    "        ##---------------------------------------------------------------------------------------\n",
    "        #去除不在cds范围内的碱基突变                   \n",
    "        dict_info_feikong = {}            \n",
    "        for i in dict_info.keys():\n",
    "            if dict_info[i]:\n",
    "                dict_info_feikong[i] = dict_info[i]\n",
    "        # print(dict_info_feikong)\n",
    "\n",
    "        #获取突变前后的三个碱基，蛋白位置不知道，可以用什么蛋白\n",
    "        dict_aa_mut = {}#非同义突变的字典\n",
    "        dict_aa_synonymous_mut = {}\n",
    "        ref_seq = 'ACGCGAAAAAATGCGTACAACAAACTTGCGTAAACCAAAAAAATGGGGCAAATAAGAATTTGATAAGTACCACTTAAATTTAACTCCCTTGGTTAGAGATGGGCAGCAATTCATTGAGTATGATAAAAGTTAGATTACAAAATTTGTTTGACAATGATGAAGTAGCATTGTTAAAAATAACATGCTATACTGACAAATTAATACATTTAACTAATGCATTGGCTAAGGCAGTGATACATACAATCAAATTGAATGGCATTGTATTTGTGCATGTTATTACAAGTAGTGATATTTGCCCTAATAATAATATTGTAGTGAAATCCAATTTCACAACAATGCCAGTGTTACAAAATGGAGGTTATATATGGGAAATGATGGAATTAACACACTGCTCTCAACCTAATGGCCTAATAGATGACAATTGTGAAATTAAATTCTCCAAAAAACTAAGTGATTCAACAATGACCAATTATATGAATCAATTATCTGAATTACTTGGATTTGATCTTAATCCATAAATTATAATAAATATCAACTAGCAAATCAATGTCACTAACACCATTAGTTAATATAAAACTTGACAGAAGATAAAAATGGGGCAAATAAATCAATTCAGCCGACCCAACCATGGACACAACACACAATGACACCACACCACAAAGACTGATGATCACAGACATGAGACCATTGTCACTTGAGACTATAATAATATCACTAACCAGAGACATCATAACACACAGATTTATATACTTGATAAATCATGAATGTATAGTGAGAAAACTTGATGAAAGACAGGCCACATTTACATTCCTGGTCAACTATGAAATGAAACTATTGCACAAAGTGGGAAGCACTAAATACAAAAAATATACTGAATACAACACAAAATATGGCACTTTTCCTATGCCAATATTTATCAATCATGATGGGTTCTTAGAATGCATTGGCATTAAGCCTACAAAGCACACTCCCATAATATACAAGTATGATCTCAATCCATGAATTTCAACACAAGAGTCACACAATCTGAAATAACAACTTCATGCATAACCACACTCCATAGTTCAAATGGAGCCTGAAAATTATAGTAATTTAAAATTAAGGAGAGACATAAGATGAAAGATGGGGCAAATACAAAAATGGCTCTTAGCAAAGTCAAGTTGAATGATACACTCAACAAAGATCAACTTCTGTCATCCAGCAAATACACCATCCAACGGAGCACAGGAGATAGTATTGATACTCCTAATTATGATGTGCAGAAACACATCAATAAGTTATGTGGCATGTTATTAATCACAGAAGATGCTAATCATAAATTCACTGGGTTAATAGGTATGTTATATGCTATGTCTAGATTAGGAAGAGAAGACACCATAAAAATACTCAGAGATGCGGGATATCATGTAAAAGCAAATGGAGTGGATGTAACAACACATCGTCAAGATATTAATGGGAAAGAAATGAAATTTGAAGTGTTAACATTGTCAAGCTTAACAACTGAAATTCAAATCAACATTGAGATAGAATCTAGAAAATCCTACAAAAAAATGCTAAAAGAAATGGGAGAGGTAGCTCCAGAATACAGGCATGACTCTCCTGATTGTGGGATGATAATATTATGTATAGCGGCATTAGTAATAACCAAATTAGCAGCAGGGGATAGATCTGGTCTTACAGCTGTGATTAGGAGGGCTAATAATGTCCTAAAAAATGAAATGAAACGTTATAAAGGCTTACTACCCAAGGATATAGCCAACAGCTTCTATGAAGTGTTTGAAAAATATCCTCACTTTATAGATGTTTTTGTTCATTTTGGTATAGCACAATCTTCTACCAGAGGTGGCAGTAGAGTTGAAGGGATTTTTGCTGGATTGTTTATGAATGCCTATGGTGCAGGGCAAGTGATGTTACGGTGGGGGGTCTTAGCAAAATCAGTTAAAAATATTATGCTAGGACACGCTAGTGTGCAAGCAGAAATGGAACAAGTTGTGGAGGTTTATGAATATGCCCAAAAATTGGGTGGAGAAGCAGGGTTCTACCATATATTGAACAACCCAAAAGCATCATTATTGTCTTTGACTCAATTTCCTCACTTCTCCAGTGTAGTATTAGGCAATGCTGCTGGCCTAGGCATAATGGGAGAATACAGAGGTACACCAAGGAATCAAGATCTATATGATGCTGCAAAAGCATATGCTGAACAACTCAAAGAAAATGGTGTGATTAACTACAGTGTATTAGACTTGACAGCAGAAGAACTAGAGGCTATCAAACATCAGCTTAATCCAAAAGATAATGATGTAGAGCTTTGAGTTAATAAAAAAAATGGGGCAAATAAAACATCATGGAAAAGTTTGCTCCTGAATTCCATGGAGAAGATGCAAACAACAGAGCTACCAAATTCCTAGAATCAATAAAGGGCAAATTCACATCACCTAAAGATCCCAAGAAAAAAGATAGTATCATATCTGTCAACTCAATAGATATAGAAGTAACCAAAGAAAGCCCTATAACATCAAATTCAACCATTATAAACCCAACAAATGAGACAGATGATACTGTAGGGAACAAGCCCAATTATCAAAGAAAACCTCTAGTAAGTTTCAAAGAAGACCCTACGCCAAGTGATAATCCCTTTTCAAAACTATACAAAGAAACCATAGAAACATTTGATAACAATGAAGAAGAATCTAGCTATTCATATGAAGAAATAAATGATCAGACAAACGATAATATAACAGCAAGATTAGATAGGATTGATGAAAAATTAAGTGAAATACTAGGAATGCTTCACACATTAGTAGTAGCGAGTGCAGGACCTACATCTGCTCGGGATGGTATAAGAGATGCCATGGTTGGTTTAAGAGAAGACATGATAGAAAAAATCAGAACTGAAGCATTAATGACCAATGACAGACTAGAAGCTATGGCAAGACTCAGGAATGAGGAAAGTGAAAAGATGGCAAAAGACACATCAGATGAAGTGTCTCTCAATCCAACATCAGAGAAATTGAACAACCTGTTGGAAGGGAATGATAGTGACAATGATCTATCACTTGATGATTTCTGATCAGTTACCAATCTGTACATCAACACACAACACCAACAGAAGACCAACAAACAAACCAACTCACCCATCCAACCAAACATCTATACGCCAATCAGCCAATCCAAAACTAGCCACCCGGAAAAAATAGATACTATAGTTACAAAAAAAGATGGGGCAAATATGGAAACATACGTGAACAAACTTCACGAAGGCTCCACATACACAGCTGCTGTTCAATACAATGTCTTAGAAAAAGACGATGACCCTGCATCACTTACAATATGGGTGCCCATGTTCCAATCATCCATGCCAGCAGATTTACTTATAAAAGAACTAGCTAATGTCAACATACTAGTGAAACAAATATCCACACCCAATGGACCTTCATTAAGAGTCATGATAAACTCAAGAAGTGCAGTGCTAGCACAAATGCCCAGCAAATTTACCATATGTGCCAATGTGTCCTTGGATGAAAGAAGCAAGCTGGCATATGATGTAACCACACCCTGTGAAATCAAGGCATGTAGTCTAACATGCCTAAAATCAAAAAATATGTTAACTACAGTTAAAGATCTCACTATGAAAACACTCAACCCAACACATGACATCATTGCTTTATGTGAATTTGAAAATATAGTAACATCAAAAAAAGTCATAATACCAACATACCTAAGATCCATCAGTGTCAGAAATAAAGATCTGAACACACTTGAAAATATAACAACCACTGAATTCAAAAATGCCATCACAAATGCAAAAATCATCCCTTACTCAGGATTACTGTTAGTCATCACAGTGACTGACAACAAAGGAGCATTCAAATACATAAAGCCACAAAGTCAATTTATAGTAGATCTTGGAGCTTACCTAGAAAAAGAAAGTATATATTATGTTACAACAAATTGGAAGCACACAGCTACACGATTTGCAATCAAACCCATGGAAGATTAACCTTTTTCTTCTACATCAGTGAGTTGATTCATACAAACTTTCTACCTACATTCTTCACTTCACCATCATAATCACCAACCCTCTGTGGTTCAACTAATCAAACAAAACCCATCTGGAGCCTCAGATCATCCCAAGTCATTGTTCATCAGATCTAGTACTCAAATAAGTTAATAAAAATATCCACATGGGGCAAATAATCATTGGAGGAAATCCAACTAATCACAATATCTGTCAACATAGACAAGTCAACACGCCAGGCAAAATCAACCAATGGAAAATACATCCATAACAATAGAATTCTCAAGCAAATTCTGGCCTTACTTTACACTAATACACATGATAACAACAATAATCTCTTTGCTAATCATAATCTCCATCATGATTGCAATACTGAACAAACTCTGTGAATATAACGTATTCCATAACAAAACCTTTGAGCTACCAAGAGCTCGAGTCAATACATAGCATTCACCAATCTGATGGCACAAAACAGTAACCTTGCATTTGTAAGTGAACAACCCTCACCTCTTTACAAAACCACATCAACATCTCACCATGCAAGCCATCATCCATATTATAAAGTAGTTAATTAAAAATAATCATAACAATGAACTAAGATATTAAGACTAACAATAACGTTGGGGCAAATGCAAACATGTCCAAAAACAAGGACCAACGCACCACCAAGACACTAGAAAAGACCTGGGACACTCTCAATCATCTATTATTCATATCATCGTGCTTATACAAGTTAAATCTTAAATCTATAGCACAAATCACATTATCCATTCTGGCAATGATAATCTCAACTTCACTTATAATTGCAGCCATCATATTCATAGCCTCGGCAAACCACAAAGTCACACTAACAACTGCAATCATACAAGATGCAACAAGCCAGATCAAGAACACAACCCCAACATACCTCACCCAGAATCCCCAGCTTGGAATCAGCTTCTCCAATCTGTCTGAAACTACATCACAAACCACCACCATACTAGCTTCAACAACACCAAGTGTCAAGTCAACCCTGCAATCCACAACAGTCAAGACCAAAAACACAACAACAACCAAAATACAACCCAGCAAGCCCACCACAAAACAACGCCAAAACAAACCACCAAACAAACCCAATAATGATTTTCACTTTGAAGTGTTCAACTTTGTACCTTGCAGCATATGCAGCAACAATCCAACCTGCTGGGCTATCTGTAAAAGAATACCAAACAAAAAACCTGGAAAGAAAACCACCACCAAGCCCACAAAAAAACCAACCATCAAGACAACCAAAAAAGATCTCAAACCTCAAACCACAAAACCAAAGGAAGTACCTACCACCAAGCCCACAGAAAAGCCAACCATCAACACCACCAAAACAAACATCAGAACTACACTGCTCACCAACAATACCACAGGAAATCCAGAACACACAAGTCAAAAGGGAACCCTCCACTCAACCTCCTCCGATGGCAATCCAAGCCCTTCACAAGTCTATACAACATCCGAGTACCTATCACAACCTCCATCTCCATCCAACACAACAAACCAGTAGTCATTAAAAAGCGTATTATTGCAAAAAGCCATGACCAAATCAACCAGAATCAAAATCAACTCTGGGGCAAATAACAATGGAGTTGCCAATCCTCAAAACAAATGCAATTACCGCAATCCTTGCTGCAGTCACACTCTGTTTTGCTTCCAGTCAAAACATCACTGAAGAATTTTATCAAACAACATGCAGTGCAGTCAGCAAAGGCTATCTTAGTGCTCTAAGAACTGGTTGGTATACTAGTGTTATAACTATAGAATTAAGTAATATCAAGGAAAATAAGTGTAATGGAACAGACGCTAAGGTAAAATTGATAAAACAAGAATTAGATAAATATAAAAGTGCTGTAACAGAATTGCAGTTGCTCATGCAAAGCACACCGGCAACCAACAATCGAGCCAGAAGAGAACTACCAAGGTTTATGAATTATACACTCAACAATACCAAAAATACCAATGTAACATTAAGCAAGAAAAGGAAAAGAAGATTTCTTGGCTTTTTGTTAGGTGTTGGATCTGCAATCGCCAGTGGCATTGCTGTATCTAAGGTCCTGCACCTAGAAGGGGAAGTGAACAAAATCAAAAGTGCTCTACTATCCACAAACAAGGCTGTAGTCAGCTTATCAAATGGAGTTAGTGTCTTAACCAGCAAAGTGTTAGACCTCAAAAACTATATAGATAAACAGTTGTTACCTATTGTGAACAAGCAAAGCTGTAGCATATCAAACATTGAAACTGTGATAGAGTTCCAACAAAAGAACAACAGACTACTAGAGATTACCAGGGAATTTAGTGTTAATGCAGGTGTAACTACACCTGTAAGCACTTATATGTTAACAAATAGTGAATTATTATCATTAATCAATGATATGCCTATAACAAATGATCAGAAAAAGTTAATGTCCAACAATGTTCAAATAGTTAGACAGCAAAGTTACTCTATCATGTCCATAATAAAGGAGGAAGTCTTAGCATATGTAGTACAATTACCACTATATGGTGTAATAGATACACCTTGTTGGAAACTGCACACATCCCCTCTATGTACAACCAACACAAAGGAAGGGTCCAACATCTGTTTAACAAGAACCGACAGAGGATGGTACTGTGACAATGCAGGATCAGTATCTTTCTTCCCACTAGCTGAAACATGTAAAGTTCAATCGAATCGAGTATTTTGTGACACAATGAACAGTTTAACATTACCAAGTGAAGTAAATCTCTGCAACATTGACATATTCAACCCCAAATATGATTGCAAAATTATGACTTCAAAAACAGATGTAAGCAGCTCCGTTATCACATCTCTAGGAGCCATTGTGTCATGCTATGGCAAAACTAAATGTACAGCATCCAATAAAAATCGTGGAATCATAAAGACATTTTCTAACGGGTGCGATTATGTATCAAATAAGGGGGTTGACACTGTGTCTGTAGGTAATACATTATATTATGTAAATAAGCAAGAAGGCAAAAGTCTCTATGTAAAAGGTGAACCAATAATAAATTTCTATGACCCATTAGTGTTCCCCTCTGATGAATTTGATGCATCAATATCTCAAGTCAATGAGAAGATTAACCAGAGCCTAGCATTTATTCGTAAATCCGATGAATTATTACATAATGTAAATGCTGGTAAATCCACCATAAATATCATGATAACTACTATAATTATAGTGATTATAGTAATATTGTTATCATTAATTGCCGTTGGACTGCTCCTATACTGCAAGGCCAGAAGCACACCAGTCACACTAAGCAAGGATCAACTGAGTGGTATAAATAATATTGCATTTAGTAACTAAATAAAAATAGCACCTAATCATGTTCTTACAATGGTTTCATATCTGCTCATAGACAACCCATCTATCATTGGATTTTCTTAAAATCTGAACTTCATCGAAACTCTCATCTATAAACCATCTCACTTACATTATTTAAGTAGATTCCTAGTTTATAGTTATATAAAACAATTGAATACCAGATTAACTTACTATTTGTAAAAAATGAGAACTGGGGCAAATATGTCACGAAGGAATCCTTGCAAATTTGAAATTCGAGGTCATTGCTTGAATGGTAAGAGGTGTCATTTTAGTCATAATTATTTTGAATGGCCACCCCATGCACTGCTTGTAAGACAAAACTTTATGTTAAACAGAATACTTAAGTCTATGGATAAAAGCATCGATACTTTATCAGAAATAAGTGGAGCTGCAGAGTTGGACAGAACAGAAGAGTATGCCCTCGGTGTAGTTGGAGTGCTAGAGAGTTATATAGGATCTATAAATAATATAACTAAACAATCAGCATGTGTTGCCATGAGCAAACTCCTCACTGAACTCAACAGTGATGACATCAAAAAACTGAGGGACAATGAAGAGCCAAATTCACCCAAGATAAGAGTGTACAATACTGTCATATCATATATTGAAAGCAACAGGAAAAACAATAAACAAACTATCCATCTGTTAAAAAGATTGCCAGCAGACGTATTGAAGAAAACCATAAAAACCACATTGGATATCCACAAGAGCATAACCATCAATAACCCAAAAGAATCAACTGTTAGTGATATAAACGACCATGCCAAAAATAATGATACTACCTGACAAATATCCTTGTAGTATAAATTCCATACTAATAACAAGTAGTTGTAGAGTTACTATGTATAATCAAAAGAACACACTATATTTCAATCAAAACAACCAAAATAACCATATATACTCACCGAATCAACCATTCAATGAAATCCATTGGACCTCTCAAGACTTGATTGATGCAATTCAAAATTTTCTACAACATCTAGGTATTACTGATGATATATACACAATATATATATTAGTGTCATAACACTCAATCCTAATGCTTACCACATCATCAAACTATTAACTCAAACAATTCAAGCCATGGGACAAAATGGATCCCATTATTAATGGAAATTCTGCTAATGTGTATCTAACCGATAGTTATTTAAAAGGTGTTATTTCTTTCTCAGAATGTAATGCTTTAGGAAGTTACATATTCAATGGTCCTTATCTCAAAAATGATTATACCAACTTAATTAGTAGACAAAATCCATTAATAGAACACATAAATCTAAAGAAACTAAATATAACACAGTCCTTAATGTCTAAGTATCATAAAGGTGAAATAAAAATAGAAGAACCTACTTATTTTCAGTCATTACTTATGACATACAAGAGTATGACCTCGTTAGAACAGATTACTACCACTAATTTACTTAAAAAGATAATAAGAAGAGCTATAGAAATTAGTGATGTCAAAGTCTATGCTATATTGAATAAACTGGGGCTTAAAGAAAAAGACAAGATTAAATCCAACAATGGACAAGATGAAGACAACTCAGTTATTACAACCATAATCAAAGATGATATACTTTTAGCTGTTAAGGATAATCAATCTCATCTTAAAGCAGTCAAAAATCACTCTACAAAACAAAAAGATACAATCAAAACAACACTCTTGAAGAAATTAATGTGTTCAATGCAACATCCTCCATCATGGTTAATACATTGGTTTAATTTATACACAAAATTAAACAACATATTAACACAGTATCGATCAAGTGAGGTAAAAAACCATGGTTTTATATTGATAGACAATCATACTCTCAATGGATTCCAATTTATTTTGAATCAATATGGTTGTATAGTTTATCATAAGGAACTCAAAAGAATTACTGTGACAACCTATAATCAATTCTTGACATGGAAAAATATTAGCCTTAGTAGATTAAATGTTTGTTTAATTACATGGATTAGTAACTGTTTGAACACATTAAATAAAAGCTTAGGCTTAAGATGCGGATTCAATAATGTTATCTTGACACAACTATTCCTCTATGGAGATTGTATACTAAAACTATTCCACAATGAGGGGTTCTACATAATAAAAGAGGTAGAGGGATTTATTATGTCTCTAATTTTAAATATAACAGAAGAAGATCAATTCAGAAAACGGTTTTATAATAGTATGCTCAACAACATCACAGATGCTGCTAATAAAGCTCAGAAAAATCTGCTATCAAGAGTATGTCATACATTATTAGATAAGACAGTATCCGATAATATAATAAATGGCAGATGGATAATTCTATTAAGTAAGTTCCTTAAATTAATTAAGCTTGCAGGTGACAATAACCTTAACAATCTGAGTGAATTATATTTTTTGTTCAGAATATTTGGACACCCAATGGTAGATGAAAGACAAGCCATGGATGCTGTTAAAGTTAATTGCAACGAGACCAAATTTTACTTGTTAAGCAGTTTGAGTATGTTAAGAGGTGCCTTTATATATAGAATTATAAAAGGATTTGTAAATAATTACAACAGATGGCCTACTTTAAGGAATGCTATTGTTTTACCCTTAAGATGGTTAACTTACTATAAACTAAACACTTATCCTTCCTTGTTGGAACTTACAGAAAGAGATTTGATTGTTTTATCAGGACTACGTTTCTATCGTGAGTTTCGGTTGCCTAAAAAAGTGGATCTTGAAATGATCATAAATGATAAGGCTATATCACCTCCTAAAAATTTGATATGGACTAGTTTCCCTAGAAATTATATGCCGTCACACATACAAAATTATATAGAACATGAAAAATTAAAATTTTCCGAGAGTGATAAATCAAGAAGAGTATTAGAGTACTATTTAAGAGATAACAAATTCAATGAATGTGATTTATACAACTGTGTAGTTAATCAAAGTTATCTTAACAACCCTAATCATGTGGTATCTTTGACAGGCAAAGAAAGAGAACTCAGTGTAGGTAGAATGTTTGCAATGCAACCAGGAATGTTCAGACAAGTTCAAATATTAGCAGAGAAAATGATAGCTGAAAACATTTTACAATTCTTTCCTGAAAGTCTTACAAGATATGGTGATCTAGAACTACAGAAAATATTAGAATTGAAAGCAGGAATAAGTAACAAATCAAATCGTTACAATGATAATTACAACAATTACATTAGTAAGTGCTCTATCATCACAGATCTCAGCAAATTCAATCAAGCATTTCGATATGAAACATCATGTATTTGTAGTGATGTACTGGATGAACTGCATGGTGTACAATCTCTATTTTTCTGGTTACATTTAGCTATTCCTCATGTCACAATAATATGCACATATAGGCATGCACCCCCCTATATAAGAGATCATATTGTAGATCTTAACAATGTAGATGAACAAAGTGGATTATATAGATATCATATGGGTGGTATCGAAGGGTGGTGTCAAAAACTATGGACCATAGAAGCTATATCACTATTGGATCTAATATCTCTCAAAGGGAAATTCTCAATTACTGCTTTAATTAATGGTGACAATCAATCAATAGATATAAGTAAACCAGTCAGACTCATGGAAGGTCAAACTCATGCTCAAGCAGATTATTTGCTAGCATTAAATAGTCTTAAATTACTGTATAAAGAGTATGCAGGCATAGGCCACAAATTAAAAGGAACTGAGACTTATATATCAAGAGATATGCAATTTATGAGTAAAACAATTCAACATAACGGTGTATATTACCCAGCTAGTATAAAGAAAGTCCTAAGAGTGGGACCGTGGATAAACACTATACTTGATGATTTCAAAGTGAGTCTAGAATCTATAGGTAGTTTGACACAAGAATTAGAATATAGAGGAGAAAGTCTATTATGCAGTTTAATATTTAGAAATGTATGGTTATATAATCAAATTGCTTTACAACTAAAAAATCATGCATTATGTAACAATAAATTATATTTGGACATATTAAAGGTTCTGAAACACTTAAAAACCTTTTTTAATCTTGATAATATTGATACAGCATTAACATTGTATATGAATTTGCCCATGTTATTTGGTGGTGGTGATCCCAACTTGTTATATCGAAGTTTCTATAGAAGAACTCCTGATTTCCTCACAGAGGCTATAGTTCACTCTGTGTTCATACTTAGTTATTATACAAACCATGATTTAAAAGATAAACTTCAAGATCTGTCAGATGATAGATTGAATAAGTTCTTAACATGCATAATCACATTTGACAAAAACCCTAATGCTGAATTCGTAACATTGATGAGAGATCCTCAAGCTTTAGGGTCTGAGAGACAAGCTAAAATTACTAGCGAAATCAATAGACTGGCAGTTACTGAGGTTTTGAGCACAGCTCCAAACAAAATATTCTCCAAAAGTGCACAACACTATACCACTACAGAGATAGATCTAAATGATATTATGCAAAATATAGAACCTACATATCCTCACGGGCTAAGAGTTGTTTATGAAAGTTTACCCTTTTATAAAGCAGAGAAAATAGTAAATCTTATATCCGGTACAAAATCTATAACTAACATACTGGAAAAGACTTCTGCCATAGACTTAACAGATATTGATAGAGCCACTGAGATGATGAGGAAAAACATAACTTTGCTTATAAGGATATTTCCATTAGATTGTAACAGAGATAAAAGGGAAATATTGAGTATGGAAAACCTAAGTATTACTGAATTAAGCAAATATGTTAGGGAAAGATCTTGGTCTTTATCCAATATAGTTGGTGTTACATCACCTAGTATCATGTATACAATGGACATCAAATATACAACAAGCACTATAGCTAGTGGCATAATCATAGAGAAATATAATGTTAACAGTTTAACACGTGGTGAGAGAGGACCCACTAAACCATGGGTTGGTTCATCTACACAAGAGAAAAAAACAATGCCAGTTTATAATAGACAAGTTTTAACCAAAAAACAGAGAGATCAAATAGATCTATTAGCAAAATTGGATTGGGTGTATGCATCTATAGATAACAAGGATGAATTCATGGAAGAACTTAGCATAGGAATTCTTGGGTTAACATATGAGAAAGCCAAAAAATTATTTCCACAATATTTAAGTGTTAACTATTTGCATCGCCTTACAGTCAGTAGTAGACCATGTGAATTCCCTGCATCAATACCAGCTTATAGAACTACAAATTATCACTTTGATACTAGCCCTATTAATCGCATATTAACAGAAAAGTATGGTGATGAAGATATTGATATAGTATTCCAAAACTGTATAAGCTTTGGCCTTAGCTTAATGTCAGTAGTAGAACAATTTACTAATGTATGTCCTAACAGAATTATTCTTATACCTAAGCTTAATGAGATACATTTAATGAAACCTCCCATATTCACAGGTGATGTTGATATTCACAAGTTAAAACAAGTGATACAAAAACAGCATATGTTTTTACCAGACAAAATAAGTTTGACTCAATATGTGGAATTATTCTTAAGTAATAAAACACTCAAATCTGGATCTCATGTTAATTCTAATTTAATATTGGCGCATAAGATATCTGACTATTTTCATAATACTTACATTTTAAGTACTAATTTAGCTGGACATTGGATTCTGATTATACAACTTATGAAAGATTCTAAGGGTATTTTTGAAAAAGATTGGGGAGAGGGATATATAACTGATCATATGTTCATTAATTTGAAAGTTTTCTTCAATGCTTATAAGACCTATCTCTTGTGTTTTCATAAAGGTTACGGCAGAGCAAAGCTGGAGTGTGATATGAATACTTCAGATCTCCTATGTGTATTGGAATTAATAGACAGTAGTTATTGGAAGTCTATGTCTAAGGTATTTTTAGAACAAAAAGTTATCAAATACATTCTTAGCCAGGATGCAAGTTTACATAGAGTAAAAGGATGTCATAGCTTCAAACTATGGTTTCTTAAACGTCTTAATGTAGCAGAATTCACAGTTTGCCCTTGGGTTGTTAACATAGATTATCATCCAACACATATGAAAGCAATATTAACTTATATAGATCTTGTTAGAATGGGATTGATAAATATAGATAAAATATACATTAAAAATAAACACAAATTCAATGATGAATTTTATACTTCTAATCTCTTTTACATTAATTATAACTTCTCAGATAATACTCATCTATTAACTAAACATATAAGGATTGCTAATTCTGAATTAGAAAATAATTACAACAAATTATATCATCCTACACCAGAAACCCTAGAAAATATACTAACCAATCCGGTTAAATGTAATGACAAAAAGACACTGAATGACTATTGTATAGGTAAAAATGTTGACTCAATAATGTTACCATTGTTATCTAATAAGAAGCTTATTAAATCGCCTACAATGATTAGAACCAATTACAGCAAACAAGATTTGTATAATTTATTTCCTACGGTTGTGATTGATAAAATTATAGATCATTCAGGTAATACAGCCAAATCTAACCAACTTTACACTACTACTTCTCATCAAATACCTTTAGTGCACAATAGCACATCACTTTATTGCATGCTTCCTTGGCATCATATTAATAGATTCAATTTTGTATTTAGTTCTACAGGTTGTAAAATTAGTATAGAGTATATTTTAAAAGACCTTATAATTAAAGATCCTAATTGTATAGCATTCATAGGTGAAGGAGCAGGGAATTTATTATTGCGTACAGTAGTGGAACTTCATCCCGATATAAGATATATTTACAGAAGTCTGAAGGATTGCAATGATCATAGTTTACCTATTGAGTTTTTAAGGCTGTACAATGGACATATCAACATTGATTATGGTGAAAATTTGACCATTCCTGCTACAGATGCAACCAACAACATTCATTGGTCTTATTTACATATAAAGTTTGCTGAACCTATCAGTCTTTTTGTCTGTGATGCTGAATTGCCTGTAACAGTCAACTGGAGTAAAATTATAATAGAGTGGAGCAAGCATGTAAGAAAATGCAAGTACTGTTCCTCAGTTAATAAATGTACGTTAATAGTAAAATATCATGCTCAAGATGATATCGATTTCAAATTAGACAATATAACTATATTAAAAACTTATGTATGCTTAGGCAGTAAGTTAAAGGGGTCTGAAGTTTACTTAGTCCTTACAATAGGTCCTGCAAATGTGTTCCCAGTATTTAATGTAGTACAAAATGCTAAATTGATACTATCAAGAACCAAAAATTTCATCATGCCTAAGAAGGCTGATAAAGAGTCTATTGATGCAAATATTAAAAGTTTGATACCCTTTCTTTGTTACCCTATAACAAAAAAAGGAATTAATACTGCATTGTCAAAACTAAAGAGTGTTGTTAGTGGAGATATACTATCATATTCTATAGCAGGACGTAATGAAGTTTTCAGCAATAAACTTATAAATCATAAGCATATGAACATCTTAAAATGGTTCAATCATGTTTTAAATTTCAGATCAACAGAACTAAACTATAATCATTTATATATGGTAGAATCTACATATCCTTATCTAAGTGAATTGTTAAACAGCTTGACAACTAATGAACTTAAAAAACTGATTAAAATCACAGGTAGTTTGTTATACAACTTTCATAATGAATAATGAATAAAAATCTTATATTAAAAATTCCCATAGCTACACACTAACACTGTATTCAATTATAGTTATTTAAAATTAAAAATTATATAATTTTTTAATAACTTTTAGTGAACTAATCCTAAAATTATCATTTTGATCTAGGAGGAATAAATTTAAATCCAAATCTAATTGGTTTATATGTATATTAACTAAACTACGAGATATTAGTTTTTGACACTTTTTTTCTCGT'\n",
    "        ref_seq_list = list(ref_seq)\n",
    "        dict_jianji = {}\n",
    "        for k,v in dict_info_feikong.items():\n",
    "            gene_pos = int(k.split('_')[1])\n",
    "            ref = k.split('_')[0]\n",
    "            alt = k.split('_')[2]\n",
    "            product = v['product']\n",
    "        #     frequency = v['frequency']\n",
    "            gene_pos_f2 = gene_pos - 2   #前面2个\n",
    "            gene_pos_f1 = gene_pos - 1   #前面1个\n",
    "            gene_pos_b1 = gene_pos + 1\n",
    "            gene_pos_b2 = gene_pos + 2\n",
    "            start = v['start']\n",
    "            end = v['end']\n",
    "            type4 = v['type']\n",
    "            chazhi = int(gene_pos) - int(start)\n",
    "            quzheng =  chazhi//3   #取整，以获取是这个蛋白质的第几个氨基酸。\n",
    "            aa_pos = quzheng + 1\n",
    "\n",
    "        # #分类讨论\n",
    "            if type4 == 'del' or type4 == 'ins':\n",
    "                del_count = len(ref) - len(alt)\n",
    "                if del_count%3 == 0:\n",
    "                    #分类讨论\n",
    "                    #1\n",
    "                    if chazhi%3 == 2 and len(alt)%3 == 2:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "                   #2\n",
    "                    if chazhi%3 == 2 and len(alt)%3 == 1:\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "\n",
    "                    #3\n",
    "                    if chazhi%3 == 2 and len(alt)%3 == 0:\n",
    "                    #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "\n",
    "                    #4\n",
    "                    if chazhi%3 == 1 and len(alt)%3 == 2:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "\n",
    "                    #5\n",
    "                    if chazhi%3 == 1 and len(alt)%3 == 1:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                #6\n",
    "                    if chazhi%3 == 1 and len(alt)%3 == 0:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "\n",
    "                    #7\n",
    "                    if chazhi%3 == 0 and len(alt)%3 == 2:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian]) \n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "\n",
    "                    #8\n",
    "                    if chazhi%3 == 0 and len(alt)%3 == 1:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "                #9\n",
    "                    if chazhi%3 == 0 and len(alt)%3 == 0:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "\n",
    "\n",
    "\n",
    "          #------------------------原来的，底下可以不变  \n",
    "            if type4 == 'snp':\n",
    "                if chazhi%3 == 2:\n",
    "                    list_jianji_pos = []#是位置不是切片，切片要-1\n",
    "                    list_jianji_pos.append(gene_pos_f2)\n",
    "                    list_jianji_pos.append(gene_pos_f1)\n",
    "                    list_jianji_pos.append(gene_pos)\n",
    "\n",
    "\n",
    "\n",
    "                if chazhi%3 == 1:\n",
    "                    list_jianji_pos = []\n",
    "                    list_jianji_pos.append(gene_pos_f1)\n",
    "                    list_jianji_pos.append(gene_pos)\n",
    "                    list_jianji_pos.append(gene_pos_b1)\n",
    "\n",
    "\n",
    "\n",
    "                if chazhi%3 == 0:\n",
    "                    list_jianji_pos = []\n",
    "                    list_jianji_pos.append(gene_pos)\n",
    "                    list_jianji_pos.append(gene_pos_b1)\n",
    "                    list_jianji_pos.append(gene_pos_b2)\n",
    "\n",
    "\n",
    "                list_jianji_qiepian = []#这个是切片\n",
    "                for i in list_jianji_pos:\n",
    "                    i_1 = i-1\n",
    "                    list_jianji_qiepian.append(i_1)\n",
    "            #     print(list_jianji_qiepian)\n",
    "\n",
    "                list_jianji_ref = []#获取参考碱基\n",
    "                for i in list_jianji_qiepian:\n",
    "                    list_jianji_ref.append(ref_seq_list[i])\n",
    "            #     print(list_jianji_ref)\n",
    "\n",
    "                #获取alt碱基\n",
    "                list_jianji_alt= []#获取参考碱基\n",
    "                for i in list_jianji_qiepian:\n",
    "                    if i == int(gene_pos - 1):\n",
    "                        list_jianji_alt.append(alt)\n",
    "                    else:\n",
    "                        list_jianji_alt.append(ref_seq_list[i])\n",
    "            #     print(list_jianji_alt)\n",
    "            #翻译蛋白质\n",
    "            str_ref = \"\".join(list_jianji_ref)\n",
    "            str_alt = \"\".join(list_jianji_alt)\n",
    "            aa_ref = str(Seq(str_ref).translate())\n",
    "            aa_alt = str(Seq(str_alt).translate())    \n",
    "            #非同义：\n",
    "            if aa_alt != aa_ref:\n",
    "                info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "                dict_aa_mut[k] = {}\n",
    "                dict_aa_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "                dict_aa_mut[k]['aa'] = info\n",
    "                dict_aa_mut[k]['num'] = dict_mut_num[k]\n",
    "            #同义\n",
    "            if aa_alt == aa_ref:\n",
    "                info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "                dict_aa_synonymous_mut[k] = {}\n",
    "                dict_aa_synonymous_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "                dict_aa_synonymous_mut[k]['aa'] = info\n",
    "                dict_aa_synonymous_mut[k]['num'] = dict_mut_num[k]\n",
    "        #             dict_aa_mut[k]['frequency'] = dict_info_feikong[k]['frequency']\n",
    "\n",
    "        # \n",
    "\n",
    "        #非同义----------------------------------------------------------------------\n",
    "        #统计各个突变的频率。\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_nonsynonymous_mut_frequency.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "            for k,v in dict_aa_mut.items():\n",
    "                writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "\n",
    "\n",
    "        #统计非同义突变在哪个蛋白上nonsynonymous mutation\n",
    "        list_product = []\n",
    "        for k,v in dict_aa_mut.items():\n",
    "                if v['product'] not in list_product:\n",
    "                    list_product.append(v['product'])\n",
    "                    dict_product = dict.fromkeys(list_product)#每个蛋白上非同义突变的个数\n",
    "        for k in dict_product.keys():\n",
    "            dict_product[k] = 0\n",
    "        for m,p in dict_aa_mut.items():\n",
    "            for k,v in dict_product.items():\n",
    "                if p['product'] == k:\n",
    "                    dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "        # print(dict_product)\n",
    "        dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "        # print(dict_product)\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_nonsynonymous_product_number.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['product','number','group'])\n",
    "            for k,v in dict_product.items():\n",
    "                writer.writerow([k,v,'nonsynonymous'])\n",
    "\n",
    "\n",
    "        #同义----------------------------------------------------------------------\n",
    "        #统计各个突变的频率。\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_synonymous_mut_frequency.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "            for k,v in dict_aa_synonymous_mut.items():\n",
    "                writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "\n",
    "\n",
    "\n",
    "        #统计同义突变在哪个蛋白上\n",
    "        list_product = []\n",
    "        for k,v in dict_aa_synonymous_mut.items():\n",
    "                if v['product'] not in list_product:\n",
    "                    list_product.append(v['product'])\n",
    "                    dict_product = dict.fromkeys(list_product)\n",
    "        for k in dict_product.keys():\n",
    "            dict_product[k] = 0\n",
    "        for m,p in dict_aa_synonymous_mut.items():\n",
    "            for k,v in dict_product.items():\n",
    "                if p['product'] == k:\n",
    "                    dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "        # print(dict_product)\n",
    "        dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "        # print(dict_product)\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_synonymous_product_number.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['product','number','group'])\n",
    "            for k,v in dict_product.items():\n",
    "                writer.writerow([k,v,'synonymous'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef4ea9bb-2b7e-4357-82b2-18fab9ea172e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06snp_mutation_PRJNA438443.csv')\n",
    "df_counts = df.isnv.value_counts()\n",
    "print(df_counts)\n",
    "# outputpath=f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/mut_sample_number/{c1}_mut_sample_number.csv'\n",
    "# df_counts.to_csv(outputpath,sep=',',index=True,header=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fb0ed4-0f3f-457c-8e36-37a4f09a381b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#加入每个突变的样本数量.不分批次的\n",
    "# 05 翻译蛋白质，要加3-80%的。isnv的\n",
    "#非同义氨基酸突变的具体情况和不同蛋白非同义突变的数量。\n",
    "#插入没加先。\n",
    "import csv\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "dict_info = {}\n",
    "info_list = []\n",
    "dict_pep_cds = {'99_518':'NS1_1C','628_1002':\"NS2_1B\",'1140_2315':\"N\",'2348_3073':\"P\",'3233_4003':\"M\",'4274_4468':\"SH\",'4659_5555':\"G\",'5632_7356':\"F\",'7576_8160':\"M2\",'8468_14965':\"L\"}\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/06isnv_mutation.csv') as csvfile:\n",
    "    csv_reader = csv.reader(csvfile)  \n",
    "    for row in csv_reader:\n",
    "        if row[0] != 'isnv':\n",
    "            mut = row[0]\n",
    "            if mut not in info_list:\n",
    "                info_list.append(mut)\n",
    "dict_info = dict.fromkeys(info_list)\n",
    "\n",
    "\n",
    "for i in dict_info.keys():#嵌套字典可以这样赋值：1、dict_info = dict.fromkeys(info_list)2、每个键都赋值成空字典，再正常加就好了\n",
    "    dict_info[i] = {}\n",
    "#新加的，统计每个突变的样本数量。\n",
    "df = pd.read_csv('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/06isnv_mutation.csv')\n",
    "df_counts = df.isnv.value_counts()\n",
    "# print(df_counts)\n",
    "outputpath=f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/mut_sample_number.csv'\n",
    "df_counts.to_csv(outputpath,sep=',',index=True,header=False)\n",
    "\n",
    "dict_mut_num = {}\n",
    "\n",
    "mut_number_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/mut_sample_number.csv'\n",
    "with open(mut_number_path) as csvfile:\n",
    "    csv_reader = csv.reader(csvfile) \n",
    "    for row in csv_reader:\n",
    "        isnv = row[0]\n",
    "        num = row[1]\n",
    "        dict_mut_num[isnv] = num    \n",
    "\n",
    "    \n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/06isnv_mutation.csv') as csvfile:\n",
    "    csv_reader = csv.reader(csvfile)  \n",
    "    for row in csv_reader:\n",
    "        if row[0] != 'isnv':\n",
    "            mut = row[0]\n",
    "            pos = int(row[0].split('_')[1])\n",
    "            fre = float(row[2])\n",
    "            type4 = row[3]\n",
    "            first = row[0].split('_')[0]\n",
    "            last = row[0].split('_')[2]\n",
    "            #分一下位置，del和ins的有可能结束位置和开始位置不在同一段蛋白里面。不太可能\n",
    "            if type4 == 'del':#按照突变前的长度算\n",
    "                pos_end = pos + len(first) -1\n",
    "            if type4 == 'ins':#按照突变后的长度算\n",
    "                pos_end = pos + len(last) -1          \n",
    "            for k,v in dict_pep_cds.items():\n",
    "                start = int(k.split('_')[0])\n",
    "                end = int(k.split('_')[1])\n",
    "                if type4 == 'snp':\n",
    "                    if start <= pos <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "                        dict_info[mut]['product'] = v\n",
    "                        dict_info[mut]['type'] = type4\n",
    "                        dict_info[mut]['start'] = start\n",
    "                        dict_info[mut]['end'] = end\n",
    "                        dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "#                         dict_info[mut]['frequency'] = fre 频率先不考虑了，有好多样本的\n",
    "                if type4 == 'del' or type4 == 'ins':\n",
    "                    if start <= pos <= end and start <= pos_end <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "                        dict_info[mut]['product'] = v\n",
    "                        dict_info[mut]['type'] = type4\n",
    "                        dict_info[mut]['start'] = start\n",
    "                        dict_info[mut]['end'] = end\n",
    "                        dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "#                         dict_info[mut]['frequency'] = fre\n",
    "\n",
    "                                       \n",
    "# print(dict_info)\n",
    "#去除不在cds范围内的碱基突变                   \n",
    "dict_info_feikong = {}            \n",
    "for i in dict_info.keys():\n",
    "    if dict_info[i]:\n",
    "        dict_info_feikong[i] = dict_info[i]\n",
    "# print(dict_info_feikong)\n",
    "\n",
    "#获取突变前后的三个碱基，蛋白位置不知道，可以用什么蛋白\n",
    "dict_aa_mut = {}#非同义突变的字典\n",
    "dict_aa_synonymous_mut = {}\n",
    "ref_seq = 'ACGCGAAAAAATGCGTACAACAAACTTGCGTAAACCAAAAAAATGGGGCAAATAAGAATTTGATAAGTACCACTTAAATTTAACTCCCTTGGTTAGAGATGGGCAGCAATTCATTGAGTATGATAAAAGTTAGATTACAAAATTTGTTTGACAATGATGAAGTAGCATTGTTAAAAATAACATGCTATACTGACAAATTAATACATTTAACTAATGCATTGGCTAAGGCAGTGATACATACAATCAAATTGAATGGCATTGTATTTGTGCATGTTATTACAAGTAGTGATATTTGCCCTAATAATAATATTGTAGTGAAATCCAATTTCACAACAATGCCAGTGTTACAAAATGGAGGTTATATATGGGAAATGATGGAATTAACACACTGCTCTCAACCTAATGGCCTAATAGATGACAATTGTGAAATTAAATTCTCCAAAAAACTAAGTGATTCAACAATGACCAATTATATGAATCAATTATCTGAATTACTTGGATTTGATCTTAATCCATAAATTATAATAAATATCAACTAGCAAATCAATGTCACTAACACCATTAGTTAATATAAAACTTGACAGAAGATAAAAATGGGGCAAATAAATCAATTCAGCCGACCCAACCATGGACACAACACACAATGACACCACACCACAAAGACTGATGATCACAGACATGAGACCATTGTCACTTGAGACTATAATAATATCACTAACCAGAGACATCATAACACACAGATTTATATACTTGATAAATCATGAATGTATAGTGAGAAAACTTGATGAAAGACAGGCCACATTTACATTCCTGGTCAACTATGAAATGAAACTATTGCACAAAGTGGGAAGCACTAAATACAAAAAATATACTGAATACAACACAAAATATGGCACTTTTCCTATGCCAATATTTATCAATCATGATGGGTTCTTAGAATGCATTGGCATTAAGCCTACAAAGCACACTCCCATAATATACAAGTATGATCTCAATCCATGAATTTCAACACAAGAGTCACACAATCTGAAATAACAACTTCATGCATAACCACACTCCATAGTTCAAATGGAGCCTGAAAATTATAGTAATTTAAAATTAAGGAGAGACATAAGATGAAAGATGGGGCAAATACAAAAATGGCTCTTAGCAAAGTCAAGTTGAATGATACACTCAACAAAGATCAACTTCTGTCATCCAGCAAATACACCATCCAACGGAGCACAGGAGATAGTATTGATACTCCTAATTATGATGTGCAGAAACACATCAATAAGTTATGTGGCATGTTATTAATCACAGAAGATGCTAATCATAAATTCACTGGGTTAATAGGTATGTTATATGCTATGTCTAGATTAGGAAGAGAAGACACCATAAAAATACTCAGAGATGCGGGATATCATGTAAAAGCAAATGGAGTGGATGTAACAACACATCGTCAAGATATTAATGGGAAAGAAATGAAATTTGAAGTGTTAACATTGTCAAGCTTAACAACTGAAATTCAAATCAACATTGAGATAGAATCTAGAAAATCCTACAAAAAAATGCTAAAAGAAATGGGAGAGGTAGCTCCAGAATACAGGCATGACTCTCCTGATTGTGGGATGATAATATTATGTATAGCGGCATTAGTAATAACCAAATTAGCAGCAGGGGATAGATCTGGTCTTACAGCTGTGATTAGGAGGGCTAATAATGTCCTAAAAAATGAAATGAAACGTTATAAAGGCTTACTACCCAAGGATATAGCCAACAGCTTCTATGAAGTGTTTGAAAAATATCCTCACTTTATAGATGTTTTTGTTCATTTTGGTATAGCACAATCTTCTACCAGAGGTGGCAGTAGAGTTGAAGGGATTTTTGCTGGATTGTTTATGAATGCCTATGGTGCAGGGCAAGTGATGTTACGGTGGGGGGTCTTAGCAAAATCAGTTAAAAATATTATGCTAGGACACGCTAGTGTGCAAGCAGAAATGGAACAAGTTGTGGAGGTTTATGAATATGCCCAAAAATTGGGTGGAGAAGCAGGGTTCTACCATATATTGAACAACCCAAAAGCATCATTATTGTCTTTGACTCAATTTCCTCACTTCTCCAGTGTAGTATTAGGCAATGCTGCTGGCCTAGGCATAATGGGAGAATACAGAGGTACACCAAGGAATCAAGATCTATATGATGCTGCAAAAGCATATGCTGAACAACTCAAAGAAAATGGTGTGATTAACTACAGTGTATTAGACTTGACAGCAGAAGAACTAGAGGCTATCAAACATCAGCTTAATCCAAAAGATAATGATGTAGAGCTTTGAGTTAATAAAAAAAATGGGGCAAATAAAACATCATGGAAAAGTTTGCTCCTGAATTCCATGGAGAAGATGCAAACAACAGAGCTACCAAATTCCTAGAATCAATAAAGGGCAAATTCACATCACCTAAAGATCCCAAGAAAAAAGATAGTATCATATCTGTCAACTCAATAGATATAGAAGTAACCAAAGAAAGCCCTATAACATCAAATTCAACCATTATAAACCCAACAAATGAGACAGATGATACTGTAGGGAACAAGCCCAATTATCAAAGAAAACCTCTAGTAAGTTTCAAAGAAGACCCTACGCCAAGTGATAATCCCTTTTCAAAACTATACAAAGAAACCATAGAAACATTTGATAACAATGAAGAAGAATCTAGCTATTCATATGAAGAAATAAATGATCAGACAAACGATAATATAACAGCAAGATTAGATAGGATTGATGAAAAATTAAGTGAAATACTAGGAATGCTTCACACATTAGTAGTAGCGAGTGCAGGACCTACATCTGCTCGGGATGGTATAAGAGATGCCATGGTTGGTTTAAGAGAAGACATGATAGAAAAAATCAGAACTGAAGCATTAATGACCAATGACAGACTAGAAGCTATGGCAAGACTCAGGAATGAGGAAAGTGAAAAGATGGCAAAAGACACATCAGATGAAGTGTCTCTCAATCCAACATCAGAGAAATTGAACAACCTGTTGGAAGGGAATGATAGTGACAATGATCTATCACTTGATGATTTCTGATCAGTTACCAATCTGTACATCAACACACAACACCAACAGAAGACCAACAAACAAACCAACTCACCCATCCAACCAAACATCTATACGCCAATCAGCCAATCCAAAACTAGCCACCCGGAAAAAATAGATACTATAGTTACAAAAAAAGATGGGGCAAATATGGAAACATACGTGAACAAACTTCACGAAGGCTCCACATACACAGCTGCTGTTCAATACAATGTCTTAGAAAAAGACGATGACCCTGCATCACTTACAATATGGGTGCCCATGTTCCAATCATCCATGCCAGCAGATTTACTTATAAAAGAACTAGCTAATGTCAACATACTAGTGAAACAAATATCCACACCCAATGGACCTTCATTAAGAGTCATGATAAACTCAAGAAGTGCAGTGCTAGCACAAATGCCCAGCAAATTTACCATATGTGCCAATGTGTCCTTGGATGAAAGAAGCAAGCTGGCATATGATGTAACCACACCCTGTGAAATCAAGGCATGTAGTCTAACATGCCTAAAATCAAAAAATATGTTAACTACAGTTAAAGATCTCACTATGAAAACACTCAACCCAACACATGACATCATTGCTTTATGTGAATTTGAAAATATAGTAACATCAAAAAAAGTCATAATACCAACATACCTAAGATCCATCAGTGTCAGAAATAAAGATCTGAACACACTTGAAAATATAACAACCACTGAATTCAAAAATGCCATCACAAATGCAAAAATCATCCCTTACTCAGGATTACTGTTAGTCATCACAGTGACTGACAACAAAGGAGCATTCAAATACATAAAGCCACAAAGTCAATTTATAGTAGATCTTGGAGCTTACCTAGAAAAAGAAAGTATATATTATGTTACAACAAATTGGAAGCACACAGCTACACGATTTGCAATCAAACCCATGGAAGATTAACCTTTTTCTTCTACATCAGTGAGTTGATTCATACAAACTTTCTACCTACATTCTTCACTTCACCATCATAATCACCAACCCTCTGTGGTTCAACTAATCAAACAAAACCCATCTGGAGCCTCAGATCATCCCAAGTCATTGTTCATCAGATCTAGTACTCAAATAAGTTAATAAAAATATCCACATGGGGCAAATAATCATTGGAGGAAATCCAACTAATCACAATATCTGTCAACATAGACAAGTCAACACGCCAGGCAAAATCAACCAATGGAAAATACATCCATAACAATAGAATTCTCAAGCAAATTCTGGCCTTACTTTACACTAATACACATGATAACAACAATAATCTCTTTGCTAATCATAATCTCCATCATGATTGCAATACTGAACAAACTCTGTGAATATAACGTATTCCATAACAAAACCTTTGAGCTACCAAGAGCTCGAGTCAATACATAGCATTCACCAATCTGATGGCACAAAACAGTAACCTTGCATTTGTAAGTGAACAACCCTCACCTCTTTACAAAACCACATCAACATCTCACCATGCAAGCCATCATCCATATTATAAAGTAGTTAATTAAAAATAATCATAACAATGAACTAAGATATTAAGACTAACAATAACGTTGGGGCAAATGCAAACATGTCCAAAAACAAGGACCAACGCACCACCAAGACACTAGAAAAGACCTGGGACACTCTCAATCATCTATTATTCATATCATCGTGCTTATACAAGTTAAATCTTAAATCTATAGCACAAATCACATTATCCATTCTGGCAATGATAATCTCAACTTCACTTATAATTGCAGCCATCATATTCATAGCCTCGGCAAACCACAAAGTCACACTAACAACTGCAATCATACAAGATGCAACAAGCCAGATCAAGAACACAACCCCAACATACCTCACCCAGAATCCCCAGCTTGGAATCAGCTTCTCCAATCTGTCTGAAACTACATCACAAACCACCACCATACTAGCTTCAACAACACCAAGTGTCAAGTCAACCCTGCAATCCACAACAGTCAAGACCAAAAACACAACAACAACCAAAATACAACCCAGCAAGCCCACCACAAAACAACGCCAAAACAAACCACCAAACAAACCCAATAATGATTTTCACTTTGAAGTGTTCAACTTTGTACCTTGCAGCATATGCAGCAACAATCCAACCTGCTGGGCTATCTGTAAAAGAATACCAAACAAAAAACCTGGAAAGAAAACCACCACCAAGCCCACAAAAAAACCAACCATCAAGACAACCAAAAAAGATCTCAAACCTCAAACCACAAAACCAAAGGAAGTACCTACCACCAAGCCCACAGAAAAGCCAACCATCAACACCACCAAAACAAACATCAGAACTACACTGCTCACCAACAATACCACAGGAAATCCAGAACACACAAGTCAAAAGGGAACCCTCCACTCAACCTCCTCCGATGGCAATCCAAGCCCTTCACAAGTCTATACAACATCCGAGTACCTATCACAACCTCCATCTCCATCCAACACAACAAACCAGTAGTCATTAAAAAGCGTATTATTGCAAAAAGCCATGACCAAATCAACCAGAATCAAAATCAACTCTGGGGCAAATAACAATGGAGTTGCCAATCCTCAAAACAAATGCAATTACCGCAATCCTTGCTGCAGTCACACTCTGTTTTGCTTCCAGTCAAAACATCACTGAAGAATTTTATCAAACAACATGCAGTGCAGTCAGCAAAGGCTATCTTAGTGCTCTAAGAACTGGTTGGTATACTAGTGTTATAACTATAGAATTAAGTAATATCAAGGAAAATAAGTGTAATGGAACAGACGCTAAGGTAAAATTGATAAAACAAGAATTAGATAAATATAAAAGTGCTGTAACAGAATTGCAGTTGCTCATGCAAAGCACACCGGCAACCAACAATCGAGCCAGAAGAGAACTACCAAGGTTTATGAATTATACACTCAACAATACCAAAAATACCAATGTAACATTAAGCAAGAAAAGGAAAAGAAGATTTCTTGGCTTTTTGTTAGGTGTTGGATCTGCAATCGCCAGTGGCATTGCTGTATCTAAGGTCCTGCACCTAGAAGGGGAAGTGAACAAAATCAAAAGTGCTCTACTATCCACAAACAAGGCTGTAGTCAGCTTATCAAATGGAGTTAGTGTCTTAACCAGCAAAGTGTTAGACCTCAAAAACTATATAGATAAACAGTTGTTACCTATTGTGAACAAGCAAAGCTGTAGCATATCAAACATTGAAACTGTGATAGAGTTCCAACAAAAGAACAACAGACTACTAGAGATTACCAGGGAATTTAGTGTTAATGCAGGTGTAACTACACCTGTAAGCACTTATATGTTAACAAATAGTGAATTATTATCATTAATCAATGATATGCCTATAACAAATGATCAGAAAAAGTTAATGTCCAACAATGTTCAAATAGTTAGACAGCAAAGTTACTCTATCATGTCCATAATAAAGGAGGAAGTCTTAGCATATGTAGTACAATTACCACTATATGGTGTAATAGATACACCTTGTTGGAAACTGCACACATCCCCTCTATGTACAACCAACACAAAGGAAGGGTCCAACATCTGTTTAACAAGAACCGACAGAGGATGGTACTGTGACAATGCAGGATCAGTATCTTTCTTCCCACTAGCTGAAACATGTAAAGTTCAATCGAATCGAGTATTTTGTGACACAATGAACAGTTTAACATTACCAAGTGAAGTAAATCTCTGCAACATTGACATATTCAACCCCAAATATGATTGCAAAATTATGACTTCAAAAACAGATGTAAGCAGCTCCGTTATCACATCTCTAGGAGCCATTGTGTCATGCTATGGCAAAACTAAATGTACAGCATCCAATAAAAATCGTGGAATCATAAAGACATTTTCTAACGGGTGCGATTATGTATCAAATAAGGGGGTTGACACTGTGTCTGTAGGTAATACATTATATTATGTAAATAAGCAAGAAGGCAAAAGTCTCTATGTAAAAGGTGAACCAATAATAAATTTCTATGACCCATTAGTGTTCCCCTCTGATGAATTTGATGCATCAATATCTCAAGTCAATGAGAAGATTAACCAGAGCCTAGCATTTATTCGTAAATCCGATGAATTATTACATAATGTAAATGCTGGTAAATCCACCATAAATATCATGATAACTACTATAATTATAGTGATTATAGTAATATTGTTATCATTAATTGCCGTTGGACTGCTCCTATACTGCAAGGCCAGAAGCACACCAGTCACACTAAGCAAGGATCAACTGAGTGGTATAAATAATATTGCATTTAGTAACTAAATAAAAATAGCACCTAATCATGTTCTTACAATGGTTTCATATCTGCTCATAGACAACCCATCTATCATTGGATTTTCTTAAAATCTGAACTTCATCGAAACTCTCATCTATAAACCATCTCACTTACATTATTTAAGTAGATTCCTAGTTTATAGTTATATAAAACAATTGAATACCAGATTAACTTACTATTTGTAAAAAATGAGAACTGGGGCAAATATGTCACGAAGGAATCCTTGCAAATTTGAAATTCGAGGTCATTGCTTGAATGGTAAGAGGTGTCATTTTAGTCATAATTATTTTGAATGGCCACCCCATGCACTGCTTGTAAGACAAAACTTTATGTTAAACAGAATACTTAAGTCTATGGATAAAAGCATCGATACTTTATCAGAAATAAGTGGAGCTGCAGAGTTGGACAGAACAGAAGAGTATGCCCTCGGTGTAGTTGGAGTGCTAGAGAGTTATATAGGATCTATAAATAATATAACTAAACAATCAGCATGTGTTGCCATGAGCAAACTCCTCACTGAACTCAACAGTGATGACATCAAAAAACTGAGGGACAATGAAGAGCCAAATTCACCCAAGATAAGAGTGTACAATACTGTCATATCATATATTGAAAGCAACAGGAAAAACAATAAACAAACTATCCATCTGTTAAAAAGATTGCCAGCAGACGTATTGAAGAAAACCATAAAAACCACATTGGATATCCACAAGAGCATAACCATCAATAACCCAAAAGAATCAACTGTTAGTGATATAAACGACCATGCCAAAAATAATGATACTACCTGACAAATATCCTTGTAGTATAAATTCCATACTAATAACAAGTAGTTGTAGAGTTACTATGTATAATCAAAAGAACACACTATATTTCAATCAAAACAACCAAAATAACCATATATACTCACCGAATCAACCATTCAATGAAATCCATTGGACCTCTCAAGACTTGATTGATGCAATTCAAAATTTTCTACAACATCTAGGTATTACTGATGATATATACACAATATATATATTAGTGTCATAACACTCAATCCTAATGCTTACCACATCATCAAACTATTAACTCAAACAATTCAAGCCATGGGACAAAATGGATCCCATTATTAATGGAAATTCTGCTAATGTGTATCTAACCGATAGTTATTTAAAAGGTGTTATTTCTTTCTCAGAATGTAATGCTTTAGGAAGTTACATATTCAATGGTCCTTATCTCAAAAATGATTATACCAACTTAATTAGTAGACAAAATCCATTAATAGAACACATAAATCTAAAGAAACTAAATATAACACAGTCCTTAATGTCTAAGTATCATAAAGGTGAAATAAAAATAGAAGAACCTACTTATTTTCAGTCATTACTTATGACATACAAGAGTATGACCTCGTTAGAACAGATTACTACCACTAATTTACTTAAAAAGATAATAAGAAGAGCTATAGAAATTAGTGATGTCAAAGTCTATGCTATATTGAATAAACTGGGGCTTAAAGAAAAAGACAAGATTAAATCCAACAATGGACAAGATGAAGACAACTCAGTTATTACAACCATAATCAAAGATGATATACTTTTAGCTGTTAAGGATAATCAATCTCATCTTAAAGCAGTCAAAAATCACTCTACAAAACAAAAAGATACAATCAAAACAACACTCTTGAAGAAATTAATGTGTTCAATGCAACATCCTCCATCATGGTTAATACATTGGTTTAATTTATACACAAAATTAAACAACATATTAACACAGTATCGATCAAGTGAGGTAAAAAACCATGGTTTTATATTGATAGACAATCATACTCTCAATGGATTCCAATTTATTTTGAATCAATATGGTTGTATAGTTTATCATAAGGAACTCAAAAGAATTACTGTGACAACCTATAATCAATTCTTGACATGGAAAAATATTAGCCTTAGTAGATTAAATGTTTGTTTAATTACATGGATTAGTAACTGTTTGAACACATTAAATAAAAGCTTAGGCTTAAGATGCGGATTCAATAATGTTATCTTGACACAACTATTCCTCTATGGAGATTGTATACTAAAACTATTCCACAATGAGGGGTTCTACATAATAAAAGAGGTAGAGGGATTTATTATGTCTCTAATTTTAAATATAACAGAAGAAGATCAATTCAGAAAACGGTTTTATAATAGTATGCTCAACAACATCACAGATGCTGCTAATAAAGCTCAGAAAAATCTGCTATCAAGAGTATGTCATACATTATTAGATAAGACAGTATCCGATAATATAATAAATGGCAGATGGATAATTCTATTAAGTAAGTTCCTTAAATTAATTAAGCTTGCAGGTGACAATAACCTTAACAATCTGAGTGAATTATATTTTTTGTTCAGAATATTTGGACACCCAATGGTAGATGAAAGACAAGCCATGGATGCTGTTAAAGTTAATTGCAACGAGACCAAATTTTACTTGTTAAGCAGTTTGAGTATGTTAAGAGGTGCCTTTATATATAGAATTATAAAAGGATTTGTAAATAATTACAACAGATGGCCTACTTTAAGGAATGCTATTGTTTTACCCTTAAGATGGTTAACTTACTATAAACTAAACACTTATCCTTCCTTGTTGGAACTTACAGAAAGAGATTTGATTGTTTTATCAGGACTACGTTTCTATCGTGAGTTTCGGTTGCCTAAAAAAGTGGATCTTGAAATGATCATAAATGATAAGGCTATATCACCTCCTAAAAATTTGATATGGACTAGTTTCCCTAGAAATTATATGCCGTCACACATACAAAATTATATAGAACATGAAAAATTAAAATTTTCCGAGAGTGATAAATCAAGAAGAGTATTAGAGTACTATTTAAGAGATAACAAATTCAATGAATGTGATTTATACAACTGTGTAGTTAATCAAAGTTATCTTAACAACCCTAATCATGTGGTATCTTTGACAGGCAAAGAAAGAGAACTCAGTGTAGGTAGAATGTTTGCAATGCAACCAGGAATGTTCAGACAAGTTCAAATATTAGCAGAGAAAATGATAGCTGAAAACATTTTACAATTCTTTCCTGAAAGTCTTACAAGATATGGTGATCTAGAACTACAGAAAATATTAGAATTGAAAGCAGGAATAAGTAACAAATCAAATCGTTACAATGATAATTACAACAATTACATTAGTAAGTGCTCTATCATCACAGATCTCAGCAAATTCAATCAAGCATTTCGATATGAAACATCATGTATTTGTAGTGATGTACTGGATGAACTGCATGGTGTACAATCTCTATTTTTCTGGTTACATTTAGCTATTCCTCATGTCACAATAATATGCACATATAGGCATGCACCCCCCTATATAAGAGATCATATTGTAGATCTTAACAATGTAGATGAACAAAGTGGATTATATAGATATCATATGGGTGGTATCGAAGGGTGGTGTCAAAAACTATGGACCATAGAAGCTATATCACTATTGGATCTAATATCTCTCAAAGGGAAATTCTCAATTACTGCTTTAATTAATGGTGACAATCAATCAATAGATATAAGTAAACCAGTCAGACTCATGGAAGGTCAAACTCATGCTCAAGCAGATTATTTGCTAGCATTAAATAGTCTTAAATTACTGTATAAAGAGTATGCAGGCATAGGCCACAAATTAAAAGGAACTGAGACTTATATATCAAGAGATATGCAATTTATGAGTAAAACAATTCAACATAACGGTGTATATTACCCAGCTAGTATAAAGAAAGTCCTAAGAGTGGGACCGTGGATAAACACTATACTTGATGATTTCAAAGTGAGTCTAGAATCTATAGGTAGTTTGACACAAGAATTAGAATATAGAGGAGAAAGTCTATTATGCAGTTTAATATTTAGAAATGTATGGTTATATAATCAAATTGCTTTACAACTAAAAAATCATGCATTATGTAACAATAAATTATATTTGGACATATTAAAGGTTCTGAAACACTTAAAAACCTTTTTTAATCTTGATAATATTGATACAGCATTAACATTGTATATGAATTTGCCCATGTTATTTGGTGGTGGTGATCCCAACTTGTTATATCGAAGTTTCTATAGAAGAACTCCTGATTTCCTCACAGAGGCTATAGTTCACTCTGTGTTCATACTTAGTTATTATACAAACCATGATTTAAAAGATAAACTTCAAGATCTGTCAGATGATAGATTGAATAAGTTCTTAACATGCATAATCACATTTGACAAAAACCCTAATGCTGAATTCGTAACATTGATGAGAGATCCTCAAGCTTTAGGGTCTGAGAGACAAGCTAAAATTACTAGCGAAATCAATAGACTGGCAGTTACTGAGGTTTTGAGCACAGCTCCAAACAAAATATTCTCCAAAAGTGCACAACACTATACCACTACAGAGATAGATCTAAATGATATTATGCAAAATATAGAACCTACATATCCTCACGGGCTAAGAGTTGTTTATGAAAGTTTACCCTTTTATAAAGCAGAGAAAATAGTAAATCTTATATCCGGTACAAAATCTATAACTAACATACTGGAAAAGACTTCTGCCATAGACTTAACAGATATTGATAGAGCCACTGAGATGATGAGGAAAAACATAACTTTGCTTATAAGGATATTTCCATTAGATTGTAACAGAGATAAAAGGGAAATATTGAGTATGGAAAACCTAAGTATTACTGAATTAAGCAAATATGTTAGGGAAAGATCTTGGTCTTTATCCAATATAGTTGGTGTTACATCACCTAGTATCATGTATACAATGGACATCAAATATACAACAAGCACTATAGCTAGTGGCATAATCATAGAGAAATATAATGTTAACAGTTTAACACGTGGTGAGAGAGGACCCACTAAACCATGGGTTGGTTCATCTACACAAGAGAAAAAAACAATGCCAGTTTATAATAGACAAGTTTTAACCAAAAAACAGAGAGATCAAATAGATCTATTAGCAAAATTGGATTGGGTGTATGCATCTATAGATAACAAGGATGAATTCATGGAAGAACTTAGCATAGGAATTCTTGGGTTAACATATGAGAAAGCCAAAAAATTATTTCCACAATATTTAAGTGTTAACTATTTGCATCGCCTTACAGTCAGTAGTAGACCATGTGAATTCCCTGCATCAATACCAGCTTATAGAACTACAAATTATCACTTTGATACTAGCCCTATTAATCGCATATTAACAGAAAAGTATGGTGATGAAGATATTGATATAGTATTCCAAAACTGTATAAGCTTTGGCCTTAGCTTAATGTCAGTAGTAGAACAATTTACTAATGTATGTCCTAACAGAATTATTCTTATACCTAAGCTTAATGAGATACATTTAATGAAACCTCCCATATTCACAGGTGATGTTGATATTCACAAGTTAAAACAAGTGATACAAAAACAGCATATGTTTTTACCAGACAAAATAAGTTTGACTCAATATGTGGAATTATTCTTAAGTAATAAAACACTCAAATCTGGATCTCATGTTAATTCTAATTTAATATTGGCGCATAAGATATCTGACTATTTTCATAATACTTACATTTTAAGTACTAATTTAGCTGGACATTGGATTCTGATTATACAACTTATGAAAGATTCTAAGGGTATTTTTGAAAAAGATTGGGGAGAGGGATATATAACTGATCATATGTTCATTAATTTGAAAGTTTTCTTCAATGCTTATAAGACCTATCTCTTGTGTTTTCATAAAGGTTACGGCAGAGCAAAGCTGGAGTGTGATATGAATACTTCAGATCTCCTATGTGTATTGGAATTAATAGACAGTAGTTATTGGAAGTCTATGTCTAAGGTATTTTTAGAACAAAAAGTTATCAAATACATTCTTAGCCAGGATGCAAGTTTACATAGAGTAAAAGGATGTCATAGCTTCAAACTATGGTTTCTTAAACGTCTTAATGTAGCAGAATTCACAGTTTGCCCTTGGGTTGTTAACATAGATTATCATCCAACACATATGAAAGCAATATTAACTTATATAGATCTTGTTAGAATGGGATTGATAAATATAGATAAAATATACATTAAAAATAAACACAAATTCAATGATGAATTTTATACTTCTAATCTCTTTTACATTAATTATAACTTCTCAGATAATACTCATCTATTAACTAAACATATAAGGATTGCTAATTCTGAATTAGAAAATAATTACAACAAATTATATCATCCTACACCAGAAACCCTAGAAAATATACTAACCAATCCGGTTAAATGTAATGACAAAAAGACACTGAATGACTATTGTATAGGTAAAAATGTTGACTCAATAATGTTACCATTGTTATCTAATAAGAAGCTTATTAAATCGCCTACAATGATTAGAACCAATTACAGCAAACAAGATTTGTATAATTTATTTCCTACGGTTGTGATTGATAAAATTATAGATCATTCAGGTAATACAGCCAAATCTAACCAACTTTACACTACTACTTCTCATCAAATACCTTTAGTGCACAATAGCACATCACTTTATTGCATGCTTCCTTGGCATCATATTAATAGATTCAATTTTGTATTTAGTTCTACAGGTTGTAAAATTAGTATAGAGTATATTTTAAAAGACCTTATAATTAAAGATCCTAATTGTATAGCATTCATAGGTGAAGGAGCAGGGAATTTATTATTGCGTACAGTAGTGGAACTTCATCCCGATATAAGATATATTTACAGAAGTCTGAAGGATTGCAATGATCATAGTTTACCTATTGAGTTTTTAAGGCTGTACAATGGACATATCAACATTGATTATGGTGAAAATTTGACCATTCCTGCTACAGATGCAACCAACAACATTCATTGGTCTTATTTACATATAAAGTTTGCTGAACCTATCAGTCTTTTTGTCTGTGATGCTGAATTGCCTGTAACAGTCAACTGGAGTAAAATTATAATAGAGTGGAGCAAGCATGTAAGAAAATGCAAGTACTGTTCCTCAGTTAATAAATGTACGTTAATAGTAAAATATCATGCTCAAGATGATATCGATTTCAAATTAGACAATATAACTATATTAAAAACTTATGTATGCTTAGGCAGTAAGTTAAAGGGGTCTGAAGTTTACTTAGTCCTTACAATAGGTCCTGCAAATGTGTTCCCAGTATTTAATGTAGTACAAAATGCTAAATTGATACTATCAAGAACCAAAAATTTCATCATGCCTAAGAAGGCTGATAAAGAGTCTATTGATGCAAATATTAAAAGTTTGATACCCTTTCTTTGTTACCCTATAACAAAAAAAGGAATTAATACTGCATTGTCAAAACTAAAGAGTGTTGTTAGTGGAGATATACTATCATATTCTATAGCAGGACGTAATGAAGTTTTCAGCAATAAACTTATAAATCATAAGCATATGAACATCTTAAAATGGTTCAATCATGTTTTAAATTTCAGATCAACAGAACTAAACTATAATCATTTATATATGGTAGAATCTACATATCCTTATCTAAGTGAATTGTTAAACAGCTTGACAACTAATGAACTTAAAAAACTGATTAAAATCACAGGTAGTTTGTTATACAACTTTCATAATGAATAATGAATAAAAATCTTATATTAAAAATTCCCATAGCTACACACTAACACTGTATTCAATTATAGTTATTTAAAATTAAAAATTATATAATTTTTTAATAACTTTTAGTGAACTAATCCTAAAATTATCATTTTGATCTAGGAGGAATAAATTTAAATCCAAATCTAATTGGTTTATATGTATATTAACTAAACTACGAGATATTAGTTTTTGACACTTTTTTTCTCGT'\n",
    "ref_seq_list = list(ref_seq)\n",
    "dict_jianji = {}\n",
    "for k,v in dict_info_feikong.items():\n",
    "    gene_pos = int(k.split('_')[1])\n",
    "    ref = k.split('_')[0]\n",
    "    alt = k.split('_')[2]\n",
    "    product = v['product']\n",
    "#     frequency = v['frequency']\n",
    "    gene_pos_f2 = gene_pos - 2   #前面2个\n",
    "    gene_pos_f1 = gene_pos - 1   #前面1个\n",
    "    gene_pos_b1 = gene_pos + 1\n",
    "    gene_pos_b2 = gene_pos + 2\n",
    "    start = v['start']\n",
    "    end = v['end']\n",
    "    type4 = v['type']\n",
    "    chazhi = int(gene_pos) - int(start)\n",
    "    quzheng =  chazhi//3   #取整，以获取是这个蛋白质的第几个氨基酸。\n",
    "    aa_pos = quzheng + 1\n",
    "\n",
    "# #分类讨论\n",
    "    if type4 == 'del' or type4 == 'ins':\n",
    "        del_count = len(ref) - len(alt)\n",
    "        if del_count%3 == 0:\n",
    "            #分类讨论\n",
    "            #1\n",
    "            if chazhi%3 == 2 and len(alt)%3 == 2:\n",
    "                #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "                list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                list_jianji_pos_alt.append(gene_pos_f1)\n",
    "                \n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "                list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "           #2\n",
    "            if chazhi%3 == 2 and len(alt)%3 == 1:\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "              \n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                list_jianji_pos_alt.append(gene_pos_f1)\n",
    "                \n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "           \n",
    "            #3\n",
    "            if chazhi%3 == 2 and len(alt)%3 == 0:\n",
    "            #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "                list_jianji_pos_ref.append(gene_pos_b1)\n",
    "\n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "                list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "            \n",
    "            #4\n",
    "            if chazhi%3 == 1 and len(alt)%3 == 2:\n",
    "                #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                list_jianji_pos_ref.append(gene_pos_f2)\n",
    "               \n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "              \n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                list_jianji_pos_alt.append(gene_pos_f2)\n",
    "               \n",
    "                \n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                \n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "      \n",
    "            #5\n",
    "            if chazhi%3 == 1 and len(alt)%3 == 1:\n",
    "                #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "                list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "                list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "        #6\n",
    "            if chazhi%3 == 1 and len(alt)%3 == 0:\n",
    "                #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                \n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "                list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                list_jianji_pos_alt.append(gene_pos_f2)\n",
    "              \n",
    "                \n",
    "                list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                \n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "                list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "                \n",
    "            #7\n",
    "            if chazhi%3 == 0 and len(alt)%3 == 2:\n",
    "                #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "       \n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "                list_jianji_pos_ref.append(gene_pos_b1)\n",
    "               \n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian]) \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "                list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                \n",
    "            #8\n",
    "            if chazhi%3 == 0 and len(alt)%3 == 1:\n",
    "                #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "                list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "                list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "        #9\n",
    "            if chazhi%3 == 0 and len(alt)%3 == 0:\n",
    "                #变异前的位置和碱基\n",
    "                list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                list_jianji_ref = []\n",
    "                for index in range(0,len(ref)):\n",
    "                    list_jianji_pos_ref.append((gene_pos + index))\n",
    "         \n",
    "                for p in list_jianji_pos_ref:\n",
    "                    p_qiepian = p-1\n",
    "                    list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "               \n",
    "                #变异后的位置\n",
    "                list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                list_jianji_alt = []\n",
    "                for index in range(0,len(alt)):\n",
    "                    list_jianji_pos_alt.append((gene_pos + index))\n",
    "                    list_jianji_alt.append(alt[index])\n",
    "      \n",
    " \n",
    "    \n",
    "  #------------------------原来的，底下可以不变  \n",
    "    if type4 == 'snp':\n",
    "        if chazhi%3 == 2:\n",
    "            list_jianji_pos = []#是位置不是切片，切片要-1\n",
    "            list_jianji_pos.append(gene_pos_f2)\n",
    "            list_jianji_pos.append(gene_pos_f1)\n",
    "            list_jianji_pos.append(gene_pos)\n",
    "\n",
    "\n",
    "\n",
    "        if chazhi%3 == 1:\n",
    "            list_jianji_pos = []\n",
    "            list_jianji_pos.append(gene_pos_f1)\n",
    "            list_jianji_pos.append(gene_pos)\n",
    "            list_jianji_pos.append(gene_pos_b1)\n",
    "\n",
    "\n",
    "\n",
    "        if chazhi%3 == 0:\n",
    "            list_jianji_pos = []\n",
    "            list_jianji_pos.append(gene_pos)\n",
    "            list_jianji_pos.append(gene_pos_b1)\n",
    "            list_jianji_pos.append(gene_pos_b2)\n",
    "\n",
    "\n",
    "        list_jianji_qiepian = []#这个是切片\n",
    "        for i in list_jianji_pos:\n",
    "            i_1 = i-1\n",
    "            list_jianji_qiepian.append(i_1)\n",
    "    #     print(list_jianji_qiepian)\n",
    "\n",
    "        list_jianji_ref = []#获取参考碱基\n",
    "        for i in list_jianji_qiepian:\n",
    "            list_jianji_ref.append(ref_seq_list[i])\n",
    "    #     print(list_jianji_ref)\n",
    "\n",
    "        #获取alt碱基\n",
    "        list_jianji_alt= []#获取参考碱基\n",
    "        for i in list_jianji_qiepian:\n",
    "            if i == int(gene_pos - 1):\n",
    "                list_jianji_alt.append(alt)\n",
    "            else:\n",
    "                list_jianji_alt.append(ref_seq_list[i])\n",
    "    #     print(list_jianji_alt)\n",
    "    #翻译蛋白质\n",
    "    str_ref = \"\".join(list_jianji_ref)\n",
    "    str_alt = \"\".join(list_jianji_alt)\n",
    "    aa_ref = str(Seq(str_ref).translate())\n",
    "    aa_alt = str(Seq(str_alt).translate())    \n",
    "    #非同义：\n",
    "    if aa_alt != aa_ref:\n",
    "        info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "        dict_aa_mut[k] = {}\n",
    "        dict_aa_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "        dict_aa_mut[k]['aa'] = info\n",
    "        dict_aa_mut[k]['num'] = dict_mut_num[k]\n",
    "    #同义\n",
    "    if aa_alt == aa_ref:\n",
    "        info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "        dict_aa_synonymous_mut[k] = {}\n",
    "        dict_aa_synonymous_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "        dict_aa_synonymous_mut[k]['aa'] = info\n",
    "        dict_aa_synonymous_mut[k]['num'] = dict_mut_num[k]\n",
    "#             dict_aa_mut[k]['frequency'] = dict_info_feikong[k]['frequency']\n",
    "    \n",
    "# \n",
    "\n",
    "#非同义----------------------------------------------------------------------\n",
    "#统计各个突变的频率。\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_nonsynonymous_mut_frequency.csv','w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "    for k,v in dict_aa_mut.items():\n",
    "        writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "    \n",
    "    \n",
    "#统计非同义突变在哪个蛋白上nonsynonymous mutation\n",
    "list_product = []\n",
    "for k,v in dict_aa_mut.items():\n",
    "        if v['product'] not in list_product:\n",
    "            list_product.append(v['product'])\n",
    "            dict_product = dict.fromkeys(list_product)#每个蛋白上非同义突变的个数\n",
    "for k in dict_product.keys():\n",
    "    dict_product[k] = 0\n",
    "for m,p in dict_aa_mut.items():\n",
    "    for k,v in dict_product.items():\n",
    "        if p['product'] == k:\n",
    "            dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "# print(dict_product)\n",
    "dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "# print(dict_product)\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_nonsynonymous_product_number.csv','w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow(['product','number','group'])\n",
    "    for k,v in dict_product.items():\n",
    "        writer.writerow([k,v,'nonsynonymous'])\n",
    "        \n",
    "\n",
    "#同义----------------------------------------------------------------------\n",
    "#统计各个突变的频率。\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_synonymous_mut_frequency.csv','w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "    \n",
    "    for k,v in dict_aa_synonymous_mut.items():\n",
    "        writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "    \n",
    "    \n",
    "#统计同义突变在哪个蛋白上\n",
    "list_product = []\n",
    "for k,v in dict_aa_synonymous_mut.items():\n",
    "        if v['product'] not in list_product:\n",
    "            list_product.append(v['product'])\n",
    "            dict_product = dict.fromkeys(list_product)\n",
    "for k in dict_product.keys():\n",
    "    dict_product[k] = 0\n",
    "for m,p in dict_aa_synonymous_mut.items():\n",
    "    for k,v in dict_product.items():\n",
    "        if p['product'] == k:\n",
    "            dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "# print(dict_product)\n",
    "dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "# print(dict_product)\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_synonymous_product_number.csv','w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow(['product','number','group'])\n",
    "    for k,v in dict_product.items():\n",
    "        writer.writerow([k,v,'synonymous'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "093d93fe-0b87-4673-b01d-8795b1206b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc22ef27-254f-4285-9aa6-99d5188091cc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe170ef-a9d8-4e5a-8627-20998cd5ef73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6cc19ad-e942-4dfb-85ae-cdd188555c15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc683bdb-ac4a-49b7-a8c1-7e56aa4c1062",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e2a3608-4127-4abe-ab47-dc5913ca4284",
   "metadata": {},
   "outputs": [],
   "source": [
    "#看一下都有哪些prj的vcF是没有找出来突变的\n",
    "import os\n",
    "def is_vcf_empty(vcf_file):\n",
    "    with open(vcf_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if not line.startswith('#'):\n",
    "                return False\n",
    "    return True\n",
    "\n",
    "empty_list = []\n",
    "prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair')\n",
    "for i in prj_list:\n",
    "    panduan_list = []\n",
    "    vcf_list = os.listdir(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair/{i}')\n",
    "    for n in vcf_list:\n",
    "        vcf_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair/{i}/{n}'\n",
    "# 调用函数来检查 VCF 文件是否为空\n",
    "        vcf_empty = is_vcf_empty(vcf_path)\n",
    "        if vcf_empty:\n",
    "            panduan_list.append('empty')\n",
    "        else:\n",
    "            panduan_list.append('not_empty')\n",
    "    if 'not_empty' in panduan_list:\n",
    "        pass\n",
    "    else:\n",
    "        empty_list.append(i)\n",
    "print(empty_list)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b35ba014-5676-4bd3-a066-c3db86f9233c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#先找出来有哪些prj是有符合要求的突变，可以用的\n",
    "import os\n",
    "import csv\n",
    "\n",
    "\n",
    "all_prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/fastq_result')\n",
    "csv_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/09snp_isnv_all_mut')\n",
    "not_empty_prj_list = []\n",
    "for i in csv_list:\n",
    "    csv_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/09snp_isnv_all_mut/{i}'\n",
    "    total = len(open(csv_path).readlines())\n",
    "    if total != 1:\n",
    "        not_empty_prj_list.append(i.split('.')[0])\n",
    "print(len(not_empty_prj_list))\n",
    "print(not_empty_prj_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55a28f79-0b55-486e-83f1-54f7a7fc39f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#先找出来有哪些prj是有符合要求的突变，可以用的\n",
    "import os\n",
    "import csv\n",
    "\n",
    "\n",
    "all_prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/fastq_result')\n",
    "csv_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/09snp_isnv_all_mut')\n",
    "not_empty_prj_list = []\n",
    "for i in csv_list:\n",
    "    csv_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/09snp_isnv_all_mut/{i}'\n",
    "    total = len(open(csv_path).readlines())\n",
    "    if total != 1:\n",
    "        not_empty_prj_list.append(i.split('.')[0])\n",
    "print(len(not_empty_prj_list))\n",
    "print(not_empty_prj_list)\n",
    "\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/11consensus/00script/01write_run_spades.sh','w')as f:\n",
    "    for i in all_prj_list:\n",
    "        if i in not_empty_prj_list:\n",
    "            prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/fastq_result/{i}'\n",
    "            fq_name_list = []\n",
    "            all_fq_list = os.listdir(prj_path)\n",
    "            for n in all_fq_list:\n",
    "                n1 = n.split('_')[0]\n",
    "                if n1 not in fq_name_list:\n",
    "                    fq_name_list.append(n1)\n",
    "            #新建目录建立prj和其中每个样本output的目录。\n",
    "            for s in fq_name_list:           \n",
    "\n",
    "                dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/11consensus/01spades_consensus_output/{i}/{s}'\n",
    "                if not os.path.exists(dirs):\n",
    "                    os.makedirs(dirs)\n",
    "            for s in fq_name_list: \n",
    "                f.write(f'docker run -id -v /mnt/alamo01/users/yangjingqi/RSV_sra/group/fastq_result/{i}:/input -v /mnt/alamo01/users/yangjingqi/RSV_sra/group/11consensus/01spades_consensus_output/{i}/{s}:/output staphb/spades spades.py -1 /input/{s}_1.fastq.gz -2 /input/{s}_2.fastq.gz -o /output\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6c8de48-b355-4f96-bf91-2c44ebbc0628",
   "metadata": {},
   "outputs": [],
   "source": [
    "#一致性序列"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83742d2c-be9b-4efb-a182-e6b6f4106c27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#构建一致性序列先用bcftools对vcf文件排序压缩。\n",
    "import os\n",
    "import csv\n",
    "\n",
    "\n",
    "all_prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair')\n",
    "csv_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/09snp_isnv_all_mut')\n",
    "not_empty_prj_list = []\n",
    "for i in csv_list:\n",
    "    csv_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/09snp_isnv_all_mut/{i}'\n",
    "    total = len(open(csv_path).readlines())\n",
    "    if total != 1:\n",
    "        not_empty_prj_list.append(i.split('.')[0])\n",
    "print(len(not_empty_prj_list))\n",
    "print(not_empty_prj_list)\n",
    "\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/00script/01write_run_bcftools_sort_gz.sh','w')as f:\n",
    "    for i in all_prj_list:\n",
    "        if i in not_empty_prj_list:\n",
    "            dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}'\n",
    "            if not os.path.exists(dirs):\n",
    "                os.makedirs(dirs)\n",
    "            prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair/{i}'\n",
    "            vcf_name_list = []\n",
    "            all_vcf_list = os.listdir(prj_path)\n",
    "            for n in all_vcf_list:\n",
    "                n1 = n.split('.')[0]\n",
    "                f.write(f'bcftools sort /mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair/{i}/{n1}.vcf -Oz -o /mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}/{n1}_sorted.vcf.gz\\n')\n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05e689e5-e481-4dad-9b2a-d6f3847d07b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tabix进行索引\n",
    "import os\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/00script/02tabix.sh','w')as f:\n",
    "    prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz')\n",
    "    for i in prj_list:\n",
    "        prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}'\n",
    "        gz_list = os.listdir(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}')\n",
    "        for n in gz_list:\n",
    "            n1 = n.split('_')[0]\n",
    "            gz_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}/{n}'\n",
    "            f.write(f'tabix {gz_path}\\n')\n",
    "\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1ca4f1d-aa85-4361-aa0a-a45d93c2d07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#构建一致性序列\n",
    "import os\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/00script/03consensus.sh','w')as f:\n",
    "    prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz')\n",
    "    for i in prj_list:\n",
    "        dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/02consensus/{i}'\n",
    "        if not os.path.exists(dirs):\n",
    "            os.makedirs(dirs)\n",
    "        prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}'\n",
    "        gz_list = os.listdir(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}')\n",
    "        for n in gz_list:\n",
    "            n1 = n.split('_')[0]\n",
    "            gz_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz/{i}/{n1}_sorted.vcf.gz'\n",
    "            f.write(f'bcftools consensus -f /mnt/alamo01/users/yangjingqi/RSV_sra/index/RSV.fasta {gz_path} > /mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/02consensus/{i}/{n1}.fa\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d7f10bd-4328-4cb2-9dd5-f473722ec477",
   "metadata": {},
   "outputs": [],
   "source": [
    "#现在的fa文件都是参考基因组的名字，先改一下。\n",
    "import os\n",
    "with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/00script/03consensus.sh','w')as f:\n",
    "    prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/01write_run_bcftools_sort_gz')\n",
    "    for i in prj_list:\n",
    "        dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/03consensus_right_name/{i}'\n",
    "        if not os.path.exists(dirs):\n",
    "            os.makedirs(dirs)\n",
    "            \n",
    "prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/02consensus')\n",
    "for i in prj_list:\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/02consensus/{i}'\n",
    "    fa_list = os.listdir(f'{prj_path}')\n",
    "    for n in fa_list:\n",
    "        n1 = n.split('.')[0]\n",
    "        fa_path = f'{prj_path}/{n}'  \n",
    "        f = open(fa_path,\"r\")\n",
    "        lines = f.readlines() #读取全部内容 ，并以列表方式返回\n",
    "        cov_seq = {}\n",
    "        seq_id = ''\n",
    "        for line in lines:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                seq_id = line\n",
    "                cov_seq[seq_id] = ''\n",
    "                seq = ''\n",
    "            else:\n",
    "                seq += line\n",
    "                cov_seq[seq_id] = seq\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/03consensus_right_name/{i}/{n1}.fasta','w')as f:        \n",
    "            for k,v in cov_seq.items():\n",
    "                f.write(f'>{n1}')\n",
    "                f.write('\\n')\n",
    "                f.write(v)\n",
    "                f.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "872b6b2d-65f1-4623-9f46-0720974c48a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#把各个prj中的一致性序列拼在一个文件中，记得在第一行加上参考基因组,每个fasta的名字要改一下。\n",
    "import os\n",
    "\n",
    "prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/03consensus_right_name')\n",
    "fa_path_list = []\n",
    "# ref_path = '/mnt/alamo01/users/yangjingqi/RSV_sra/index/RSV.fasta'\n",
    "# fa_path_list.append(ref_path)\n",
    "for i in prj_list:\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/03consensus_right_name/{i}'\n",
    "    fa_list = os.listdir(f'{prj_path}')\n",
    "    for n in fa_list:\n",
    "        fa_path = f'{prj_path}/{n}'\n",
    "        fa_path_list.append(fa_path)\n",
    "# print(fa_path_list)\n",
    "        \n",
    "for i in fa_path_list:\n",
    "    os.system(f'cat {i} >> /mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/04consensus_all_in_one/RSV.fasta')\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "# list_str = ' '.join(fa_path_list)\n",
    "# print(list_str)\n",
    "# command = f'cat {list_str} >/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/04consensus_all_in_one/RSV_consensus.fa'\n",
    "# print(command)\n",
    "\n",
    "# os.system(f'/usr/bin/cat {list_str} >/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/04consensus_all_in_one/RSV_consensus.fa')\n",
    "# os.system(f'cat /mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/03consensus_right_name/PRJNA61367/SRR527716.fasta /mnt/alamo01/users/yangjingqi/RSV_sra/index/RSV.fasta >/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/04consensus_all_in_one/hh.fasta')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "9da70291-36a6-480d-9eee-ecf513a47965",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Cluster 6': 1, 'Cluster 7': 1, 'Cluster 10': 1, 'Cluster 12': 1, 'Cluster 18': 1, 'Cluster 27': 1, 'Cluster 35': 1, 'Cluster 36': 1, 'Cluster 37': 1, 'Cluster 38': 1, 'Cluster 39': 1, 'Cluster 45': 1, 'Cluster 48': 1, 'Cluster 50': 1, 'Cluster 51': 1, 'Cluster 53': 1, 'Cluster 54': 1, 'Cluster 58': 1, 'Cluster 63': 1, 'Cluster 67': 1, 'Cluster 68': 1, 'Cluster 71': 1, 'Cluster 78': 1, 'Cluster 86': 1, 'Cluster 87': 1, 'Cluster 89': 1, 'Cluster 104': 1, 'Cluster 105': 1, 'Cluster 107': 1, 'Cluster 112': 1, 'Cluster 115': 1, 'Cluster 119': 1, 'Cluster 120': 1, 'Cluster 121': 1, 'Cluster 123': 1, 'Cluster 125': 1, 'Cluster 134': 1, 'Cluster 137': 1, 'Cluster 139': 1, 'Cluster 140': 1, 'Cluster 141': 1, 'Cluster 142': 1, 'Cluster 144': 1, 'Cluster 147': 1, 'Cluster 149': 1, 'Cluster 151': 1, 'Cluster 153': 1, 'Cluster 154': 1, 'Cluster 155': 1, 'Cluster 158': 1, 'Cluster 159': 1, 'Cluster 160': 1, 'Cluster 161': 1, 'Cluster 163': 1, 'Cluster 164': 1, 'Cluster 167': 1, 'Cluster 168': 1, 'Cluster 169': 1, 'Cluster 170': 1, 'Cluster 171': 1, 'Cluster 172': 1, 'Cluster 173': 1, 'Cluster 175': 1, 'Cluster 176': 1, 'Cluster 177': 1, 'Cluster 178': 1, 'Cluster 179': 1, 'Cluster 180': 1, 'Cluster 181': 1, 'Cluster 183': 1, 'Cluster 184': 1, 'Cluster 186': 1, 'Cluster 188': 1, 'Cluster 189': 1, 'Cluster 190': 1, 'Cluster 191': 1, 'Cluster 192': 1, 'Cluster 193': 1, 'Cluster 194': 1, 'Cluster 196': 1, 'Cluster 197': 1, 'Cluster 199': 1, 'Cluster 200': 1, 'Cluster 202': 1, 'Cluster 203': 1, 'Cluster 205': 1, 'Cluster 206': 1, 'Cluster 207': 1, 'Cluster 209': 1, 'Cluster 210': 1, 'Cluster 211': 1, 'Cluster 0': 2, 'Cluster 13': 2, 'Cluster 40': 2, 'Cluster 43': 2, 'Cluster 46': 2, 'Cluster 56': 2, 'Cluster 57': 2, 'Cluster 61': 2, 'Cluster 62': 2, 'Cluster 64': 2, 'Cluster 66': 2, 'Cluster 70': 2, 'Cluster 88': 2, 'Cluster 90': 2, 'Cluster 96': 2, 'Cluster 110': 2, 'Cluster 126': 2, 'Cluster 127': 2, 'Cluster 133': 2, 'Cluster 143': 2, 'Cluster 146': 2, 'Cluster 150': 2, 'Cluster 156': 2, 'Cluster 174': 2, 'Cluster 182': 2, 'Cluster 185': 2, 'Cluster 201': 2, 'Cluster 204': 2, 'Cluster 11': 3, 'Cluster 15': 3, 'Cluster 17': 3, 'Cluster 21': 3, 'Cluster 47': 3, 'Cluster 59': 3, 'Cluster 93': 3, 'Cluster 94': 3, 'Cluster 98': 3, 'Cluster 99': 3, 'Cluster 101': 3, 'Cluster 102': 3, 'Cluster 130': 3, 'Cluster 131': 3, 'Cluster 145': 3, 'Cluster 162': 3, 'Cluster 165': 3, 'Cluster 166': 3, 'Cluster 198': 3, 'Cluster 208': 3, 'Cluster 8': 4, 'Cluster 26': 4, 'Cluster 28': 4, 'Cluster 34': 4, 'Cluster 74': 4, 'Cluster 84': 4, 'Cluster 103': 4, 'Cluster 113': 4, 'Cluster 114': 4, 'Cluster 124': 4, 'Cluster 128': 4, 'Cluster 135': 4, 'Cluster 136': 4, 'Cluster 148': 4, 'Cluster 157': 4, 'Cluster 187': 4, 'Cluster 1': 5, 'Cluster 4': 5, 'Cluster 5': 5, 'Cluster 20': 5, 'Cluster 30': 5, 'Cluster 31': 5, 'Cluster 76': 5, 'Cluster 97': 5, 'Cluster 106': 5, 'Cluster 122': 5, 'Cluster 132': 5, 'Cluster 9': 6, 'Cluster 14': 6, 'Cluster 16': 6, 'Cluster 55': 6, 'Cluster 92': 6, 'Cluster 111': 6, 'Cluster 116': 6, 'Cluster 138': 6, 'Cluster 152': 6, 'Cluster 195': 6, 'Cluster 29': 7, 'Cluster 33': 7, 'Cluster 23': 8, 'Cluster 75': 8, 'Cluster 83': 8, 'Cluster 129': 8, 'Cluster 69': 9, 'Cluster 24': 11, 'Cluster 41': 11, 'Cluster 3': 12, 'Cluster 85': 12, 'Cluster 44': 13, 'Cluster 49': 13, 'Cluster 72': 14, 'Cluster 91': 14, 'Cluster 82': 16, 'Cluster 109': 16, 'Cluster 73': 20, 'Cluster 77': 21, 'Cluster 118': 24, 'Cluster 79': 31, 'Cluster 60': 35, 'Cluster 25': 38, 'Cluster 32': 44, 'Cluster 81': 44, 'Cluster 117': 46, 'Cluster 65': 55, 'Cluster 95': 57, 'Cluster 19': 58, 'Cluster 42': 62, 'Cluster 100': 73, 'Cluster 2': 87, 'Cluster 22': 98, 'Cluster 108': 156, 'Cluster 80': 199, 'Cluster 52': 1018}\n",
      "['SRR1050547', 'SRR1180077', 'SRR1180650', 'SRR1050571', 'SRR1180037', 'SRR1180381', 'SRR1180914', 'SRR1180078', 'SRR1180679', 'SRR1180430', 'SRR1181334', 'SRR5408810', 'SRR1050576', 'SRR1181339', 'SRR1050535', 'SRR1180051', 'SRR1181361', 'SRR5408830', 'SRR1180065', 'SRR1050578', 'SRR1180852', 'SRR1180045', 'SRR1181410', 'SRR1050538', 'SRR1180050', 'SRR1050561', 'SRR1050540', 'SRR1180070', 'SRR1050570', 'SRR1180056', 'SRR1050551', 'SRR1180574', 'SRR1050549', 'SRR1180350', 'SRR1180874', 'SRR1050573', 'SRR1181365', 'SRR1181344', 'SRR1180683', 'SRR5408838', 'SRR1180347', 'SRR1180054', 'SRR1181364', 'SRR5408836', 'SRR1181442', 'SRR1180383', 'SRR1180028', 'SRR1180907', 'SRR1180701', 'SRR1181388', 'SRR1050542', 'SRR1180064', 'SRR1180853', 'SRR1180069', 'SRR5408808', 'SRR1180827', 'SRR1180027', 'SRR1180684', 'SRR1180602', 'SRR1050577', 'SRR1181378', 'SRR1180898', 'SRR1180865', 'SRR1180044', 'SRR1180864', 'SRR1180756', 'SRR1180366', 'SRR5408840', 'SRR1180036', 'SRR1181416', 'SRR1180877', 'SRR1180373', 'SRR1181420']\n",
      "['SRR1180562', 'SRR1181437', 'SRR1180839', 'SRR1180517', 'SRR1181400', 'SRR1180841', 'SRR1180344', 'SRR1180365', 'SRR1181380', 'SRR1181417', 'SRR1180819', 'SRR1181443', 'SRR5408817', 'SRR1180916', 'SRR1180710', 'SRR1180728', 'SRR5408814', 'SRR1181373', 'SRR1180664', 'SRR1180586', 'SRR1180859', 'SRR1181379', 'SRR1180390', 'SRR1180546', 'SRR5408831', 'SRR1180389', 'SRR1181444', 'SRR1180721', 'SRR1180603', 'SRR1180906', 'SRR1180779', 'SRR5408811', 'SRR1180870', 'SRR1181374', 'SRR1181425', 'SRR1181362', 'SRR1180873', 'SRR1181429', 'SRR1181343', 'SRR1181408', 'SRR1181432', 'SRR1180601', 'SRR1180904', 'SRR1181337', 'SRR1180068', 'SRR1180678', 'SRR1181424', 'SRR1181431', 'SRR1180867', 'SRR1180408', 'SRR1180412', 'SRR1180850', 'SRR1180792', 'SRR1180409', 'SRR5408821', 'SRR1180749', 'SRR1180593', 'SRR5408837', 'SRR1180818', 'SRR1180035', 'SRR1181436', 'SRR5408843', 'SRR1180438', 'SRR1181346', 'SRR1180382', 'SRR5413220', 'SRR1180042', 'SRR1180370', 'SRR1181423', 'SRR1181415', 'SRR1180577', 'SRR5408834', 'SRR1180380', 'SRR1181394', 'SRR1180085', 'SRR1181419', 'SRR1181434', 'SRR1181439', 'SRR5408823', 'SRR1180522', 'SRR1180627', 'SRR1180515', 'SRR1181333', 'SRR1180076', 'SRR1180038', 'SRR1180423', 'SRR5408842', 'SRR1181381', 'SRR1180901', 'SRR1181366', 'SRR1181430', 'SRR1180742', 'SRR1180363', 'SRR1180582', 'SRR1180354', 'SRR1180762', 'SRR1180545', 'SRR1180785', 'SRR1180812', 'SRR1180789', 'SRR1181341', 'SRR5408828', 'SRR1180845', 'SRR1180912', 'SRR1180533', 'SRR1181409', 'SRR1050575', 'SRR1180777', 'SRR1180894', 'SRR1180913', 'SRR1180570', 'SRR1180393', 'SRR5408844', 'SRR1180846', 'SRR1181338', 'SRR1180029', 'SRR1180778', 'SRR1180374', 'SRR1181335', 'SRR1180847', 'SRR1180573', 'SRR1181336', 'SRR1181433', 'SRR1180791', 'SRR5408825', 'SRR1181389', 'SRR1180360', 'SRR1180741', 'SRR1180882', 'SRR1180757', 'SRR1180790', 'SRR5408847', 'SRR1180426', 'SRR1180386', 'SRR1180411', 'SRR1180377', 'SRR1181363', 'SRR1181414', 'SRR1181440', 'SRR1181399', 'SRR1180576', 'SRR1180084', 'SRR1180915', 'SRR1180682', 'SRR1180645', 'SRR1180708', 'SRR1181438', 'SRR1180893', 'SRR1180401', 'SRR1181441', 'SRR1180514', 'SRR1180890', 'SRR1180022', 'SRR1180891', 'SRR7649921', 'SRR7649928']\n"
     ]
    }
   ],
   "source": [
    "#找一下这些cluster都包含几个序列\n",
    "#chatgpt写的\n",
    "clstr_file = \"/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/12consensus_bcf/04consensus_all_in_one/cdhit/RSV_all_with_ref_986.fasta.clstr\"\n",
    "clusters = []\n",
    "current_cluster = None\n",
    "\n",
    "with open(clstr_file, \"r\") as file:\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "        if line.startswith(\">\"):\n",
    "            if current_cluster is not None:\n",
    "                clusters.append(current_cluster)\n",
    "\n",
    "            current_cluster = []\n",
    "        else:\n",
    "            sequence_id = line.split(\">\")[1].split(\"...\")[0]\n",
    "            status = line.split(\">\")[1].split(\"...\")[1].strip()  #是不是在去冗余中留下，*是留下了\n",
    "            current_cluster.append(sequence_id)\n",
    "            \n",
    "if current_cluster is not None:\n",
    "    clusters.append(current_cluster)    \n",
    "# print(clusters)\n",
    "#--------------------------------------------------------------------------------------------\n",
    "dict_cluster_num = {}\n",
    "for i, cluster in enumerate(clusters):    \n",
    "    dict_cluster_num[f\"Cluster {i}\"] = len(cluster)\n",
    "# print(dict_cluster_num)\n",
    "\n",
    "dict_cluster_num_sorted = dict(sorted(dict_cluster_num.items(),key = lambda x:x[1]))\n",
    "print(dict_cluster_num_sorted)\n",
    "#--------------------------------------------------------------------------------------------\n",
    "\n",
    "dict_cluster_id = {}\n",
    "for i, cluster in enumerate(clusters):  \n",
    "    dict_cluster_id[f\"Cluster {i}\"] = cluster\n",
    "# print(dict_cluster_id)\n",
    "print(dict_cluster_id['Cluster 100'])\n",
    "print(dict_cluster_id['Cluster 108'])\n",
    "#--------------------------------------------------------------------------------------------\n",
    "# 看看带星号的序列都有哪些\n",
    "list_seq_in_tree = []\n",
    "with open(clstr_file, \"r\") as file:\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "        if line.startswith(\">\"):\n",
    "            pass\n",
    "        else:\n",
    "            sequence_id = line.split(\">\")[1].split(\"...\")[0]\n",
    "            status = line.split(\">\")[1].split(\"...\")[1].strip()    #是不是在去冗余中留下，*是留下了\n",
    "            if status == '*':\n",
    "                list_seq_in_tree.append(sequence_id)\n",
    "\n",
    "# print(len(list_seq_in_tree))\n",
    "\n",
    "#-------------\n",
    "#每个cluster都是什么代表的seq\n",
    "dict_cluster_seq_tree = {}\n",
    "for k,v in dict_cluster_id.items():\n",
    "    for i in v:\n",
    "        if i in list_seq_in_tree:\n",
    "            dict_cluster_seq_tree[k] = i\n",
    "# print(dict_cluster_seq_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "095bc0b6-f58e-459c-b794-029aacabb8e4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
