{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c921c4-4457-4409-9ced-9309fbfe322f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#05 统计bayes的每个样本reads数，写成A256G的形式，过滤，只过滤测序深度,不过滤频率。\n",
    "#统计每个样本的突变数量\n",
    "#dp>100并且在至少2个样本中出现过的才算。\n",
    "#变异类型是snp或者del或ins，这里也筛选了。\n",
    "#最终输出csv文件。\n",
    "\n",
    "import csv\n",
    "import os\n",
    "import vcf\n",
    "wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "for i in wenjianjia_list:\n",
    "    dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{i}'\n",
    "    if not os.path.exists(dirs):\n",
    "        os.makedirs(dirs)\n",
    "\n",
    "for p in os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair'):\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/vcf_pair/{p}'\n",
    "    \n",
    "    bayes_filenames=os.listdir(prj_path)\n",
    "    type_list= []\n",
    "    dict_mut_num = {}#在几个样本中出现过。形成列表，在列表中的才算真正的突变。\n",
    "    #找出突变在几个样本中出现\n",
    "    for i in bayes_filenames:\n",
    "        if i.endswith('vcf'):\n",
    "            i_1 = i.split('.')[0]\n",
    "            reads_pos_dict = {}\n",
    "            path1 = prj_path+'/'+i\n",
    "            vcf_reader = vcf.Reader(filename=path1)\n",
    "            for record in vcf_reader:\n",
    "    #             print(record)\n",
    "    #             print(record.INFO)\n",
    "\n",
    "                alt_count = record.INFO['AO']\n",
    "                ref_count = record.INFO['RO']\n",
    "                ref = str(record).split(',')[2].split('=')[1]\n",
    "                type0 = record.INFO['TYPE']\n",
    "                position = str(record).split(',')[1].split('=')[1]\n",
    "                dp = record.INFO['DP']\n",
    "                if type0 not in type_list:\n",
    "                    type_list.append(type0)\n",
    "                for i in range(0,len(record.INFO['TYPE'])):\n",
    "                    alt1 = str(record).split('=')[4].replace(')','').replace('[','').replace(']','').replace(' ','')\n",
    "                    alt2 = alt1.split(\",\")#转成列表,碱基\n",
    "                    type2 = record.INFO['TYPE'][i]\n",
    "                    info = f'{ref}_{position}_{alt2[i]}'\n",
    "                    alt_count1 = alt_count[i]\n",
    "                    af = record.INFO['AO'][i]/(record.INFO['AO'][i] + record.INFO['RO'])\n",
    "    #                 print(af)\n",
    "                    if dp > 100:\n",
    "    #                     reads_pos_dict[info] = {'alt':alt_count1,'ref':ref_count,'type':type2,'allele_frequency':af}\n",
    "                        if info not in dict_mut_num:\n",
    "                            dict_mut_num[info] = 1\n",
    "                        else:\n",
    "                            dict_mut_num[info] += 1\n",
    "    #         print(reads_pos_dict)\n",
    "\n",
    "\n",
    "\n",
    "    filtered_list = []# 测序深度>100且出现在至少2个样本中的突变。\n",
    "    for k,v in dict_mut_num.items():\n",
    "        if v > 1:\n",
    "            filtered_list.append(k)\n",
    "    # print(filtered_list)\n",
    "\n",
    "\n",
    "    #根据filtered_list筛选突变，输出csv。\n",
    "    for i in bayes_filenames:\n",
    "        if i.endswith('vcf'):\n",
    "            i_1 = i.split('.')[0]\n",
    "            reads_pos_dict = {}\n",
    "            path1 = prj_path +'/'+i\n",
    "            vcf_reader = vcf.Reader(filename=path1)\n",
    "            for record in vcf_reader:\n",
    "    #             print(record)\n",
    "    #             print(record.INFO)\n",
    "\n",
    "                alt_count = record.INFO['AO']\n",
    "                ref_count = record.INFO['RO']\n",
    "                ref = str(record).split(',')[2].split('=')[1]\n",
    "                type0 = record.INFO['TYPE']\n",
    "                position = str(record).split(',')[1].split('=')[1]\n",
    "                dp = record.INFO['DP']\n",
    "                if type0 not in type_list:\n",
    "                    type_list.append(type0)\n",
    "                for i in range(0,len(record.INFO['TYPE'])):\n",
    "                    alt1 = str(record).split('=')[4].replace(')','').replace('[','').replace(']','').replace(' ','')\n",
    "                    alt2 = alt1.split(\",\")#转成列表,碱基\n",
    "                    type2 = record.INFO['TYPE'][i]\n",
    "                    info = f'{ref}_{position}_{alt2[i]}'\n",
    "                    alt_count1 = alt_count[i]\n",
    "                    af = record.INFO['AO'][i]/(record.INFO['AO'][i] + record.INFO['RO'])    \n",
    "                    if dp > 100:#即使有列表了也还是需要测序深度>100的这个条件，不然有可能是出现在样本多但是测序深度不行的。\n",
    "                        if info in filtered_list:\n",
    "                            if type2 == 'snp' or type2 == 'del' or type2 == 'ins':#筛选变异类型\n",
    "                                reads_pos_dict[info] = {'alt':alt_count1,'ref':ref_count,'type':type2,'allele_frequency':af,'sample_exist':dict_mut_num[info]}\n",
    "    #         print(reads_pos_dict)\n",
    "\n",
    "            with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{p}/{i_1}.csv','w') as f:#表格可以加个频率\n",
    "                writer = csv.writer(f)\n",
    "                writer.writerow([\"mutation\",\"alt_reads\",\"ref_reads\",\"type\",\"app\",\"allele_frequency\",\"sample_exist'\"])\n",
    "                for k,v in reads_pos_dict.items():\n",
    "                    writer.writerow([k,v['alt'],v['ref'],v['type'],'freebayes',v['allele_frequency'],v['sample_exist']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc88bc62-de72-4e38-8e3d-29acee0b6d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install pyvcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e32282e-e97a-46a1-8379-cb4848602bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "#04统计突变类型，3%-80%和80%以上的比例，先不管allele frequency，只看次数\n",
    "#二代测序中3%-80%的突变，对频率有错误影响，对蛋白分布没有影响。可以加一列属于哪个样本的。加上了，次数是准的了。\n",
    "\n",
    "#80%以下：\n",
    "import os\n",
    "import csv\n",
    "\n",
    "# wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "# for i in wenjianjia_list:\n",
    "#     dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/{i}'\n",
    "#     if not os.path.exists(dirs):\n",
    "#         os.makedirs(dirs)\n",
    "\n",
    "# wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "# for i in wenjianjia_list:\n",
    "#     dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/07mutation_proportion/{i}'\n",
    "#     if not os.path.exists(dirs):\n",
    "#         os.makedirs(dirs)\n",
    "\n",
    "for p in os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample'):\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{p}'\n",
    "\n",
    "    dict_003_08_next_generation = {}\n",
    "    for i in os.listdir(prj_path):\n",
    "        if i.endswith('.csv'):\n",
    "            csv_path = f'{prj_path}/{i}'\n",
    "            with open(csv_path) as csvfile:\n",
    "                csv_reader = csv.reader(csvfile)  \n",
    "                for row in csv_reader:\n",
    "                    if row[0] != 'mutation':\n",
    "                        first = row[0].split('_')[0]\n",
    "    #                     last = row[0].split('_')[2]\n",
    "    #                     position = row[0].split('_')[1]\n",
    "                        frequnecy = float(row[5])\n",
    "                        mutation = row[0]\n",
    "                        type1 = row[3]\n",
    "                        if 0.03 <= frequnecy <= 0.8:\n",
    "    #                         if len(first) == 1:#找的snv很奇怪，明明是单个突变非要写一堆，所以就不用type筛选了。\n",
    "    #                             dict_003_08_next_generation[mutation] = frequnecy\n",
    "                            if mutation not in dict_003_08_next_generation.keys():\n",
    "                                dict_003_08_next_generation[mutation] = {}\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "                            else:#一定要加，不然每次都是赋值一个空字典就没办法新加进来样本。\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "    # print(dict_003_08_next_generation)\n",
    "    # print(p)\n",
    "# ----------------------------------------------------------------------------------------------------------    \n",
    "# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!统计所有的突变及其数量，频率。 后面看在哪个突变上要用。             \n",
    "    with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06isnv_mutation_{p}.csv','w') as f:\n",
    "        writer = csv.writer(f)\n",
    "        writer.writerow(['isnv','sample','frequency','type'])\n",
    "        for k,v in dict_003_08_next_generation.items():\n",
    "            for s,info in v.items():\n",
    "                frequency = info[0]\n",
    "                type3 = info[1]\n",
    "                writer.writerow([k,s.split('.')[0],frequency,type3])\n",
    "            \n",
    "#-----------------------------------------------------------------------------------------------------------\n",
    "    \n",
    "#统计数量\n",
    "    dict_num_mutation = {'A_G':0,'A_T':0,'A_C':0,'G_A':0,'G_C':0,'G_T':0,'T_C':0,'T_A':0,'T_G':0,'C_A':0,'C_T':0,'C_G':0,'del':0,'ins':0}\n",
    "    for k,v in dict_003_08_next_generation.items():\n",
    "        ref = k.split('_')[0]\n",
    "        alt = k.split('_')[2]\n",
    "        for s in v.values():\n",
    "            type2 = s[1]\n",
    "    #         print(type2)\n",
    "        if type2 =='snp':\n",
    "            name = f'{ref}_{alt}'\n",
    "            add_count = len(v)\n",
    "        #     print(name)\n",
    "            for m,c in dict_num_mutation.items():\n",
    "                if name == m:\n",
    "                    dict_num_mutation[m] += add_count  #!!!赋值不可以只写c，一定要写成键的值。\n",
    "        if type2 =='del':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['del'] += add_count\n",
    "        if type2 =='ins':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['ins'] += add_count\n",
    "\n",
    "\n",
    "\n",
    "    print(dict_num_mutation)\n",
    "    count_all = 0\n",
    "    for k,v in dict_num_mutation.items():\n",
    "        count_all += v\n",
    "    print(count_all)\n",
    "    \n",
    "    dict_proportion =  dict.fromkeys(dict_num_mutation.keys())\n",
    "    for k in dict_proportion.keys():\n",
    "        dict_proportion[k] = {}\n",
    "        dict_proportion[k]['count'] = dict_num_mutation[k]\n",
    "        if count_all!= 0:\n",
    "            dict_proportion[k]['proportion'] = dict_proportion[k]['count']/count_all\n",
    "        else:\n",
    "            dict_proportion[k]['proportion'] = '/'\n",
    "    print(dict_proportion)  \n",
    "    with open(f\"/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/07mutation_proportion/07isnv_mutation_type_{p}.csv\",\"w\") as csvfile: \n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow(['mutation_type','count','proportion','group'])\n",
    "        for k,v in dict_proportion.items():\n",
    "            writer.writerow([k,v['count'],v['proportion'],'iSNV'])\n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "# #-----------------------------------------------------------------------\n",
    "# #80%上：\n",
    "\n",
    "    dict_003_08_next_generation = {}\n",
    "    for i in os.listdir(prj_path):\n",
    "        if i.endswith('.csv'):\n",
    "            csv_path = f'{prj_path}/{i}'\n",
    "            with open(csv_path) as csvfile:\n",
    "                csv_reader = csv.reader(csvfile)  \n",
    "                for row in csv_reader:\n",
    "                    if row[0] != 'mutation':\n",
    "                        first = row[0].split('_')[0]\n",
    "    #                     last = row[0].split('_')[2]\n",
    "    #                     position = row[0].split('_')[1]\n",
    "                        frequnecy = float(row[5])\n",
    "                        mutation = row[0]\n",
    "                        type1 = row[3]\n",
    "                        if  frequnecy > 0.8:\n",
    "    #                         if len(first) == 1:#找的snv很奇怪，明明是单个突变非要写一堆，所以就不用type筛选了。\n",
    "    #                             dict_003_08_next_generation[mutation] = frequnecy\n",
    "                            if mutation not in dict_003_08_next_generation.keys():\n",
    "                                dict_003_08_next_generation[mutation] = {}\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "                            else:#一定要加，不然每次都是赋值一个空字典就没办法新加进来样本。\n",
    "                                dict_003_08_next_generation[mutation][i] = [frequnecy,type1]\n",
    "\n",
    "    #统计数量\n",
    "    dict_num_mutation = {'A_G':0,'A_T':0,'A_C':0,'G_A':0,'G_C':0,'G_T':0,'T_C':0,'T_A':0,'T_G':0,'C_A':0,'C_T':0,'C_G':0,'del':0,'ins':0}\n",
    "    for k,v in dict_003_08_next_generation.items():\n",
    "        ref = k.split('_')[0]\n",
    "        alt = k.split('_')[2]\n",
    "        for s in v.values():\n",
    "            type2 = s[1]\n",
    "    #         print(type2)\n",
    "        if type2 =='snp':\n",
    "            name = f'{ref}_{alt}'\n",
    "            add_count = len(v)\n",
    "        #     print(name)\n",
    "            for m,c in dict_num_mutation.items():\n",
    "                if name == m:\n",
    "                    dict_num_mutation[m] += add_count  #!!!赋值不可以只写c，一定要写成键的值。\n",
    "        if type2 =='del':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['del'] += add_count\n",
    "        if type2 =='ins':\n",
    "            add_count = len(v)\n",
    "\n",
    "            dict_num_mutation['ins'] += add_count\n",
    "\n",
    "\n",
    "\n",
    "    # print(dict_num_mutation)\n",
    "\n",
    "    #算比例的分母\n",
    "    count_all = 0\n",
    "    for k,v in dict_num_mutation.items():\n",
    "        count_all += v\n",
    "    # print(count_all)\n",
    "    dict_proportion =  dict.fromkeys(dict_num_mutation.keys())\n",
    "    for k in dict_proportion.keys():\n",
    "        dict_proportion[k] = {}\n",
    "        dict_proportion[k]['count'] = dict_num_mutation[k]\n",
    "        if count_all!= 0:\n",
    "            dict_proportion[k]['proportion'] = dict_proportion[k]['count']/count_all\n",
    "        else:\n",
    "            dict_proportion[k]['proportion'] = '/'\n",
    "    print(dict_proportion)  \n",
    "    with open(f\"/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/07mutation_proportion/07snp_mutation_type_{p}.csv\",\"w\") as csvfile: \n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow(['mutation_type','count','proportion','group'])\n",
    "        for k,v in dict_proportion.items():\n",
    "            writer.writerow([k,v['count'],v['proportion'],'SNP'])\n",
    "\n",
    "    #!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!统计所有的突变及其数量，频率。 后面看在哪个突变上要用。             \n",
    "    with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06snp_mutation_{p}.csv','w') as f:\n",
    "        writer = csv.writer(f)\n",
    "        writer.writerow(['isnv','sample','frequency','type'])\n",
    "        for k,v in dict_003_08_next_generation.items():\n",
    "            for s,info in v.items():\n",
    "                frequency = info[0]\n",
    "                type3 = info[1]\n",
    "                writer.writerow([k,s.split('.')[0],frequency,type3])\n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2eb0b72-f9f2-4a4a-9c2c-fa798d036898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 05 翻译蛋白质，要加3-80%的。isnv的\n",
    "#分批次\n",
    "#非同义氨基酸突变的具体情况和不同蛋白非同义突变的数量。\n",
    "#插入加了。\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "for i in wenjianjia_list:\n",
    "    dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{i}'\n",
    "    if not os.path.exists(dirs):\n",
    "        os.makedirs(dirs)\n",
    "dict_pep_cds = {'99_518':'NS1_1C','628_1002':\"NS2_1B\",'1140_2315':\"N\",'2348_3073':\"P\",'3233_4003':\"M\",'4274_4468':\"SH\",'4659_5555':\"G\",'5632_7356':\"F\",'7576_8160':\"M2\",'8468_14965':\"L\"}\n",
    "\n",
    "csv_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist')\n",
    "for c in csv_list:\n",
    "    if c.endswith('csv'):\n",
    "        c1 = c.split('_')[2].split('.')[0]\n",
    "        csv_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06isnv_mutation_{c1}.csv'\n",
    "        with open(csv_path) as csvfile:\n",
    "            dict_info = {}\n",
    "            info_list = []\n",
    "            csv_reader = csv.reader(csvfile)  \n",
    "            for row in csv_reader:\n",
    "                if row[0] != 'isnv':\n",
    "                    mut = row[0]\n",
    "                    if mut not in info_list:\n",
    "                        info_list.append(mut)\n",
    "        dict_info = dict.fromkeys(info_list)\n",
    "\n",
    "\n",
    "        for i in dict_info.keys():#嵌套字典可以这样赋值：1、dict_info = dict.fromkeys(info_list)2、每个键都赋值成空字典，再正常加就好了\n",
    "            dict_info[i] = {}\n",
    "\n",
    "\n",
    "    #----------------\n",
    "    #新加的，统计每个突变的样本数量。\n",
    "        df = pd.read_csv(csv_path)\n",
    "        df_counts = df.isnv.value_counts()\n",
    "        #print(df_counts)\n",
    "        outputpath=f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/mut_sample_number/{c1}_mut_sample_number.csv'\n",
    "        df_counts.to_csv(outputpath,sep=',',index=True,header=False)\n",
    "\n",
    "        dict_mut_num = {}\n",
    "        \n",
    "        mut_number_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/mut_sample_number/{c1}_mut_sample_number.csv'\n",
    "        with open(mut_number_path) as csvfile:\n",
    "            csv_reader = csv.reader(csvfile) \n",
    "            for row in csv_reader:\n",
    "                isnv = row[0]\n",
    "                num = row[1]\n",
    "                dict_mut_num[isnv] = num\n",
    "        # print(dict_mut_num)\n",
    "\n",
    "\n",
    "    #--------------\n",
    "        with open(csv_path) as csvfile:\n",
    "            csv_reader = csv.reader(csvfile)  \n",
    "            for row in csv_reader:\n",
    "                if row[0] != 'isnv':\n",
    "                    mut = row[0]\n",
    "                    pos = int(row[0].split('_')[1])\n",
    "                    fre = float(row[2])\n",
    "                    type4 = row[3]\n",
    "                    first = row[0].split('_')[0]\n",
    "                    last = row[0].split('_')[2]\n",
    "                    #分一下位置，del和ins的有可能结束位置和开始位置不在同一段蛋白里面。不太可能\n",
    "                    if type4 == 'del':#按照突变前的长度算\n",
    "                        pos_end = pos + len(first) -1\n",
    "                    if type4 == 'ins':#按照突变后的长度算\n",
    "                        pos_end = pos + len(last) -1          \n",
    "                    for k,v in dict_pep_cds.items():\n",
    "                        start = int(k.split('_')[0])\n",
    "                        end = int(k.split('_')[1])\n",
    "                        if type4 == 'snp':\n",
    "                            if start <= pos <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "                                dict_info[mut]['product'] = v\n",
    "                                dict_info[mut]['type'] = type4\n",
    "                                dict_info[mut]['start'] = start\n",
    "                                dict_info[mut]['end'] = end\n",
    "                                dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "        #                         dict_info[mut]['frequency'] = fre 频率先不考虑了，有好多样本的\n",
    "                        if type4 == 'del' or type4 == 'ins':\n",
    "                            if start <= pos <= end and start <= pos_end <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "                                dict_info[mut]['product'] = v\n",
    "                                dict_info[mut]['type'] = type4\n",
    "                                dict_info[mut]['start'] = start\n",
    "                                dict_info[mut]['end'] = end\n",
    "        #                         dict_info[mut]['frequency'] = fre\n",
    "                                dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # print(dict_info)\n",
    "        ##---------------------------------------------------------------------------------------\n",
    "        #去除不在cds范围内的碱基突变                   \n",
    "        dict_info_feikong = {}            \n",
    "        for i in dict_info.keys():\n",
    "            if dict_info[i]:\n",
    "                dict_info_feikong[i] = dict_info[i]\n",
    "        # print(dict_info_feikong)\n",
    "\n",
    "        #获取突变前后的三个碱基，蛋白位置不知道，可以用什么蛋白\n",
    "        dict_aa_mut = {}#非同义突变的字典\n",
    "        dict_aa_synonymous_mut = {}\n",
    "        ref_seq = 'ACGCGAAAAAATGCGTACAACAAACTTGCGTAAACCAAAAAAATGGGGCAAATAAGAATTTGATAAGTACCACTTAAATTTAACTCCCTTGGTTAGAGATGGGCAGCAATTCATTGAGTATGATAAAAGTTAGATTACAAAATTTGTTTGACAATGATGAAGTAGCATTGTTAAAAATAACATGCTATACTGACAAATTAATACATTTAACTAATGCATTGGCTAAGGCAGTGATACATACAATCAAATTGAATGGCATTGTATTTGTGCATGTTATTACAAGTAGTGATATTTGCCCTAATAATAATATTGTAGTGAAATCCAATTTCACAACAATGCCAGTGTTACAAAATGGAGGTTATATATGGGAAATGATGGAATTAACACACTGCTCTCAACCTAATGGCCTAATAGATGACAATTGTGAAATTAAATTCTCCAAAAAACTAAGTGATTCAACAATGACCAATTATATGAATCAATTATCTGAATTACTTGGATTTGATCTTAATCCATAAATTATAATAAATATCAACTAGCAAATCAATGTCACTAACACCATTAGTTAATATAAAACTTGACAGAAGATAAAAATGGGGCAAATAAATCAATTCAGCCGACCCAACCATGGACACAACACACAATGACACCACACCACAAAGACTGATGATCACAGACATGAGACCATTGTCACTTGAGACTATAATAATATCACTAACCAGAGACATCATAACACACAGATTTATATACTTGATAAATCATGAATGTATAGTGAGAAAACTTGATGAAAGACAGGCCACATTTACATTCCTGGTCAACTATGAAATGAAACTATTGCACAAAGTGGGAAGCACTAAATACAAAAAATATACTGAATACAACACAAAATATGGCACTTTTCCTATGCCAATATTTATCAATCATGATGGGTTCTTAGAATGCATTGGCATTAAGCCTACAAAGCACACTCCCATAATATACAAGTATGATCTCAATCCATGAATTTCAACACAAGAGTCACACAATCTGAAATAACAACTTCATGCATAACCACACTCCATAGTTCAAATGGAGCCTGAAAATTATAGTAATTTAAAATTAAGGAGAGACATAAGATGAAAGATGGGGCAAATACAAAAATGGCTCTTAGCAAAGTCAAGTTGAATGATACACTCAACAAAGATCAACTTCTGTCATCCAGCAAATACACCATCCAACGGAGCACAGGAGATAGTATTGATACTCCTAATTATGATGTGCAGAAACACATCAATAAGTTATGTGGCATGTTATTAATCACAGAAGATGCTAATCATAAATTCACTGGGTTAATAGGTATGTTATATGCTATGTCTAGATTAGGAAGAGAAGACACCATAAAAATACTCAGAGATGCGGGATATCATGTAAAAGCAAATGGAGTGGATGTAACAACACATCGTCAAGATATTAATGGGAAAGAAATGAAATTTGAAGTGTTAACATTGTCAAGCTTAACAACTGAAATTCAAATCAACATTGAGATAGAATCTAGAAAATCCTACAAAAAAATGCTAAAAGAAATGGGAGAGGTAGCTCCAGAATACAGGCATGACTCTCCTGATTGTGGGATGATAATATTATGTATAGCGGCATTAGTAATAACCAAATTAGCAGCAGGGGATAGATCTGGTCTTACAGCTGTGATTAGGAGGGCTAATAATGTCCTAAAAAATGAAATGAAACGTTATAAAGGCTTACTACCCAAGGATATAGCCAACAGCTTCTATGAAGTGTTTGAAAAATATCCTCACTTTATAGATGTTTTTGTTCATTTTGGTATAGCACAATCTTCTACCAGAGGTGGCAGTAGAGTTGAAGGGATTTTTGCTGGATTGTTTATGAATGCCTATGGTGCAGGGCAAGTGATGTTACGGTGGGGGGTCTTAGCAAAATCAGTTAAAAATATTATGCTAGGACACGCTAGTGTGCAAGCAGAAATGGAACAAGTTGTGGAGGTTTATGAATATGCCCAAAAATTGGGTGGAGAAGCAGGGTTCTACCATATATTGAACAACCCAAAAGCATCATTATTGTCTTTGACTCAATTTCCTCACTTCTCCAGTGTAGTATTAGGCAATGCTGCTGGCCTAGGCATAATGGGAGAATACAGAGGTACACCAAGGAATCAAGATCTATATGATGCTGCAAAAGCATATGCTGAACAACTCAAAGAAAATGGTGTGATTAACTACAGTGTATTAGACTTGACAGCAGAAGAACTAGAGGCTATCAAACATCAGCTTAATCCAAAAGATAATGATGTAGAGCTTTGAGTTAATAAAAAAAATGGGGCAAATAAAACATCATGGAAAAGTTTGCTCCTGAATTCCATGGAGAAGATGCAAACAACAGAGCTACCAAATTCCTAGAATCAATAAAGGGCAAATTCACATCACCTAAAGATCCCAAGAAAAAAGATAGTATCATATCTGTCAACTCAATAGATATAGAAGTAACCAAAGAAAGCCCTATAACATCAAATTCAACCATTATAAACCCAACAAATGAGACAGATGATACTGTAGGGAACAAGCCCAATTATCAAAGAAAACCTCTAGTAAGTTTCAAAGAAGACCCTACGCCAAGTGATAATCCCTTTTCAAAACTATACAAAGAAACCATAGAAACATTTGATAACAATGAAGAAGAATCTAGCTATTCATATGAAGAAATAAATGATCAGACAAACGATAATATAACAGCAAGATTAGATAGGATTGATGAAAAATTAAGTGAAATACTAGGAATGCTTCACACATTAGTAGTAGCGAGTGCAGGACCTACATCTGCTCGGGATGGTATAAGAGATGCCATGGTTGGTTTAAGAGAAGACATGATAGAAAAAATCAGAACTGAAGCATTAATGACCAATGACAGACTAGAAGCTATGGCAAGACTCAGGAATGAGGAAAGTGAAAAGATGGCAAAAGACACATCAGATGAAGTGTCTCTCAATCCAACATCAGAGAAATTGAACAACCTGTTGGAAGGGAATGATAGTGACAATGATCTATCACTTGATGATTTCTGATCAGTTACCAATCTGTACATCAACACACAACACCAACAGAAGACCAACAAACAAACCAACTCACCCATCCAACCAAACATCTATACGCCAATCAGCCAATCCAAAACTAGCCACCCGGAAAAAATAGATACTATAGTTACAAAAAAAGATGGGGCAAATATGGAAACATACGTGAACAAACTTCACGAAGGCTCCACATACACAGCTGCTGTTCAATACAATGTCTTAGAAAAAGACGATGACCCTGCATCACTTACAATATGGGTGCCCATGTTCCAATCATCCATGCCAGCAGATTTACTTATAAAAGAACTAGCTAATGTCAACATACTAGTGAAACAAATATCCACACCCAATGGACCTTCATTAAGAGTCATGATAAACTCAAGAAGTGCAGTGCTAGCACAAATGCCCAGCAAATTTACCATATGTGCCAATGTGTCCTTGGATGAAAGAAGCAAGCTGGCATATGATGTAACCACACCCTGTGAAATCAAGGCATGTAGTCTAACATGCCTAAAATCAAAAAATATGTTAACTACAGTTAAAGATCTCACTATGAAAACACTCAACCCAACACATGACATCATTGCTTTATGTGAATTTGAAAATATAGTAACATCAAAAAAAGTCATAATACCAACATACCTAAGATCCATCAGTGTCAGAAATAAAGATCTGAACACACTTGAAAATATAACAACCACTGAATTCAAAAATGCCATCACAAATGCAAAAATCATCCCTTACTCAGGATTACTGTTAGTCATCACAGTGACTGACAACAAAGGAGCATTCAAATACATAAAGCCACAAAGTCAATTTATAGTAGATCTTGGAGCTTACCTAGAAAAAGAAAGTATATATTATGTTACAACAAATTGGAAGCACACAGCTACACGATTTGCAATCAAACCCATGGAAGATTAACCTTTTTCTTCTACATCAGTGAGTTGATTCATACAAACTTTCTACCTACATTCTTCACTTCACCATCATAATCACCAACCCTCTGTGGTTCAACTAATCAAACAAAACCCATCTGGAGCCTCAGATCATCCCAAGTCATTGTTCATCAGATCTAGTACTCAAATAAGTTAATAAAAATATCCACATGGGGCAAATAATCATTGGAGGAAATCCAACTAATCACAATATCTGTCAACATAGACAAGTCAACACGCCAGGCAAAATCAACCAATGGAAAATACATCCATAACAATAGAATTCTCAAGCAAATTCTGGCCTTACTTTACACTAATACACATGATAACAACAATAATCTCTTTGCTAATCATAATCTCCATCATGATTGCAATACTGAACAAACTCTGTGAATATAACGTATTCCATAACAAAACCTTTGAGCTACCAAGAGCTCGAGTCAATACATAGCATTCACCAATCTGATGGCACAAAACAGTAACCTTGCATTTGTAAGTGAACAACCCTCACCTCTTTACAAAACCACATCAACATCTCACCATGCAAGCCATCATCCATATTATAAAGTAGTTAATTAAAAATAATCATAACAATGAACTAAGATATTAAGACTAACAATAACGTTGGGGCAAATGCAAACATGTCCAAAAACAAGGACCAACGCACCACCAAGACACTAGAAAAGACCTGGGACACTCTCAATCATCTATTATTCATATCATCGTGCTTATACAAGTTAAATCTTAAATCTATAGCACAAATCACATTATCCATTCTGGCAATGATAATCTCAACTTCACTTATAATTGCAGCCATCATATTCATAGCCTCGGCAAACCACAAAGTCACACTAACAACTGCAATCATACAAGATGCAACAAGCCAGATCAAGAACACAACCCCAACATACCTCACCCAGAATCCCCAGCTTGGAATCAGCTTCTCCAATCTGTCTGAAACTACATCACAAACCACCACCATACTAGCTTCAACAACACCAAGTGTCAAGTCAACCCTGCAATCCACAACAGTCAAGACCAAAAACACAACAACAACCAAAATACAACCCAGCAAGCCCACCACAAAACAACGCCAAAACAAACCACCAAACAAACCCAATAATGATTTTCACTTTGAAGTGTTCAACTTTGTACCTTGCAGCATATGCAGCAACAATCCAACCTGCTGGGCTATCTGTAAAAGAATACCAAACAAAAAACCTGGAAAGAAAACCACCACCAAGCCCACAAAAAAACCAACCATCAAGACAACCAAAAAAGATCTCAAACCTCAAACCACAAAACCAAAGGAAGTACCTACCACCAAGCCCACAGAAAAGCCAACCATCAACACCACCAAAACAAACATCAGAACTACACTGCTCACCAACAATACCACAGGAAATCCAGAACACACAAGTCAAAAGGGAACCCTCCACTCAACCTCCTCCGATGGCAATCCAAGCCCTTCACAAGTCTATACAACATCCGAGTACCTATCACAACCTCCATCTCCATCCAACACAACAAACCAGTAGTCATTAAAAAGCGTATTATTGCAAAAAGCCATGACCAAATCAACCAGAATCAAAATCAACTCTGGGGCAAATAACAATGGAGTTGCCAATCCTCAAAACAAATGCAATTACCGCAATCCTTGCTGCAGTCACACTCTGTTTTGCTTCCAGTCAAAACATCACTGAAGAATTTTATCAAACAACATGCAGTGCAGTCAGCAAAGGCTATCTTAGTGCTCTAAGAACTGGTTGGTATACTAGTGTTATAACTATAGAATTAAGTAATATCAAGGAAAATAAGTGTAATGGAACAGACGCTAAGGTAAAATTGATAAAACAAGAATTAGATAAATATAAAAGTGCTGTAACAGAATTGCAGTTGCTCATGCAAAGCACACCGGCAACCAACAATCGAGCCAGAAGAGAACTACCAAGGTTTATGAATTATACACTCAACAATACCAAAAATACCAATGTAACATTAAGCAAGAAAAGGAAAAGAAGATTTCTTGGCTTTTTGTTAGGTGTTGGATCTGCAATCGCCAGTGGCATTGCTGTATCTAAGGTCCTGCACCTAGAAGGGGAAGTGAACAAAATCAAAAGTGCTCTACTATCCACAAACAAGGCTGTAGTCAGCTTATCAAATGGAGTTAGTGTCTTAACCAGCAAAGTGTTAGACCTCAAAAACTATATAGATAAACAGTTGTTACCTATTGTGAACAAGCAAAGCTGTAGCATATCAAACATTGAAACTGTGATAGAGTTCCAACAAAAGAACAACAGACTACTAGAGATTACCAGGGAATTTAGTGTTAATGCAGGTGTAACTACACCTGTAAGCACTTATATGTTAACAAATAGTGAATTATTATCATTAATCAATGATATGCCTATAACAAATGATCAGAAAAAGTTAATGTCCAACAATGTTCAAATAGTTAGACAGCAAAGTTACTCTATCATGTCCATAATAAAGGAGGAAGTCTTAGCATATGTAGTACAATTACCACTATATGGTGTAATAGATACACCTTGTTGGAAACTGCACACATCCCCTCTATGTACAACCAACACAAAGGAAGGGTCCAACATCTGTTTAACAAGAACCGACAGAGGATGGTACTGTGACAATGCAGGATCAGTATCTTTCTTCCCACTAGCTGAAACATGTAAAGTTCAATCGAATCGAGTATTTTGTGACACAATGAACAGTTTAACATTACCAAGTGAAGTAAATCTCTGCAACATTGACATATTCAACCCCAAATATGATTGCAAAATTATGACTTCAAAAACAGATGTAAGCAGCTCCGTTATCACATCTCTAGGAGCCATTGTGTCATGCTATGGCAAAACTAAATGTACAGCATCCAATAAAAATCGTGGAATCATAAAGACATTTTCTAACGGGTGCGATTATGTATCAAATAAGGGGGTTGACACTGTGTCTGTAGGTAATACATTATATTATGTAAATAAGCAAGAAGGCAAAAGTCTCTATGTAAAAGGTGAACCAATAATAAATTTCTATGACCCATTAGTGTTCCCCTCTGATGAATTTGATGCATCAATATCTCAAGTCAATGAGAAGATTAACCAGAGCCTAGCATTTATTCGTAAATCCGATGAATTATTACATAATGTAAATGCTGGTAAATCCACCATAAATATCATGATAACTACTATAATTATAGTGATTATAGTAATATTGTTATCATTAATTGCCGTTGGACTGCTCCTATACTGCAAGGCCAGAAGCACACCAGTCACACTAAGCAAGGATCAACTGAGTGGTATAAATAATATTGCATTTAGTAACTAAATAAAAATAGCACCTAATCATGTTCTTACAATGGTTTCATATCTGCTCATAGACAACCCATCTATCATTGGATTTTCTTAAAATCTGAACTTCATCGAAACTCTCATCTATAAACCATCTCACTTACATTATTTAAGTAGATTCCTAGTTTATAGTTATATAAAACAATTGAATACCAGATTAACTTACTATTTGTAAAAAATGAGAACTGGGGCAAATATGTCACGAAGGAATCCTTGCAAATTTGAAATTCGAGGTCATTGCTTGAATGGTAAGAGGTGTCATTTTAGTCATAATTATTTTGAATGGCCACCCCATGCACTGCTTGTAAGACAAAACTTTATGTTAAACAGAATACTTAAGTCTATGGATAAAAGCATCGATACTTTATCAGAAATAAGTGGAGCTGCAGAGTTGGACAGAACAGAAGAGTATGCCCTCGGTGTAGTTGGAGTGCTAGAGAGTTATATAGGATCTATAAATAATATAACTAAACAATCAGCATGTGTTGCCATGAGCAAACTCCTCACTGAACTCAACAGTGATGACATCAAAAAACTGAGGGACAATGAAGAGCCAAATTCACCCAAGATAAGAGTGTACAATACTGTCATATCATATATTGAAAGCAACAGGAAAAACAATAAACAAACTATCCATCTGTTAAAAAGATTGCCAGCAGACGTATTGAAGAAAACCATAAAAACCACATTGGATATCCACAAGAGCATAACCATCAATAACCCAAAAGAATCAACTGTTAGTGATATAAACGACCATGCCAAAAATAATGATACTACCTGACAAATATCCTTGTAGTATAAATTCCATACTAATAACAAGTAGTTGTAGAGTTACTATGTATAATCAAAAGAACACACTATATTTCAATCAAAACAACCAAAATAACCATATATACTCACCGAATCAACCATTCAATGAAATCCATTGGACCTCTCAAGACTTGATTGATGCAATTCAAAATTTTCTACAACATCTAGGTATTACTGATGATATATACACAATATATATATTAGTGTCATAACACTCAATCCTAATGCTTACCACATCATCAAACTATTAACTCAAACAATTCAAGCCATGGGACAAAATGGATCCCATTATTAATGGAAATTCTGCTAATGTGTATCTAACCGATAGTTATTTAAAAGGTGTTATTTCTTTCTCAGAATGTAATGCTTTAGGAAGTTACATATTCAATGGTCCTTATCTCAAAAATGATTATACCAACTTAATTAGTAGACAAAATCCATTAATAGAACACATAAATCTAAAGAAACTAAATATAACACAGTCCTTAATGTCTAAGTATCATAAAGGTGAAATAAAAATAGAAGAACCTACTTATTTTCAGTCATTACTTATGACATACAAGAGTATGACCTCGTTAGAACAGATTACTACCACTAATTTACTTAAAAAGATAATAAGAAGAGCTATAGAAATTAGTGATGTCAAAGTCTATGCTATATTGAATAAACTGGGGCTTAAAGAAAAAGACAAGATTAAATCCAACAATGGACAAGATGAAGACAACTCAGTTATTACAACCATAATCAAAGATGATATACTTTTAGCTGTTAAGGATAATCAATCTCATCTTAAAGCAGTCAAAAATCACTCTACAAAACAAAAAGATACAATCAAAACAACACTCTTGAAGAAATTAATGTGTTCAATGCAACATCCTCCATCATGGTTAATACATTGGTTTAATTTATACACAAAATTAAACAACATATTAACACAGTATCGATCAAGTGAGGTAAAAAACCATGGTTTTATATTGATAGACAATCATACTCTCAATGGATTCCAATTTATTTTGAATCAATATGGTTGTATAGTTTATCATAAGGAACTCAAAAGAATTACTGTGACAACCTATAATCAATTCTTGACATGGAAAAATATTAGCCTTAGTAGATTAAATGTTTGTTTAATTACATGGATTAGTAACTGTTTGAACACATTAAATAAAAGCTTAGGCTTAAGATGCGGATTCAATAATGTTATCTTGACACAACTATTCCTCTATGGAGATTGTATACTAAAACTATTCCACAATGAGGGGTTCTACATAATAAAAGAGGTAGAGGGATTTATTATGTCTCTAATTTTAAATATAACAGAAGAAGATCAATTCAGAAAACGGTTTTATAATAGTATGCTCAACAACATCACAGATGCTGCTAATAAAGCTCAGAAAAATCTGCTATCAAGAGTATGTCATACATTATTAGATAAGACAGTATCCGATAATATAATAAATGGCAGATGGATAATTCTATTAAGTAAGTTCCTTAAATTAATTAAGCTTGCAGGTGACAATAACCTTAACAATCTGAGTGAATTATATTTTTTGTTCAGAATATTTGGACACCCAATGGTAGATGAAAGACAAGCCATGGATGCTGTTAAAGTTAATTGCAACGAGACCAAATTTTACTTGTTAAGCAGTTTGAGTATGTTAAGAGGTGCCTTTATATATAGAATTATAAAAGGATTTGTAAATAATTACAACAGATGGCCTACTTTAAGGAATGCTATTGTTTTACCCTTAAGATGGTTAACTTACTATAAACTAAACACTTATCCTTCCTTGTTGGAACTTACAGAAAGAGATTTGATTGTTTTATCAGGACTACGTTTCTATCGTGAGTTTCGGTTGCCTAAAAAAGTGGATCTTGAAATGATCATAAATGATAAGGCTATATCACCTCCTAAAAATTTGATATGGACTAGTTTCCCTAGAAATTATATGCCGTCACACATACAAAATTATATAGAACATGAAAAATTAAAATTTTCCGAGAGTGATAAATCAAGAAGAGTATTAGAGTACTATTTAAGAGATAACAAATTCAATGAATGTGATTTATACAACTGTGTAGTTAATCAAAGTTATCTTAACAACCCTAATCATGTGGTATCTTTGACAGGCAAAGAAAGAGAACTCAGTGTAGGTAGAATGTTTGCAATGCAACCAGGAATGTTCAGACAAGTTCAAATATTAGCAGAGAAAATGATAGCTGAAAACATTTTACAATTCTTTCCTGAAAGTCTTACAAGATATGGTGATCTAGAACTACAGAAAATATTAGAATTGAAAGCAGGAATAAGTAACAAATCAAATCGTTACAATGATAATTACAACAATTACATTAGTAAGTGCTCTATCATCACAGATCTCAGCAAATTCAATCAAGCATTTCGATATGAAACATCATGTATTTGTAGTGATGTACTGGATGAACTGCATGGTGTACAATCTCTATTTTTCTGGTTACATTTAGCTATTCCTCATGTCACAATAATATGCACATATAGGCATGCACCCCCCTATATAAGAGATCATATTGTAGATCTTAACAATGTAGATGAACAAAGTGGATTATATAGATATCATATGGGTGGTATCGAAGGGTGGTGTCAAAAACTATGGACCATAGAAGCTATATCACTATTGGATCTAATATCTCTCAAAGGGAAATTCTCAATTACTGCTTTAATTAATGGTGACAATCAATCAATAGATATAAGTAAACCAGTCAGACTCATGGAAGGTCAAACTCATGCTCAAGCAGATTATTTGCTAGCATTAAATAGTCTTAAATTACTGTATAAAGAGTATGCAGGCATAGGCCACAAATTAAAAGGAACTGAGACTTATATATCAAGAGATATGCAATTTATGAGTAAAACAATTCAACATAACGGTGTATATTACCCAGCTAGTATAAAGAAAGTCCTAAGAGTGGGACCGTGGATAAACACTATACTTGATGATTTCAAAGTGAGTCTAGAATCTATAGGTAGTTTGACACAAGAATTAGAATATAGAGGAGAAAGTCTATTATGCAGTTTAATATTTAGAAATGTATGGTTATATAATCAAATTGCTTTACAACTAAAAAATCATGCATTATGTAACAATAAATTATATTTGGACATATTAAAGGTTCTGAAACACTTAAAAACCTTTTTTAATCTTGATAATATTGATACAGCATTAACATTGTATATGAATTTGCCCATGTTATTTGGTGGTGGTGATCCCAACTTGTTATATCGAAGTTTCTATAGAAGAACTCCTGATTTCCTCACAGAGGCTATAGTTCACTCTGTGTTCATACTTAGTTATTATACAAACCATGATTTAAAAGATAAACTTCAAGATCTGTCAGATGATAGATTGAATAAGTTCTTAACATGCATAATCACATTTGACAAAAACCCTAATGCTGAATTCGTAACATTGATGAGAGATCCTCAAGCTTTAGGGTCTGAGAGACAAGCTAAAATTACTAGCGAAATCAATAGACTGGCAGTTACTGAGGTTTTGAGCACAGCTCCAAACAAAATATTCTCCAAAAGTGCACAACACTATACCACTACAGAGATAGATCTAAATGATATTATGCAAAATATAGAACCTACATATCCTCACGGGCTAAGAGTTGTTTATGAAAGTTTACCCTTTTATAAAGCAGAGAAAATAGTAAATCTTATATCCGGTACAAAATCTATAACTAACATACTGGAAAAGACTTCTGCCATAGACTTAACAGATATTGATAGAGCCACTGAGATGATGAGGAAAAACATAACTTTGCTTATAAGGATATTTCCATTAGATTGTAACAGAGATAAAAGGGAAATATTGAGTATGGAAAACCTAAGTATTACTGAATTAAGCAAATATGTTAGGGAAAGATCTTGGTCTTTATCCAATATAGTTGGTGTTACATCACCTAGTATCATGTATACAATGGACATCAAATATACAACAAGCACTATAGCTAGTGGCATAATCATAGAGAAATATAATGTTAACAGTTTAACACGTGGTGAGAGAGGACCCACTAAACCATGGGTTGGTTCATCTACACAAGAGAAAAAAACAATGCCAGTTTATAATAGACAAGTTTTAACCAAAAAACAGAGAGATCAAATAGATCTATTAGCAAAATTGGATTGGGTGTATGCATCTATAGATAACAAGGATGAATTCATGGAAGAACTTAGCATAGGAATTCTTGGGTTAACATATGAGAAAGCCAAAAAATTATTTCCACAATATTTAAGTGTTAACTATTTGCATCGCCTTACAGTCAGTAGTAGACCATGTGAATTCCCTGCATCAATACCAGCTTATAGAACTACAAATTATCACTTTGATACTAGCCCTATTAATCGCATATTAACAGAAAAGTATGGTGATGAAGATATTGATATAGTATTCCAAAACTGTATAAGCTTTGGCCTTAGCTTAATGTCAGTAGTAGAACAATTTACTAATGTATGTCCTAACAGAATTATTCTTATACCTAAGCTTAATGAGATACATTTAATGAAACCTCCCATATTCACAGGTGATGTTGATATTCACAAGTTAAAACAAGTGATACAAAAACAGCATATGTTTTTACCAGACAAAATAAGTTTGACTCAATATGTGGAATTATTCTTAAGTAATAAAACACTCAAATCTGGATCTCATGTTAATTCTAATTTAATATTGGCGCATAAGATATCTGACTATTTTCATAATACTTACATTTTAAGTACTAATTTAGCTGGACATTGGATTCTGATTATACAACTTATGAAAGATTCTAAGGGTATTTTTGAAAAAGATTGGGGAGAGGGATATATAACTGATCATATGTTCATTAATTTGAAAGTTTTCTTCAATGCTTATAAGACCTATCTCTTGTGTTTTCATAAAGGTTACGGCAGAGCAAAGCTGGAGTGTGATATGAATACTTCAGATCTCCTATGTGTATTGGAATTAATAGACAGTAGTTATTGGAAGTCTATGTCTAAGGTATTTTTAGAACAAAAAGTTATCAAATACATTCTTAGCCAGGATGCAAGTTTACATAGAGTAAAAGGATGTCATAGCTTCAAACTATGGTTTCTTAAACGTCTTAATGTAGCAGAATTCACAGTTTGCCCTTGGGTTGTTAACATAGATTATCATCCAACACATATGAAAGCAATATTAACTTATATAGATCTTGTTAGAATGGGATTGATAAATATAGATAAAATATACATTAAAAATAAACACAAATTCAATGATGAATTTTATACTTCTAATCTCTTTTACATTAATTATAACTTCTCAGATAATACTCATCTATTAACTAAACATATAAGGATTGCTAATTCTGAATTAGAAAATAATTACAACAAATTATATCATCCTACACCAGAAACCCTAGAAAATATACTAACCAATCCGGTTAAATGTAATGACAAAAAGACACTGAATGACTATTGTATAGGTAAAAATGTTGACTCAATAATGTTACCATTGTTATCTAATAAGAAGCTTATTAAATCGCCTACAATGATTAGAACCAATTACAGCAAACAAGATTTGTATAATTTATTTCCTACGGTTGTGATTGATAAAATTATAGATCATTCAGGTAATACAGCCAAATCTAACCAACTTTACACTACTACTTCTCATCAAATACCTTTAGTGCACAATAGCACATCACTTTATTGCATGCTTCCTTGGCATCATATTAATAGATTCAATTTTGTATTTAGTTCTACAGGTTGTAAAATTAGTATAGAGTATATTTTAAAAGACCTTATAATTAAAGATCCTAATTGTATAGCATTCATAGGTGAAGGAGCAGGGAATTTATTATTGCGTACAGTAGTGGAACTTCATCCCGATATAAGATATATTTACAGAAGTCTGAAGGATTGCAATGATCATAGTTTACCTATTGAGTTTTTAAGGCTGTACAATGGACATATCAACATTGATTATGGTGAAAATTTGACCATTCCTGCTACAGATGCAACCAACAACATTCATTGGTCTTATTTACATATAAAGTTTGCTGAACCTATCAGTCTTTTTGTCTGTGATGCTGAATTGCCTGTAACAGTCAACTGGAGTAAAATTATAATAGAGTGGAGCAAGCATGTAAGAAAATGCAAGTACTGTTCCTCAGTTAATAAATGTACGTTAATAGTAAAATATCATGCTCAAGATGATATCGATTTCAAATTAGACAATATAACTATATTAAAAACTTATGTATGCTTAGGCAGTAAGTTAAAGGGGTCTGAAGTTTACTTAGTCCTTACAATAGGTCCTGCAAATGTGTTCCCAGTATTTAATGTAGTACAAAATGCTAAATTGATACTATCAAGAACCAAAAATTTCATCATGCCTAAGAAGGCTGATAAAGAGTCTATTGATGCAAATATTAAAAGTTTGATACCCTTTCTTTGTTACCCTATAACAAAAAAAGGAATTAATACTGCATTGTCAAAACTAAAGAGTGTTGTTAGTGGAGATATACTATCATATTCTATAGCAGGACGTAATGAAGTTTTCAGCAATAAACTTATAAATCATAAGCATATGAACATCTTAAAATGGTTCAATCATGTTTTAAATTTCAGATCAACAGAACTAAACTATAATCATTTATATATGGTAGAATCTACATATCCTTATCTAAGTGAATTGTTAAACAGCTTGACAACTAATGAACTTAAAAAACTGATTAAAATCACAGGTAGTTTGTTATACAACTTTCATAATGAATAATGAATAAAAATCTTATATTAAAAATTCCCATAGCTACACACTAACACTGTATTCAATTATAGTTATTTAAAATTAAAAATTATATAATTTTTTAATAACTTTTAGTGAACTAATCCTAAAATTATCATTTTGATCTAGGAGGAATAAATTTAAATCCAAATCTAATTGGTTTATATGTATATTAACTAAACTACGAGATATTAGTTTTTGACACTTTTTTTCTCGT'\n",
    "        ref_seq_list = list(ref_seq)\n",
    "        dict_jianji = {}\n",
    "        for k,v in dict_info_feikong.items():\n",
    "            gene_pos = int(k.split('_')[1])\n",
    "            ref = k.split('_')[0]\n",
    "            alt = k.split('_')[2]\n",
    "            product = v['product']\n",
    "        #     frequency = v['frequency']\n",
    "            gene_pos_f2 = gene_pos - 2   #前面2个\n",
    "            gene_pos_f1 = gene_pos - 1   #前面1个\n",
    "            gene_pos_b1 = gene_pos + 1\n",
    "            gene_pos_b2 = gene_pos + 2\n",
    "            start = v['start']\n",
    "            end = v['end']\n",
    "            type4 = v['type']\n",
    "            chazhi = int(gene_pos) - int(start)\n",
    "            quzheng =  chazhi//3   #取整，以获取是这个蛋白质的第几个氨基酸。\n",
    "            aa_pos = quzheng + 1\n",
    "\n",
    "        # #分类讨论\n",
    "            if type4 == 'del' or type4 == 'ins':\n",
    "                del_count = len(ref) - len(alt)\n",
    "                if del_count%3 == 0:\n",
    "                    #分类讨论\n",
    "                    #1\n",
    "                    if chazhi%3 == 2 and len(alt)%3 == 2:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "                   #2\n",
    "                    if chazhi%3 == 2 and len(alt)%3 == 1:\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "\n",
    "                    #3\n",
    "                    if chazhi%3 == 2 and len(alt)%3 == 0:\n",
    "                    #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        list_jianji_pos_ref.append(gene_pos_f1)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "\n",
    "                    #4\n",
    "                    if chazhi%3 == 1 and len(alt)%3 == 2:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "\n",
    "                    #5\n",
    "                    if chazhi%3 == 1 and len(alt)%3 == 1:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                #6\n",
    "                    if chazhi%3 == 1 and len(alt)%3 == 0:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        list_jianji_pos_ref.append(gene_pos_f2)\n",
    "\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        list_jianji_pos_alt.append(gene_pos_f2)\n",
    "\n",
    "\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "\n",
    "                    #7\n",
    "                    if chazhi%3 == 0 and len(alt)%3 == 2:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian]) \n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "\n",
    "                    #8\n",
    "                    if chazhi%3 == 0 and len(alt)%3 == 1:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "                        list_jianji_pos_ref.append(gene_pos_b1)\n",
    "                        list_jianji_pos_ref.append(gene_pos_b2)\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "                        list_jianji_pos_alt.append(gene_pos_b1)\n",
    "                        list_jianji_pos_alt.append(gene_pos_b2)\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                        list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "                #9\n",
    "                    if chazhi%3 == 0 and len(alt)%3 == 0:\n",
    "                        #变异前的位置和碱基\n",
    "                        list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_ref = []\n",
    "                        for index in range(0,len(ref)):\n",
    "                            list_jianji_pos_ref.append((gene_pos + index))\n",
    "\n",
    "                        for p in list_jianji_pos_ref:\n",
    "                            p_qiepian = p-1\n",
    "                            list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "                        #变异后的位置\n",
    "                        list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "                        list_jianji_alt = []\n",
    "                        for index in range(0,len(alt)):\n",
    "                            list_jianji_pos_alt.append((gene_pos + index))\n",
    "                            list_jianji_alt.append(alt[index])\n",
    "\n",
    "\n",
    "\n",
    "          #------------------------原来的，底下可以不变  \n",
    "            if type4 == 'snp':\n",
    "                if chazhi%3 == 2:\n",
    "                    list_jianji_pos = []#是位置不是切片，切片要-1\n",
    "                    list_jianji_pos.append(gene_pos_f2)\n",
    "                    list_jianji_pos.append(gene_pos_f1)\n",
    "                    list_jianji_pos.append(gene_pos)\n",
    "\n",
    "\n",
    "\n",
    "                if chazhi%3 == 1:\n",
    "                    list_jianji_pos = []\n",
    "                    list_jianji_pos.append(gene_pos_f1)\n",
    "                    list_jianji_pos.append(gene_pos)\n",
    "                    list_jianji_pos.append(gene_pos_b1)\n",
    "\n",
    "\n",
    "\n",
    "                if chazhi%3 == 0:\n",
    "                    list_jianji_pos = []\n",
    "                    list_jianji_pos.append(gene_pos)\n",
    "                    list_jianji_pos.append(gene_pos_b1)\n",
    "                    list_jianji_pos.append(gene_pos_b2)\n",
    "\n",
    "\n",
    "                list_jianji_qiepian = []#这个是切片\n",
    "                for i in list_jianji_pos:\n",
    "                    i_1 = i-1\n",
    "                    list_jianji_qiepian.append(i_1)\n",
    "            #     print(list_jianji_qiepian)\n",
    "\n",
    "                list_jianji_ref = []#获取参考碱基\n",
    "                for i in list_jianji_qiepian:\n",
    "                    list_jianji_ref.append(ref_seq_list[i])\n",
    "            #     print(list_jianji_ref)\n",
    "\n",
    "                #获取alt碱基\n",
    "                list_jianji_alt= []#获取参考碱基\n",
    "                for i in list_jianji_qiepian:\n",
    "                    if i == int(gene_pos - 1):\n",
    "                        list_jianji_alt.append(alt)\n",
    "                    else:\n",
    "                        list_jianji_alt.append(ref_seq_list[i])\n",
    "            #     print(list_jianji_alt)\n",
    "            #翻译蛋白质\n",
    "            str_ref = \"\".join(list_jianji_ref)\n",
    "            str_alt = \"\".join(list_jianji_alt)\n",
    "            aa_ref = str(Seq(str_ref).translate())\n",
    "            aa_alt = str(Seq(str_alt).translate())    \n",
    "            #非同义：\n",
    "            if aa_alt != aa_ref:\n",
    "                info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "                dict_aa_mut[k] = {}\n",
    "                dict_aa_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "                dict_aa_mut[k]['aa'] = info\n",
    "                dict_aa_mut[k]['num'] = dict_mut_num[k]\n",
    "            #同义\n",
    "            if aa_alt == aa_ref:\n",
    "                info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "                dict_aa_synonymous_mut[k] = {}\n",
    "                dict_aa_synonymous_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "                dict_aa_synonymous_mut[k]['aa'] = info\n",
    "                dict_aa_synonymous_mut[k]['num'] = dict_mut_num[k]\n",
    "        #             dict_aa_mut[k]['frequency'] = dict_info_feikong[k]['frequency']\n",
    "\n",
    "        # \n",
    "\n",
    "        #非同义----------------------------------------------------------------------\n",
    "        #统计各个突变的频率。\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_nonsynonymous_mut_frequency.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "            for k,v in dict_aa_mut.items():\n",
    "                writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "\n",
    "\n",
    "        #统计非同义突变在哪个蛋白上nonsynonymous mutation\n",
    "        list_product = []\n",
    "        for k,v in dict_aa_mut.items():\n",
    "                if v['product'] not in list_product:\n",
    "                    list_product.append(v['product'])\n",
    "                    dict_product = dict.fromkeys(list_product)#每个蛋白上非同义突变的个数\n",
    "        for k in dict_product.keys():\n",
    "            dict_product[k] = 0\n",
    "        for m,p in dict_aa_mut.items():\n",
    "            for k,v in dict_product.items():\n",
    "                if p['product'] == k:\n",
    "                    dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "        # print(dict_product)\n",
    "        dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "        # print(dict_product)\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_nonsynonymous_product_number.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['product','number','group'])\n",
    "            for k,v in dict_product.items():\n",
    "                writer.writerow([k,v,'nonsynonymous'])\n",
    "\n",
    "\n",
    "        #同义----------------------------------------------------------------------\n",
    "        #统计各个突变的频率。\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_synonymous_mut_frequency.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "            for k,v in dict_aa_synonymous_mut.items():\n",
    "                writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "\n",
    "\n",
    "\n",
    "        #统计同义突变在哪个蛋白上\n",
    "        list_product = []\n",
    "        for k,v in dict_aa_synonymous_mut.items():\n",
    "                if v['product'] not in list_product:\n",
    "                    list_product.append(v['product'])\n",
    "                    dict_product = dict.fromkeys(list_product)\n",
    "        for k in dict_product.keys():\n",
    "            dict_product[k] = 0\n",
    "        for m,p in dict_aa_synonymous_mut.items():\n",
    "            for k,v in dict_product.items():\n",
    "                if p['product'] == k:\n",
    "                    dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "        # print(dict_product)\n",
    "        dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "        # print(dict_product)\n",
    "        with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/08product_frequency/{c1}/08isnv_synonymous_product_number.csv','w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            writer.writerow(['product','number','group'])\n",
    "            for k,v in dict_product.items():\n",
    "                writer.writerow([k,v,'synonymous'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef4ea9bb-2b7e-4357-82b2-18fab9ea172e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/06snp_mutation_PRJNA438443.csv')\n",
    "df_counts = df.isnv.value_counts()\n",
    "print(df_counts)\n",
    "# outputpath=f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/06product_exist/mut_sample_number/{c1}_mut_sample_number.csv'\n",
    "# df_counts.to_csv(outputpath,sep=',',index=True,header=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fb0ed4-0f3f-457c-8e36-37a4f09a381b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #加入每个突变的样本数量.不分批次的\n",
    "# # 05 翻译蛋白质，要加3-80%的。isnv的\n",
    "# #非同义氨基酸突变的具体情况和不同蛋白非同义突变的数量。\n",
    "# #插入没加先。\n",
    "# import csv\n",
    "# import pandas as pd\n",
    "# from Bio.Seq import Seq\n",
    "# dict_info = {}\n",
    "# info_list = []\n",
    "# dict_pep_cds = {'99_518':'NS1_1C','628_1002':\"NS2_1B\",'1140_2315':\"N\",'2348_3073':\"P\",'3233_4003':\"M\",'4274_4468':\"SH\",'4659_5555':\"G\",'5632_7356':\"F\",'7576_8160':\"M2\",'8468_14965':\"L\"}\n",
    "# with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/06isnv_mutation.csv') as csvfile:\n",
    "#     csv_reader = csv.reader(csvfile)  \n",
    "#     for row in csv_reader:\n",
    "#         if row[0] != 'isnv':\n",
    "#             mut = row[0]\n",
    "#             if mut not in info_list:\n",
    "#                 info_list.append(mut)\n",
    "# dict_info = dict.fromkeys(info_list)\n",
    "\n",
    "\n",
    "# for i in dict_info.keys():#嵌套字典可以这样赋值：1、dict_info = dict.fromkeys(info_list)2、每个键都赋值成空字典，再正常加就好了\n",
    "#     dict_info[i] = {}\n",
    "# #新加的，统计每个突变的样本数量。\n",
    "# df = pd.read_csv('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/06isnv_mutation.csv')\n",
    "# df_counts = df.isnv.value_counts()\n",
    "# # print(df_counts)\n",
    "# outputpath=f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/mut_sample_number.csv'\n",
    "# df_counts.to_csv(outputpath,sep=',',index=True,header=False)\n",
    "\n",
    "# dict_mut_num = {}\n",
    "\n",
    "# mut_number_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/mut_sample_number.csv'\n",
    "# with open(mut_number_path) as csvfile:\n",
    "#     csv_reader = csv.reader(csvfile) \n",
    "#     for row in csv_reader:\n",
    "#         isnv = row[0]\n",
    "#         num = row[1]\n",
    "#         dict_mut_num[isnv] = num    \n",
    "\n",
    "    \n",
    "# with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/06product_exist/06isnv_mutation.csv') as csvfile:\n",
    "#     csv_reader = csv.reader(csvfile)  \n",
    "#     for row in csv_reader:\n",
    "#         if row[0] != 'isnv':\n",
    "#             mut = row[0]\n",
    "#             pos = int(row[0].split('_')[1])\n",
    "#             fre = float(row[2])\n",
    "#             type4 = row[3]\n",
    "#             first = row[0].split('_')[0]\n",
    "#             last = row[0].split('_')[2]\n",
    "#             #分一下位置，del和ins的有可能结束位置和开始位置不在同一段蛋白里面。不太可能\n",
    "#             if type4 == 'del':#按照突变前的长度算\n",
    "#                 pos_end = pos + len(first) -1\n",
    "#             if type4 == 'ins':#按照突变后的长度算\n",
    "#                 pos_end = pos + len(last) -1          \n",
    "#             for k,v in dict_pep_cds.items():\n",
    "#                 start = int(k.split('_')[0])\n",
    "#                 end = int(k.split('_')[1])\n",
    "#                 if type4 == 'snp':\n",
    "#                     if start <= pos <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "#                         dict_info[mut]['product'] = v\n",
    "#                         dict_info[mut]['type'] = type4\n",
    "#                         dict_info[mut]['start'] = start\n",
    "#                         dict_info[mut]['end'] = end\n",
    "#                         dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "# #                         dict_info[mut]['frequency'] = fre 频率先不考虑了，有好多样本的\n",
    "#                 if type4 == 'del' or type4 == 'ins':\n",
    "#                     if start <= pos <= end and start <= pos_end <= end: #这里一定要转化成int，不然就是字符串是不对的。    \n",
    "#                         dict_info[mut]['product'] = v\n",
    "#                         dict_info[mut]['type'] = type4\n",
    "#                         dict_info[mut]['start'] = start\n",
    "#                         dict_info[mut]['end'] = end\n",
    "#                         dict_info[mut]['sample_number'] = dict_mut_num[mut]\n",
    "# #                         dict_info[mut]['frequency'] = fre\n",
    "\n",
    "                                       \n",
    "# # print(dict_info)\n",
    "# #去除不在cds范围内的碱基突变                   \n",
    "# dict_info_feikong = {}            \n",
    "# for i in dict_info.keys():\n",
    "#     if dict_info[i]:\n",
    "#         dict_info_feikong[i] = dict_info[i]\n",
    "# # print(dict_info_feikong)\n",
    "\n",
    "# #获取突变前后的三个碱基，蛋白位置不知道，可以用什么蛋白\n",
    "# dict_aa_mut = {}#非同义突变的字典\n",
    "# dict_aa_synonymous_mut = {}\n",
    "# ref_seq = 'ACGCGAAAAAATGCGTACAACAAACTTGCGTAAACCAAAAAAATGGGGCAAATAAGAATTTGATAAGTACCACTTAAATTTAACTCCCTTGGTTAGAGATGGGCAGCAATTCATTGAGTATGATAAAAGTTAGATTACAAAATTTGTTTGACAATGATGAAGTAGCATTGTTAAAAATAACATGCTATACTGACAAATTAATACATTTAACTAATGCATTGGCTAAGGCAGTGATACATACAATCAAATTGAATGGCATTGTATTTGTGCATGTTATTACAAGTAGTGATATTTGCCCTAATAATAATATTGTAGTGAAATCCAATTTCACAACAATGCCAGTGTTACAAAATGGAGGTTATATATGGGAAATGATGGAATTAACACACTGCTCTCAACCTAATGGCCTAATAGATGACAATTGTGAAATTAAATTCTCCAAAAAACTAAGTGATTCAACAATGACCAATTATATGAATCAATTATCTGAATTACTTGGATTTGATCTTAATCCATAAATTATAATAAATATCAACTAGCAAATCAATGTCACTAACACCATTAGTTAATATAAAACTTGACAGAAGATAAAAATGGGGCAAATAAATCAATTCAGCCGACCCAACCATGGACACAACACACAATGACACCACACCACAAAGACTGATGATCACAGACATGAGACCATTGTCACTTGAGACTATAATAATATCACTAACCAGAGACATCATAACACACAGATTTATATACTTGATAAATCATGAATGTATAGTGAGAAAACTTGATGAAAGACAGGCCACATTTACATTCCTGGTCAACTATGAAATGAAACTATTGCACAAAGTGGGAAGCACTAAATACAAAAAATATACTGAATACAACACAAAATATGGCACTTTTCCTATGCCAATATTTATCAATCATGATGGGTTCTTAGAATGCATTGGCATTAAGCCTACAAAGCACACTCCCATAATATACAAGTATGATCTCAATCCATGAATTTCAACACAAGAGTCACACAATCTGAAATAACAACTTCATGCATAACCACACTCCATAGTTCAAATGGAGCCTGAAAATTATAGTAATTTAAAATTAAGGAGAGACATAAGATGAAAGATGGGGCAAATACAAAAATGGCTCTTAGCAAAGTCAAGTTGAATGATACACTCAACAAAGATCAACTTCTGTCATCCAGCAAATACACCATCCAACGGAGCACAGGAGATAGTATTGATACTCCTAATTATGATGTGCAGAAACACATCAATAAGTTATGTGGCATGTTATTAATCACAGAAGATGCTAATCATAAATTCACTGGGTTAATAGGTATGTTATATGCTATGTCTAGATTAGGAAGAGAAGACACCATAAAAATACTCAGAGATGCGGGATATCATGTAAAAGCAAATGGAGTGGATGTAACAACACATCGTCAAGATATTAATGGGAAAGAAATGAAATTTGAAGTGTTAACATTGTCAAGCTTAACAACTGAAATTCAAATCAACATTGAGATAGAATCTAGAAAATCCTACAAAAAAATGCTAAAAGAAATGGGAGAGGTAGCTCCAGAATACAGGCATGACTCTCCTGATTGTGGGATGATAATATTATGTATAGCGGCATTAGTAATAACCAAATTAGCAGCAGGGGATAGATCTGGTCTTACAGCTGTGATTAGGAGGGCTAATAATGTCCTAAAAAATGAAATGAAACGTTATAAAGGCTTACTACCCAAGGATATAGCCAACAGCTTCTATGAAGTGTTTGAAAAATATCCTCACTTTATAGATGTTTTTGTTCATTTTGGTATAGCACAATCTTCTACCAGAGGTGGCAGTAGAGTTGAAGGGATTTTTGCTGGATTGTTTATGAATGCCTATGGTGCAGGGCAAGTGATGTTACGGTGGGGGGTCTTAGCAAAATCAGTTAAAAATATTATGCTAGGACACGCTAGTGTGCAAGCAGAAATGGAACAAGTTGTGGAGGTTTATGAATATGCCCAAAAATTGGGTGGAGAAGCAGGGTTCTACCATATATTGAACAACCCAAAAGCATCATTATTGTCTTTGACTCAATTTCCTCACTTCTCCAGTGTAGTATTAGGCAATGCTGCTGGCCTAGGCATAATGGGAGAATACAGAGGTACACCAAGGAATCAAGATCTATATGATGCTGCAAAAGCATATGCTGAACAACTCAAAGAAAATGGTGTGATTAACTACAGTGTATTAGACTTGACAGCAGAAGAACTAGAGGCTATCAAACATCAGCTTAATCCAAAAGATAATGATGTAGAGCTTTGAGTTAATAAAAAAAATGGGGCAAATAAAACATCATGGAAAAGTTTGCTCCTGAATTCCATGGAGAAGATGCAAACAACAGAGCTACCAAATTCCTAGAATCAATAAAGGGCAAATTCACATCACCTAAAGATCCCAAGAAAAAAGATAGTATCATATCTGTCAACTCAATAGATATAGAAGTAACCAAAGAAAGCCCTATAACATCAAATTCAACCATTATAAACCCAACAAATGAGACAGATGATACTGTAGGGAACAAGCCCAATTATCAAAGAAAACCTCTAGTAAGTTTCAAAGAAGACCCTACGCCAAGTGATAATCCCTTTTCAAAACTATACAAAGAAACCATAGAAACATTTGATAACAATGAAGAAGAATCTAGCTATTCATATGAAGAAATAAATGATCAGACAAACGATAATATAACAGCAAGATTAGATAGGATTGATGAAAAATTAAGTGAAATACTAGGAATGCTTCACACATTAGTAGTAGCGAGTGCAGGACCTACATCTGCTCGGGATGGTATAAGAGATGCCATGGTTGGTTTAAGAGAAGACATGATAGAAAAAATCAGAACTGAAGCATTAATGACCAATGACAGACTAGAAGCTATGGCAAGACTCAGGAATGAGGAAAGTGAAAAGATGGCAAAAGACACATCAGATGAAGTGTCTCTCAATCCAACATCAGAGAAATTGAACAACCTGTTGGAAGGGAATGATAGTGACAATGATCTATCACTTGATGATTTCTGATCAGTTACCAATCTGTACATCAACACACAACACCAACAGAAGACCAACAAACAAACCAACTCACCCATCCAACCAAACATCTATACGCCAATCAGCCAATCCAAAACTAGCCACCCGGAAAAAATAGATACTATAGTTACAAAAAAAGATGGGGCAAATATGGAAACATACGTGAACAAACTTCACGAAGGCTCCACATACACAGCTGCTGTTCAATACAATGTCTTAGAAAAAGACGATGACCCTGCATCACTTACAATATGGGTGCCCATGTTCCAATCATCCATGCCAGCAGATTTACTTATAAAAGAACTAGCTAATGTCAACATACTAGTGAAACAAATATCCACACCCAATGGACCTTCATTAAGAGTCATGATAAACTCAAGAAGTGCAGTGCTAGCACAAATGCCCAGCAAATTTACCATATGTGCCAATGTGTCCTTGGATGAAAGAAGCAAGCTGGCATATGATGTAACCACACCCTGTGAAATCAAGGCATGTAGTCTAACATGCCTAAAATCAAAAAATATGTTAACTACAGTTAAAGATCTCACTATGAAAACACTCAACCCAACACATGACATCATTGCTTTATGTGAATTTGAAAATATAGTAACATCAAAAAAAGTCATAATACCAACATACCTAAGATCCATCAGTGTCAGAAATAAAGATCTGAACACACTTGAAAATATAACAACCACTGAATTCAAAAATGCCATCACAAATGCAAAAATCATCCCTTACTCAGGATTACTGTTAGTCATCACAGTGACTGACAACAAAGGAGCATTCAAATACATAAAGCCACAAAGTCAATTTATAGTAGATCTTGGAGCTTACCTAGAAAAAGAAAGTATATATTATGTTACAACAAATTGGAAGCACACAGCTACACGATTTGCAATCAAACCCATGGAAGATTAACCTTTTTCTTCTACATCAGTGAGTTGATTCATACAAACTTTCTACCTACATTCTTCACTTCACCATCATAATCACCAACCCTCTGTGGTTCAACTAATCAAACAAAACCCATCTGGAGCCTCAGATCATCCCAAGTCATTGTTCATCAGATCTAGTACTCAAATAAGTTAATAAAAATATCCACATGGGGCAAATAATCATTGGAGGAAATCCAACTAATCACAATATCTGTCAACATAGACAAGTCAACACGCCAGGCAAAATCAACCAATGGAAAATACATCCATAACAATAGAATTCTCAAGCAAATTCTGGCCTTACTTTACACTAATACACATGATAACAACAATAATCTCTTTGCTAATCATAATCTCCATCATGATTGCAATACTGAACAAACTCTGTGAATATAACGTATTCCATAACAAAACCTTTGAGCTACCAAGAGCTCGAGTCAATACATAGCATTCACCAATCTGATGGCACAAAACAGTAACCTTGCATTTGTAAGTGAACAACCCTCACCTCTTTACAAAACCACATCAACATCTCACCATGCAAGCCATCATCCATATTATAAAGTAGTTAATTAAAAATAATCATAACAATGAACTAAGATATTAAGACTAACAATAACGTTGGGGCAAATGCAAACATGTCCAAAAACAAGGACCAACGCACCACCAAGACACTAGAAAAGACCTGGGACACTCTCAATCATCTATTATTCATATCATCGTGCTTATACAAGTTAAATCTTAAATCTATAGCACAAATCACATTATCCATTCTGGCAATGATAATCTCAACTTCACTTATAATTGCAGCCATCATATTCATAGCCTCGGCAAACCACAAAGTCACACTAACAACTGCAATCATACAAGATGCAACAAGCCAGATCAAGAACACAACCCCAACATACCTCACCCAGAATCCCCAGCTTGGAATCAGCTTCTCCAATCTGTCTGAAACTACATCACAAACCACCACCATACTAGCTTCAACAACACCAAGTGTCAAGTCAACCCTGCAATCCACAACAGTCAAGACCAAAAACACAACAACAACCAAAATACAACCCAGCAAGCCCACCACAAAACAACGCCAAAACAAACCACCAAACAAACCCAATAATGATTTTCACTTTGAAGTGTTCAACTTTGTACCTTGCAGCATATGCAGCAACAATCCAACCTGCTGGGCTATCTGTAAAAGAATACCAAACAAAAAACCTGGAAAGAAAACCACCACCAAGCCCACAAAAAAACCAACCATCAAGACAACCAAAAAAGATCTCAAACCTCAAACCACAAAACCAAAGGAAGTACCTACCACCAAGCCCACAGAAAAGCCAACCATCAACACCACCAAAACAAACATCAGAACTACACTGCTCACCAACAATACCACAGGAAATCCAGAACACACAAGTCAAAAGGGAACCCTCCACTCAACCTCCTCCGATGGCAATCCAAGCCCTTCACAAGTCTATACAACATCCGAGTACCTATCACAACCTCCATCTCCATCCAACACAACAAACCAGTAGTCATTAAAAAGCGTATTATTGCAAAAAGCCATGACCAAATCAACCAGAATCAAAATCAACTCTGGGGCAAATAACAATGGAGTTGCCAATCCTCAAAACAAATGCAATTACCGCAATCCTTGCTGCAGTCACACTCTGTTTTGCTTCCAGTCAAAACATCACTGAAGAATTTTATCAAACAACATGCAGTGCAGTCAGCAAAGGCTATCTTAGTGCTCTAAGAACTGGTTGGTATACTAGTGTTATAACTATAGAATTAAGTAATATCAAGGAAAATAAGTGTAATGGAACAGACGCTAAGGTAAAATTGATAAAACAAGAATTAGATAAATATAAAAGTGCTGTAACAGAATTGCAGTTGCTCATGCAAAGCACACCGGCAACCAACAATCGAGCCAGAAGAGAACTACCAAGGTTTATGAATTATACACTCAACAATACCAAAAATACCAATGTAACATTAAGCAAGAAAAGGAAAAGAAGATTTCTTGGCTTTTTGTTAGGTGTTGGATCTGCAATCGCCAGTGGCATTGCTGTATCTAAGGTCCTGCACCTAGAAGGGGAAGTGAACAAAATCAAAAGTGCTCTACTATCCACAAACAAGGCTGTAGTCAGCTTATCAAATGGAGTTAGTGTCTTAACCAGCAAAGTGTTAGACCTCAAAAACTATATAGATAAACAGTTGTTACCTATTGTGAACAAGCAAAGCTGTAGCATATCAAACATTGAAACTGTGATAGAGTTCCAACAAAAGAACAACAGACTACTAGAGATTACCAGGGAATTTAGTGTTAATGCAGGTGTAACTACACCTGTAAGCACTTATATGTTAACAAATAGTGAATTATTATCATTAATCAATGATATGCCTATAACAAATGATCAGAAAAAGTTAATGTCCAACAATGTTCAAATAGTTAGACAGCAAAGTTACTCTATCATGTCCATAATAAAGGAGGAAGTCTTAGCATATGTAGTACAATTACCACTATATGGTGTAATAGATACACCTTGTTGGAAACTGCACACATCCCCTCTATGTACAACCAACACAAAGGAAGGGTCCAACATCTGTTTAACAAGAACCGACAGAGGATGGTACTGTGACAATGCAGGATCAGTATCTTTCTTCCCACTAGCTGAAACATGTAAAGTTCAATCGAATCGAGTATTTTGTGACACAATGAACAGTTTAACATTACCAAGTGAAGTAAATCTCTGCAACATTGACATATTCAACCCCAAATATGATTGCAAAATTATGACTTCAAAAACAGATGTAAGCAGCTCCGTTATCACATCTCTAGGAGCCATTGTGTCATGCTATGGCAAAACTAAATGTACAGCATCCAATAAAAATCGTGGAATCATAAAGACATTTTCTAACGGGTGCGATTATGTATCAAATAAGGGGGTTGACACTGTGTCTGTAGGTAATACATTATATTATGTAAATAAGCAAGAAGGCAAAAGTCTCTATGTAAAAGGTGAACCAATAATAAATTTCTATGACCCATTAGTGTTCCCCTCTGATGAATTTGATGCATCAATATCTCAAGTCAATGAGAAGATTAACCAGAGCCTAGCATTTATTCGTAAATCCGATGAATTATTACATAATGTAAATGCTGGTAAATCCACCATAAATATCATGATAACTACTATAATTATAGTGATTATAGTAATATTGTTATCATTAATTGCCGTTGGACTGCTCCTATACTGCAAGGCCAGAAGCACACCAGTCACACTAAGCAAGGATCAACTGAGTGGTATAAATAATATTGCATTTAGTAACTAAATAAAAATAGCACCTAATCATGTTCTTACAATGGTTTCATATCTGCTCATAGACAACCCATCTATCATTGGATTTTCTTAAAATCTGAACTTCATCGAAACTCTCATCTATAAACCATCTCACTTACATTATTTAAGTAGATTCCTAGTTTATAGTTATATAAAACAATTGAATACCAGATTAACTTACTATTTGTAAAAAATGAGAACTGGGGCAAATATGTCACGAAGGAATCCTTGCAAATTTGAAATTCGAGGTCATTGCTTGAATGGTAAGAGGTGTCATTTTAGTCATAATTATTTTGAATGGCCACCCCATGCACTGCTTGTAAGACAAAACTTTATGTTAAACAGAATACTTAAGTCTATGGATAAAAGCATCGATACTTTATCAGAAATAAGTGGAGCTGCAGAGTTGGACAGAACAGAAGAGTATGCCCTCGGTGTAGTTGGAGTGCTAGAGAGTTATATAGGATCTATAAATAATATAACTAAACAATCAGCATGTGTTGCCATGAGCAAACTCCTCACTGAACTCAACAGTGATGACATCAAAAAACTGAGGGACAATGAAGAGCCAAATTCACCCAAGATAAGAGTGTACAATACTGTCATATCATATATTGAAAGCAACAGGAAAAACAATAAACAAACTATCCATCTGTTAAAAAGATTGCCAGCAGACGTATTGAAGAAAACCATAAAAACCACATTGGATATCCACAAGAGCATAACCATCAATAACCCAAAAGAATCAACTGTTAGTGATATAAACGACCATGCCAAAAATAATGATACTACCTGACAAATATCCTTGTAGTATAAATTCCATACTAATAACAAGTAGTTGTAGAGTTACTATGTATAATCAAAAGAACACACTATATTTCAATCAAAACAACCAAAATAACCATATATACTCACCGAATCAACCATTCAATGAAATCCATTGGACCTCTCAAGACTTGATTGATGCAATTCAAAATTTTCTACAACATCTAGGTATTACTGATGATATATACACAATATATATATTAGTGTCATAACACTCAATCCTAATGCTTACCACATCATCAAACTATTAACTCAAACAATTCAAGCCATGGGACAAAATGGATCCCATTATTAATGGAAATTCTGCTAATGTGTATCTAACCGATAGTTATTTAAAAGGTGTTATTTCTTTCTCAGAATGTAATGCTTTAGGAAGTTACATATTCAATGGTCCTTATCTCAAAAATGATTATACCAACTTAATTAGTAGACAAAATCCATTAATAGAACACATAAATCTAAAGAAACTAAATATAACACAGTCCTTAATGTCTAAGTATCATAAAGGTGAAATAAAAATAGAAGAACCTACTTATTTTCAGTCATTACTTATGACATACAAGAGTATGACCTCGTTAGAACAGATTACTACCACTAATTTACTTAAAAAGATAATAAGAAGAGCTATAGAAATTAGTGATGTCAAAGTCTATGCTATATTGAATAAACTGGGGCTTAAAGAAAAAGACAAGATTAAATCCAACAATGGACAAGATGAAGACAACTCAGTTATTACAACCATAATCAAAGATGATATACTTTTAGCTGTTAAGGATAATCAATCTCATCTTAAAGCAGTCAAAAATCACTCTACAAAACAAAAAGATACAATCAAAACAACACTCTTGAAGAAATTAATGTGTTCAATGCAACATCCTCCATCATGGTTAATACATTGGTTTAATTTATACACAAAATTAAACAACATATTAACACAGTATCGATCAAGTGAGGTAAAAAACCATGGTTTTATATTGATAGACAATCATACTCTCAATGGATTCCAATTTATTTTGAATCAATATGGTTGTATAGTTTATCATAAGGAACTCAAAAGAATTACTGTGACAACCTATAATCAATTCTTGACATGGAAAAATATTAGCCTTAGTAGATTAAATGTTTGTTTAATTACATGGATTAGTAACTGTTTGAACACATTAAATAAAAGCTTAGGCTTAAGATGCGGATTCAATAATGTTATCTTGACACAACTATTCCTCTATGGAGATTGTATACTAAAACTATTCCACAATGAGGGGTTCTACATAATAAAAGAGGTAGAGGGATTTATTATGTCTCTAATTTTAAATATAACAGAAGAAGATCAATTCAGAAAACGGTTTTATAATAGTATGCTCAACAACATCACAGATGCTGCTAATAAAGCTCAGAAAAATCTGCTATCAAGAGTATGTCATACATTATTAGATAAGACAGTATCCGATAATATAATAAATGGCAGATGGATAATTCTATTAAGTAAGTTCCTTAAATTAATTAAGCTTGCAGGTGACAATAACCTTAACAATCTGAGTGAATTATATTTTTTGTTCAGAATATTTGGACACCCAATGGTAGATGAAAGACAAGCCATGGATGCTGTTAAAGTTAATTGCAACGAGACCAAATTTTACTTGTTAAGCAGTTTGAGTATGTTAAGAGGTGCCTTTATATATAGAATTATAAAAGGATTTGTAAATAATTACAACAGATGGCCTACTTTAAGGAATGCTATTGTTTTACCCTTAAGATGGTTAACTTACTATAAACTAAACACTTATCCTTCCTTGTTGGAACTTACAGAAAGAGATTTGATTGTTTTATCAGGACTACGTTTCTATCGTGAGTTTCGGTTGCCTAAAAAAGTGGATCTTGAAATGATCATAAATGATAAGGCTATATCACCTCCTAAAAATTTGATATGGACTAGTTTCCCTAGAAATTATATGCCGTCACACATACAAAATTATATAGAACATGAAAAATTAAAATTTTCCGAGAGTGATAAATCAAGAAGAGTATTAGAGTACTATTTAAGAGATAACAAATTCAATGAATGTGATTTATACAACTGTGTAGTTAATCAAAGTTATCTTAACAACCCTAATCATGTGGTATCTTTGACAGGCAAAGAAAGAGAACTCAGTGTAGGTAGAATGTTTGCAATGCAACCAGGAATGTTCAGACAAGTTCAAATATTAGCAGAGAAAATGATAGCTGAAAACATTTTACAATTCTTTCCTGAAAGTCTTACAAGATATGGTGATCTAGAACTACAGAAAATATTAGAATTGAAAGCAGGAATAAGTAACAAATCAAATCGTTACAATGATAATTACAACAATTACATTAGTAAGTGCTCTATCATCACAGATCTCAGCAAATTCAATCAAGCATTTCGATATGAAACATCATGTATTTGTAGTGATGTACTGGATGAACTGCATGGTGTACAATCTCTATTTTTCTGGTTACATTTAGCTATTCCTCATGTCACAATAATATGCACATATAGGCATGCACCCCCCTATATAAGAGATCATATTGTAGATCTTAACAATGTAGATGAACAAAGTGGATTATATAGATATCATATGGGTGGTATCGAAGGGTGGTGTCAAAAACTATGGACCATAGAAGCTATATCACTATTGGATCTAATATCTCTCAAAGGGAAATTCTCAATTACTGCTTTAATTAATGGTGACAATCAATCAATAGATATAAGTAAACCAGTCAGACTCATGGAAGGTCAAACTCATGCTCAAGCAGATTATTTGCTAGCATTAAATAGTCTTAAATTACTGTATAAAGAGTATGCAGGCATAGGCCACAAATTAAAAGGAACTGAGACTTATATATCAAGAGATATGCAATTTATGAGTAAAACAATTCAACATAACGGTGTATATTACCCAGCTAGTATAAAGAAAGTCCTAAGAGTGGGACCGTGGATAAACACTATACTTGATGATTTCAAAGTGAGTCTAGAATCTATAGGTAGTTTGACACAAGAATTAGAATATAGAGGAGAAAGTCTATTATGCAGTTTAATATTTAGAAATGTATGGTTATATAATCAAATTGCTTTACAACTAAAAAATCATGCATTATGTAACAATAAATTATATTTGGACATATTAAAGGTTCTGAAACACTTAAAAACCTTTTTTAATCTTGATAATATTGATACAGCATTAACATTGTATATGAATTTGCCCATGTTATTTGGTGGTGGTGATCCCAACTTGTTATATCGAAGTTTCTATAGAAGAACTCCTGATTTCCTCACAGAGGCTATAGTTCACTCTGTGTTCATACTTAGTTATTATACAAACCATGATTTAAAAGATAAACTTCAAGATCTGTCAGATGATAGATTGAATAAGTTCTTAACATGCATAATCACATTTGACAAAAACCCTAATGCTGAATTCGTAACATTGATGAGAGATCCTCAAGCTTTAGGGTCTGAGAGACAAGCTAAAATTACTAGCGAAATCAATAGACTGGCAGTTACTGAGGTTTTGAGCACAGCTCCAAACAAAATATTCTCCAAAAGTGCACAACACTATACCACTACAGAGATAGATCTAAATGATATTATGCAAAATATAGAACCTACATATCCTCACGGGCTAAGAGTTGTTTATGAAAGTTTACCCTTTTATAAAGCAGAGAAAATAGTAAATCTTATATCCGGTACAAAATCTATAACTAACATACTGGAAAAGACTTCTGCCATAGACTTAACAGATATTGATAGAGCCACTGAGATGATGAGGAAAAACATAACTTTGCTTATAAGGATATTTCCATTAGATTGTAACAGAGATAAAAGGGAAATATTGAGTATGGAAAACCTAAGTATTACTGAATTAAGCAAATATGTTAGGGAAAGATCTTGGTCTTTATCCAATATAGTTGGTGTTACATCACCTAGTATCATGTATACAATGGACATCAAATATACAACAAGCACTATAGCTAGTGGCATAATCATAGAGAAATATAATGTTAACAGTTTAACACGTGGTGAGAGAGGACCCACTAAACCATGGGTTGGTTCATCTACACAAGAGAAAAAAACAATGCCAGTTTATAATAGACAAGTTTTAACCAAAAAACAGAGAGATCAAATAGATCTATTAGCAAAATTGGATTGGGTGTATGCATCTATAGATAACAAGGATGAATTCATGGAAGAACTTAGCATAGGAATTCTTGGGTTAACATATGAGAAAGCCAAAAAATTATTTCCACAATATTTAAGTGTTAACTATTTGCATCGCCTTACAGTCAGTAGTAGACCATGTGAATTCCCTGCATCAATACCAGCTTATAGAACTACAAATTATCACTTTGATACTAGCCCTATTAATCGCATATTAACAGAAAAGTATGGTGATGAAGATATTGATATAGTATTCCAAAACTGTATAAGCTTTGGCCTTAGCTTAATGTCAGTAGTAGAACAATTTACTAATGTATGTCCTAACAGAATTATTCTTATACCTAAGCTTAATGAGATACATTTAATGAAACCTCCCATATTCACAGGTGATGTTGATATTCACAAGTTAAAACAAGTGATACAAAAACAGCATATGTTTTTACCAGACAAAATAAGTTTGACTCAATATGTGGAATTATTCTTAAGTAATAAAACACTCAAATCTGGATCTCATGTTAATTCTAATTTAATATTGGCGCATAAGATATCTGACTATTTTCATAATACTTACATTTTAAGTACTAATTTAGCTGGACATTGGATTCTGATTATACAACTTATGAAAGATTCTAAGGGTATTTTTGAAAAAGATTGGGGAGAGGGATATATAACTGATCATATGTTCATTAATTTGAAAGTTTTCTTCAATGCTTATAAGACCTATCTCTTGTGTTTTCATAAAGGTTACGGCAGAGCAAAGCTGGAGTGTGATATGAATACTTCAGATCTCCTATGTGTATTGGAATTAATAGACAGTAGTTATTGGAAGTCTATGTCTAAGGTATTTTTAGAACAAAAAGTTATCAAATACATTCTTAGCCAGGATGCAAGTTTACATAGAGTAAAAGGATGTCATAGCTTCAAACTATGGTTTCTTAAACGTCTTAATGTAGCAGAATTCACAGTTTGCCCTTGGGTTGTTAACATAGATTATCATCCAACACATATGAAAGCAATATTAACTTATATAGATCTTGTTAGAATGGGATTGATAAATATAGATAAAATATACATTAAAAATAAACACAAATTCAATGATGAATTTTATACTTCTAATCTCTTTTACATTAATTATAACTTCTCAGATAATACTCATCTATTAACTAAACATATAAGGATTGCTAATTCTGAATTAGAAAATAATTACAACAAATTATATCATCCTACACCAGAAACCCTAGAAAATATACTAACCAATCCGGTTAAATGTAATGACAAAAAGACACTGAATGACTATTGTATAGGTAAAAATGTTGACTCAATAATGTTACCATTGTTATCTAATAAGAAGCTTATTAAATCGCCTACAATGATTAGAACCAATTACAGCAAACAAGATTTGTATAATTTATTTCCTACGGTTGTGATTGATAAAATTATAGATCATTCAGGTAATACAGCCAAATCTAACCAACTTTACACTACTACTTCTCATCAAATACCTTTAGTGCACAATAGCACATCACTTTATTGCATGCTTCCTTGGCATCATATTAATAGATTCAATTTTGTATTTAGTTCTACAGGTTGTAAAATTAGTATAGAGTATATTTTAAAAGACCTTATAATTAAAGATCCTAATTGTATAGCATTCATAGGTGAAGGAGCAGGGAATTTATTATTGCGTACAGTAGTGGAACTTCATCCCGATATAAGATATATTTACAGAAGTCTGAAGGATTGCAATGATCATAGTTTACCTATTGAGTTTTTAAGGCTGTACAATGGACATATCAACATTGATTATGGTGAAAATTTGACCATTCCTGCTACAGATGCAACCAACAACATTCATTGGTCTTATTTACATATAAAGTTTGCTGAACCTATCAGTCTTTTTGTCTGTGATGCTGAATTGCCTGTAACAGTCAACTGGAGTAAAATTATAATAGAGTGGAGCAAGCATGTAAGAAAATGCAAGTACTGTTCCTCAGTTAATAAATGTACGTTAATAGTAAAATATCATGCTCAAGATGATATCGATTTCAAATTAGACAATATAACTATATTAAAAACTTATGTATGCTTAGGCAGTAAGTTAAAGGGGTCTGAAGTTTACTTAGTCCTTACAATAGGTCCTGCAAATGTGTTCCCAGTATTTAATGTAGTACAAAATGCTAAATTGATACTATCAAGAACCAAAAATTTCATCATGCCTAAGAAGGCTGATAAAGAGTCTATTGATGCAAATATTAAAAGTTTGATACCCTTTCTTTGTTACCCTATAACAAAAAAAGGAATTAATACTGCATTGTCAAAACTAAAGAGTGTTGTTAGTGGAGATATACTATCATATTCTATAGCAGGACGTAATGAAGTTTTCAGCAATAAACTTATAAATCATAAGCATATGAACATCTTAAAATGGTTCAATCATGTTTTAAATTTCAGATCAACAGAACTAAACTATAATCATTTATATATGGTAGAATCTACATATCCTTATCTAAGTGAATTGTTAAACAGCTTGACAACTAATGAACTTAAAAAACTGATTAAAATCACAGGTAGTTTGTTATACAACTTTCATAATGAATAATGAATAAAAATCTTATATTAAAAATTCCCATAGCTACACACTAACACTGTATTCAATTATAGTTATTTAAAATTAAAAATTATATAATTTTTTAATAACTTTTAGTGAACTAATCCTAAAATTATCATTTTGATCTAGGAGGAATAAATTTAAATCCAAATCTAATTGGTTTATATGTATATTAACTAAACTACGAGATATTAGTTTTTGACACTTTTTTTCTCGT'\n",
    "# ref_seq_list = list(ref_seq)\n",
    "# dict_jianji = {}\n",
    "# for k,v in dict_info_feikong.items():\n",
    "#     gene_pos = int(k.split('_')[1])\n",
    "#     ref = k.split('_')[0]\n",
    "#     alt = k.split('_')[2]\n",
    "#     product = v['product']\n",
    "# #     frequency = v['frequency']\n",
    "#     gene_pos_f2 = gene_pos - 2   #前面2个\n",
    "#     gene_pos_f1 = gene_pos - 1   #前面1个\n",
    "#     gene_pos_b1 = gene_pos + 1\n",
    "#     gene_pos_b2 = gene_pos + 2\n",
    "#     start = v['start']\n",
    "#     end = v['end']\n",
    "#     type4 = v['type']\n",
    "#     chazhi = int(gene_pos) - int(start)\n",
    "#     quzheng =  chazhi//3   #取整，以获取是这个蛋白质的第几个氨基酸。\n",
    "#     aa_pos = quzheng + 1\n",
    "\n",
    "# # #分类讨论\n",
    "#     if type4 == 'del' or type4 == 'ins':\n",
    "#         del_count = len(ref) - len(alt)\n",
    "#         if del_count%3 == 0:\n",
    "#             #分类讨论\n",
    "#             #1\n",
    "#             if chazhi%3 == 2 and len(alt)%3 == 2:\n",
    "#                 #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f2)\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f1)\n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b1)\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b2)\n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f2)\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f1)\n",
    "                \n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b1)\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b2)\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "#            #2\n",
    "#             if chazhi%3 == 2 and len(alt)%3 == 1:\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f2)\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f1)\n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "              \n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f2)\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f1)\n",
    "                \n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "           \n",
    "#             #3\n",
    "#             if chazhi%3 == 2 and len(alt)%3 == 0:\n",
    "#             #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f2)\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f1)\n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b1)\n",
    "\n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "\n",
    "\n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f2)\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f1)\n",
    "\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f1-1])\n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b1)\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "            \n",
    "#             #4\n",
    "#             if chazhi%3 == 1 and len(alt)%3 == 2:\n",
    "#                 #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f2)\n",
    "               \n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "              \n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f2)\n",
    "               \n",
    "                \n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                \n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "      \n",
    "#             #5\n",
    "#             if chazhi%3 == 1 and len(alt)%3 == 1:\n",
    "#                 #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f2)\n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b1)\n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f2)\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b1)\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "#         #6\n",
    "#             if chazhi%3 == 1 and len(alt)%3 == 0:\n",
    "#                 #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 list_jianji_pos_ref.append(gene_pos_f2)\n",
    "                \n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b1)\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b2)\n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 list_jianji_pos_alt.append(gene_pos_f2)\n",
    "              \n",
    "                \n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_f2-1])\n",
    "                \n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b1)\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b2)\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "                \n",
    "#             #7\n",
    "#             if chazhi%3 == 0 and len(alt)%3 == 2:\n",
    "#                 #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "       \n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b1)\n",
    "               \n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian]) \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b1)\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "                \n",
    "#             #8\n",
    "#             if chazhi%3 == 0 and len(alt)%3 == 1:\n",
    "#                 #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b1)\n",
    "#                 list_jianji_pos_ref.append(gene_pos_b2)\n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "                    \n",
    "                \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b1)\n",
    "#                 list_jianji_pos_alt.append(gene_pos_b2)\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b1-1])\n",
    "#                 list_jianji_alt.append(ref_seq[gene_pos_b2-1])\n",
    "#         #9\n",
    "#             if chazhi%3 == 0 and len(alt)%3 == 0:\n",
    "#                 #变异前的位置和碱基\n",
    "#                 list_jianji_pos_ref = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_ref = []\n",
    "#                 for index in range(0,len(ref)):\n",
    "#                     list_jianji_pos_ref.append((gene_pos + index))\n",
    "         \n",
    "#                 for p in list_jianji_pos_ref:\n",
    "#                     p_qiepian = p-1\n",
    "#                     list_jianji_ref.append(ref_seq[p_qiepian])\n",
    "               \n",
    "#                 #变异后的位置\n",
    "#                 list_jianji_pos_alt = []#是位置不是切片，切片要-1\n",
    "#                 list_jianji_alt = []\n",
    "#                 for index in range(0,len(alt)):\n",
    "#                     list_jianji_pos_alt.append((gene_pos + index))\n",
    "#                     list_jianji_alt.append(alt[index])\n",
    "      \n",
    " \n",
    "    \n",
    "#   #------------------------原来的，底下可以不变  \n",
    "#     if type4 == 'snp':\n",
    "#         if chazhi%3 == 2:\n",
    "#             list_jianji_pos = []#是位置不是切片，切片要-1\n",
    "#             list_jianji_pos.append(gene_pos_f2)\n",
    "#             list_jianji_pos.append(gene_pos_f1)\n",
    "#             list_jianji_pos.append(gene_pos)\n",
    "\n",
    "\n",
    "\n",
    "#         if chazhi%3 == 1:\n",
    "#             list_jianji_pos = []\n",
    "#             list_jianji_pos.append(gene_pos_f1)\n",
    "#             list_jianji_pos.append(gene_pos)\n",
    "#             list_jianji_pos.append(gene_pos_b1)\n",
    "\n",
    "\n",
    "\n",
    "#         if chazhi%3 == 0:\n",
    "#             list_jianji_pos = []\n",
    "#             list_jianji_pos.append(gene_pos)\n",
    "#             list_jianji_pos.append(gene_pos_b1)\n",
    "#             list_jianji_pos.append(gene_pos_b2)\n",
    "\n",
    "\n",
    "#         list_jianji_qiepian = []#这个是切片\n",
    "#         for i in list_jianji_pos:\n",
    "#             i_1 = i-1\n",
    "#             list_jianji_qiepian.append(i_1)\n",
    "#     #     print(list_jianji_qiepian)\n",
    "\n",
    "#         list_jianji_ref = []#获取参考碱基\n",
    "#         for i in list_jianji_qiepian:\n",
    "#             list_jianji_ref.append(ref_seq_list[i])\n",
    "#     #     print(list_jianji_ref)\n",
    "\n",
    "#         #获取alt碱基\n",
    "#         list_jianji_alt= []#获取参考碱基\n",
    "#         for i in list_jianji_qiepian:\n",
    "#             if i == int(gene_pos - 1):\n",
    "#                 list_jianji_alt.append(alt)\n",
    "#             else:\n",
    "#                 list_jianji_alt.append(ref_seq_list[i])\n",
    "#     #     print(list_jianji_alt)\n",
    "#     #翻译蛋白质\n",
    "#     str_ref = \"\".join(list_jianji_ref)\n",
    "#     str_alt = \"\".join(list_jianji_alt)\n",
    "#     aa_ref = str(Seq(str_ref).translate())\n",
    "#     aa_alt = str(Seq(str_alt).translate())    \n",
    "#     #非同义：\n",
    "#     if aa_alt != aa_ref:\n",
    "#         info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "#         dict_aa_mut[k] = {}\n",
    "#         dict_aa_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "#         dict_aa_mut[k]['aa'] = info\n",
    "#         dict_aa_mut[k]['num'] = dict_mut_num[k]\n",
    "#     #同义\n",
    "#     if aa_alt == aa_ref:\n",
    "#         info = f'{aa_ref}_{product}_{aa_pos}_{aa_alt}'\n",
    "#         dict_aa_synonymous_mut[k] = {}\n",
    "#         dict_aa_synonymous_mut[k]['product'] = dict_info_feikong[k]['product']\n",
    "#         dict_aa_synonymous_mut[k]['aa'] = info\n",
    "#         dict_aa_synonymous_mut[k]['num'] = dict_mut_num[k]\n",
    "# #             dict_aa_mut[k]['frequency'] = dict_info_feikong[k]['frequency']\n",
    "    \n",
    "# # \n",
    "\n",
    "# #非同义----------------------------------------------------------------------\n",
    "# #统计各个突变的频率。\n",
    "# with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_nonsynonymous_mut_frequency.csv','w') as f:\n",
    "#     writer = csv.writer(f)\n",
    "#     writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "#     for k,v in dict_aa_mut.items():\n",
    "#         writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "    \n",
    "    \n",
    "# #统计非同义突变在哪个蛋白上nonsynonymous mutation\n",
    "# list_product = []\n",
    "# for k,v in dict_aa_mut.items():\n",
    "#         if v['product'] not in list_product:\n",
    "#             list_product.append(v['product'])\n",
    "#             dict_product = dict.fromkeys(list_product)#每个蛋白上非同义突变的个数\n",
    "# for k in dict_product.keys():\n",
    "#     dict_product[k] = 0\n",
    "# for m,p in dict_aa_mut.items():\n",
    "#     for k,v in dict_product.items():\n",
    "#         if p['product'] == k:\n",
    "#             dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "# # print(dict_product)\n",
    "# dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "# # print(dict_product)\n",
    "# with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_nonsynonymous_product_number.csv','w') as f:\n",
    "#     writer = csv.writer(f)\n",
    "#     writer.writerow(['product','number','group'])\n",
    "#     for k,v in dict_product.items():\n",
    "#         writer.writerow([k,v,'nonsynonymous'])\n",
    "        \n",
    "\n",
    "# #同义----------------------------------------------------------------------\n",
    "# #统计各个突变的频率。\n",
    "# with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_synonymous_mut_frequency.csv','w') as f:\n",
    "#     writer = csv.writer(f)\n",
    "#     writer.writerow(['isnv','product','aa_mutation','sample_number'])\n",
    "    \n",
    "#     for k,v in dict_aa_synonymous_mut.items():\n",
    "#         writer.writerow([k,v['product'],v['aa'],v['num']])\n",
    "    \n",
    "    \n",
    "# #统计同义突变在哪个蛋白上\n",
    "# list_product = []\n",
    "# for k,v in dict_aa_synonymous_mut.items():\n",
    "#         if v['product'] not in list_product:\n",
    "#             list_product.append(v['product'])\n",
    "#             dict_product = dict.fromkeys(list_product)\n",
    "# for k in dict_product.keys():\n",
    "#     dict_product[k] = 0\n",
    "# for m,p in dict_aa_synonymous_mut.items():\n",
    "#     for k,v in dict_product.items():\n",
    "#         if p['product'] == k:\n",
    "#             dict_product[k] += int(p['num']) #赋值的时候不可以直接写v，要写成k的值\n",
    "# # print(dict_product)\n",
    "# dict_product = dict(sorted(dict_product.items(),key=lambda x:x[0]))\n",
    "# # print(dict_product)\n",
    "# with open('/mnt/alamo01/users/yangjingqi/RSV_sra/group/workflow_result_pair/08product_frequency/08isnv_synonymous_product_number.csv','w') as f:\n",
    "#     writer = csv.writer(f)\n",
    "#     writer.writerow(['product','number','group'])\n",
    "#     for k,v in dict_product.items():\n",
    "#         writer.writerow([k,v,'synonymous'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "093d93fe-0b87-4673-b01d-8795b1206b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "#找出来所有的突变isnv和snp的 09\n",
    "import os\n",
    "import csv\n",
    "prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/')\n",
    "for i in prj_list:\n",
    "    mutation_list = []\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{i}'\n",
    "    for n in os.listdir(prj_path):\n",
    "        csv_path = f'{prj_path}/{n}'\n",
    "        with open(csv_path) as csvfile:\n",
    "            csv_reader = csv.reader(csvfile)  \n",
    "            for row in csv_reader:\n",
    "                if row[0] != 'mutation':\n",
    "                    if row[0] not in mutation_list:\n",
    "                        mutation_list.append(row[0])\n",
    "    with open(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/09snp_isnv_all_mut/{i}.csv','w')as f:\n",
    "        csv_writer = csv.writer(f)\n",
    "        csv_writer.writerow(['mutation'])\n",
    "        for m in mutation_list:\n",
    "            csv_writer.writerow([m])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1681dd14-8078-4313-b130-77f91ee7a4bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #按照reads计算熵值\n",
    "# import csv\n",
    "# import os\n",
    "# prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample')\n",
    "# for i in prj_list:\n",
    "#     list_site_reads_num = []\n",
    "#     prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{i}'\n",
    "#     csv_list = os.listdir(prj_path)\n",
    "#     for n in csv_list:\n",
    "#         csv_path = f'{prj_path}/{n}'\n",
    "#         with open(csv_path) as csvfile:\n",
    "#             csv_reader = csv.reader(csvfile)  \n",
    "#             for row in csv_reader:\n",
    "#                 if row[0] != 'mutation':\n",
    "#                     site_pos = row[0].split('_')[1]\n",
    "#                     ref = row[0].split('_')[0]\n",
    "#                     alt = row[0].split('_')[2]\n",
    "#                     if site_pos not in list_site_reads_num:\n",
    "#                         list_site_reads_num.append(site_pos)\n",
    "#     dict_site_reads_num = dict.fromkeys(list_site_reads_num)\n",
    "#     for k,v in dict_site_reads_num.items():\n",
    "#         dict_site_reads_num[k] = {}\n",
    "        \n",
    "#     for n in csv_list:\n",
    "#         csv_path = f'{prj_path}/{n}'\n",
    "#         with open(csv_path) as csvfile:\n",
    "#             csv_reader = csv.reader(csvfile)  \n",
    "#             for row in csv_reader:\n",
    "#                 if row[0] != 'mutation':\n",
    "#                     site_pos = row[0].split('_')[1]\n",
    "#                     ref = row[0].split('_')[0]\n",
    "#                     alt = row[0].split('_')[2]\n",
    "#                     ref_num = int(row[2])\n",
    "#                     alt_num = int(row[1])\n",
    "#                     for k,v in dict_site_reads_num.items():\n",
    "#                         if ref not in dict_site_reads_num[k]:\n",
    "#                             dict_site_reads_num[k][ref] = ref_num\n",
    "#                         else:\n",
    "#                             dict_site_reads_num[k][ref] += ref_num\n",
    "#                         if alt not in dict_site_reads_num[k]:\n",
    "#                             dict_site_reads_num[k][alt] = alt_num\n",
    "#                         else:\n",
    "#                             dict_site_reads_num[k][alt] += ref_num\n",
    "#     print(dict_site_reads_num)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4da1f951-9cb3-49e6-8762-0cc8b3dbe8d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #按照reads计算熵值\n",
    "# import csv\n",
    "# import os\n",
    "# import math\n",
    "# import matplotlib.pyplot as plt\n",
    "# prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample')\n",
    "# for i in prj_list:\n",
    "#     list_site_reads_num = []\n",
    "#     prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{i}'\n",
    "#     csv_list = os.listdir(prj_path)\n",
    "#     for n in csv_list:\n",
    "#         csv_path = f'{prj_path}/{n}'\n",
    "#         with open(csv_path) as csvfile:\n",
    "#             csv_reader = csv.reader(csvfile)  \n",
    "#             for row in csv_reader:\n",
    "#                 if row[0] != 'mutation':\n",
    "#                     site_pos = row[0].split('_')[1]\n",
    "#                     ref = row[0].split('_')[0]\n",
    "#                     alt = row[0].split('_')[2]\n",
    "#                     if site_pos not in list_site_reads_num:\n",
    "#                         list_site_reads_num.append(site_pos)\n",
    "#     dict_site_reads_num = dict.fromkeys(list_site_reads_num)\n",
    "#     for k,v in dict_site_reads_num.items():\n",
    "#         dict_site_reads_num[k] = {}\n",
    "\n",
    "        \n",
    "        \n",
    "#     for k,v in dict_site_reads_num.items():  \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "#         for n in csv_list:\n",
    "#             csv_path = f'{prj_path}/{n}'\n",
    "#             with open(csv_path) as csvfile:\n",
    "#                 csv_reader = csv.reader(csvfile)  \n",
    "#                 for row in csv_reader:\n",
    "#                     if row[0] != 'mutation':\n",
    "#                         site_pos = row[0].split('_')[1]\n",
    "#                         ref = row[0].split('_')[0]\n",
    "#                         alt = row[0].split('_')[2]\n",
    "#                         ref_num = int(row[2])\n",
    "#                         alt_num = int(row[1])\n",
    "#                         if k == site_pos:\n",
    "#                             if ref not in dict_site_reads_num[k]:\n",
    "#                                 dict_site_reads_num[k][ref] = ref_num\n",
    "#                             else:\n",
    "#                                 dict_site_reads_num[k][ref] += ref_num\n",
    "#                             if alt not in dict_site_reads_num[k]:\n",
    "#                                 dict_site_reads_num[k][alt] = alt_num\n",
    "#                             else:\n",
    "#                                 dict_site_reads_num[k][alt] += ref_num\n",
    "#     # print(dict_site_reads_num)\n",
    "    \n",
    "#     dict_eachsite_entropy = {}\n",
    "#     for k,v in dict_site_reads_num.items():\n",
    "#         each_site_all_reads = 0\n",
    "#         h_all = 0\n",
    "#         for j in v.keys():\n",
    "#             each_site_all_reads += dict_site_reads_num[k][j]\n",
    "#         for j in v.keys():\n",
    "#             p = dict_site_reads_num[k][j]/each_site_all_reads\n",
    "#             if p!= 0:\n",
    "#                 h = -p*math.log2(p)\n",
    "#                 h_all += h\n",
    "#         dict_eachsite_entropy[k] = h_all\n",
    "#     dict_all_site = {}\n",
    "    \n",
    "    \n",
    "#     for s in range(1,15192):\n",
    "#         s1 = str(s)\n",
    "#         dict_all_site[s1] = 0\n",
    "#     for s,e in dict_all_site.items():\n",
    "#         for d,p in dict_eachsite_entropy.items():\n",
    "#             if s == d:\n",
    "#                 dict_all_site[s] = dict_eachsite_entropy[s]\n",
    "                \n",
    "        \n",
    "#     # print(i)\n",
    "#     # print(dict_eachsite_entropy)\n",
    "#     list_entropy =list(dict_all_site.values())\n",
    "#     list_position = list(dict_all_site.keys())\n",
    "#     # print(list_entropy)\n",
    "#     # print(list_position)\n",
    "#     plt.bar(list_position, list_entropy)\n",
    "#     plt.xlabel('position')\n",
    "#     plt.ylabel('entropy')\n",
    "#     plt.title('Entropy Plot')\n",
    "#     plt.show()\n",
    "#     plt.savefig(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/10entropy/{i}_entropy.pdf')\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f0ab570-0ca6-42d7-a892-0be680b2b57c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59445de3-7d87-4048-9a55-70bae45d57cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68778ae8-0bc2-4f13-bea1-c8254e626e52",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "import csv\n",
    "#新建分组的文件夹\n",
    "wenjianjia_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/yjq_sra_bioproject_rsv')\n",
    "for i in wenjianjia_list:\n",
    "    dirs = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/10entropy/{i}'\n",
    "    if not os.path.exists(dirs):\n",
    "        os.makedirs(dirs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a7d4489-99f6-4a28-9afd-bac7b5a28b22",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d20dbdb-5429-4d4f-96d0-84bb3576f1e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#按照reads计算熵值,画柱状图并且加上所有位点包括没有在vcf里出现的\n",
    "#熵图分蛋白。\n",
    "dict_pep_cds = {'99_518':'NS1_1C','628_1002':\"NS2_1B\",'1140_2315':\"N\",'2348_3073':\"P\",'3233_4003':\"M\",'4274_4468':\"SH\",'4659_5555':\"G\",'5632_7356':\"F\",'7576_8160':\"M2\",'8468_14965':\"L\"}\n",
    "list_pep_cds = list(dict_pep_cds.keys())\n",
    "print(list_pep_cds)\n",
    "\n",
    "\n",
    "import csv\n",
    "import os\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "prj_list = os.listdir('/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample')\n",
    "for i in prj_list:\n",
    "    list_site_reads_num = []\n",
    "    prj_path = f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/01filtered_csv_deep_sample/{i}'\n",
    "    csv_list = os.listdir(prj_path)\n",
    "    for n in csv_list:\n",
    "        csv_path = f'{prj_path}/{n}'\n",
    "        with open(csv_path) as csvfile:\n",
    "            csv_reader = csv.reader(csvfile)  \n",
    "            for row in csv_reader:\n",
    "                if row[0] != 'mutation':\n",
    "                    site_pos = row[0].split('_')[1]\n",
    "                    ref = row[0].split('_')[0]\n",
    "                    alt = row[0].split('_')[2]\n",
    "                    if site_pos not in list_site_reads_num:\n",
    "                        list_site_reads_num.append(site_pos)\n",
    "    dict_site_reads_num = dict.fromkeys(list_site_reads_num)\n",
    "    for k,v in dict_site_reads_num.items():\n",
    "        dict_site_reads_num[k] = {}\n",
    "\n",
    "        \n",
    "        \n",
    "    for k,v in dict_site_reads_num.items():  \n",
    "         for n in csv_list:\n",
    "            csv_path = f'{prj_path}/{n}'\n",
    "            with open(csv_path) as csvfile:\n",
    "                csv_reader = csv.reader(csvfile)  \n",
    "                for row in csv_reader:\n",
    "                    if row[0] != 'mutation':\n",
    "                        site_pos = row[0].split('_')[1]\n",
    "                        ref = row[0].split('_')[0]\n",
    "                        alt = row[0].split('_')[2]\n",
    "                        ref_num = int(row[2])\n",
    "                        alt_num = int(row[1])\n",
    "                        if k == site_pos:\n",
    "                            if ref not in dict_site_reads_num[k]:\n",
    "                                dict_site_reads_num[k][ref] = ref_num\n",
    "                            else:\n",
    "                                dict_site_reads_num[k][ref] += ref_num\n",
    "                            if alt not in dict_site_reads_num[k]:\n",
    "                                dict_site_reads_num[k][alt] = alt_num\n",
    "                            else:\n",
    "                                dict_site_reads_num[k][alt] += ref_num\n",
    "    # print(dict_site_reads_num)\n",
    "    \n",
    "    \n",
    "    list_mutation_kind = []\n",
    "    for k,v in dict_site_reads_num.items():\n",
    "        for j,num in v.items():\n",
    "            if j not in list_mutation_kind:\n",
    "                list_mutation_kind.append(j)\n",
    "    print(list_mutation_kind)\n",
    "    \n",
    "    dict_each_pep_entropy = dict.fromkeys(list_pep_cds)\n",
    "    # print(dict_each_pep_entropy)\n",
    "    for k,v in dict_each_pep_entropy.items():\n",
    "        dict_each_pep_entropy[k] = dict.fromkeys(list_mutation_kind)\n",
    "        \n",
    "    for k,v in dict_each_pep_entropy.items():\n",
    "        for p,s in v.items():\n",
    "            dict_each_pep_entropy[k][p] = 0\n",
    "        \n",
    "        \n",
    "    # print(dict_each_pep_entropy)\n",
    "    for k,v in dict_each_pep_entropy.items():\n",
    "        if len(v) !=0:\n",
    "            for y,r in v.items():\n",
    "                start = k.split('_')[0]\n",
    "                end = k.split('_')[1]\n",
    "                for s,p in dict_site_reads_num.items():\n",
    "                    for j,num in p.items():\n",
    "                        if int(start) <= int(s) <= int(end):\n",
    "                            # print(j,y)\n",
    "                            if j==y:\n",
    "                                dict_each_pep_entropy[k][j] += num \n",
    "    # print(dict_each_pep_entropy)\n",
    "    \n",
    "    dict_entropy = {}\n",
    "    for k,v in dict_each_pep_entropy.items():\n",
    "        if len(v) !=0:\n",
    "            each_pep_all_reads = 0\n",
    "            h_all = 0\n",
    "            for j in v.keys():\n",
    "                each_pep_all_reads += dict_each_pep_entropy[k][j]\n",
    "                # print(each_pep_all_reads)\n",
    "            if each_pep_all_reads!=0:\n",
    "                for j in v.keys():\n",
    "                    p = dict_each_pep_entropy[k][j]/each_pep_all_reads\n",
    "                    if p!= 0:\n",
    "                        h = -p*math.log2(p)\n",
    "                        h_all += h\n",
    "                dict_entropy[k] = h_all\n",
    "    # print(dict_entropy)\n",
    "    \n",
    "    \n",
    "    dict_entropy_all_pep = dict.fromkeys(list_pep_cds)\n",
    "    for k,v in dict_entropy_all_pep.items():\n",
    "        dict_entropy_all_pep[k] = 0\n",
    "    \n",
    "    for k,v in dict_entropy_all_pep.items():\n",
    "        for d,p in dict_entropy.items():\n",
    "            if k == d:\n",
    "                dict_entropy_all_pep[k] = dict_entropy[k]\n",
    "    # print(dict_entropy_all_pep)            \n",
    "        \n",
    "    # print(i)\n",
    "    # print(dict_eachsite_entropy)\n",
    "    list_entropy =list(dict_entropy_all_pep.values())\n",
    "    list_position = list(dict_entropy_all_pep.keys())\n",
    "    print(list_entropy)\n",
    "    print(list_position)\n",
    "    plt.figure(figsize=(12, 6))\n",
    "    plt.bar(list_position, list_entropy)\n",
    "    plt.xlabel('position')\n",
    "    plt.ylabel('entropy')\n",
    "    plt.title('Entropy Plot')\n",
    "    plt.show()\n",
    "    plt.savefig(f'/mnt/alamo01/users/yangjingqi/RSV_sra/group/08workflow_result_pair_each_project/11entropy_each_pep/{i}_entropy.pdf')\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3922f4c-d32b-4088-8c46-ed17e7bc96f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_mutation_kind"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
